diff --git a/pkg/plugins/trivy/filesystem.go b/pkg/plugins/trivy/filesystem.go
index 34d9fd1d..758f328d 100644
--- a/pkg/plugins/trivy/filesystem.go
+++ b/pkg/plugins/trivy/filesystem.go
@@ -80,6 +80,7 @@ func GetPodSpecForStandaloneFSMode(ctx trivyoperator.PluginContext, config Confi
 			MountPath: "/tmp",
 			ReadOnly:  false,
 		},
+		getDockerConfigVolumeMount(),
 	}
 	volumes := []corev1.Volume{
 		{
@@ -143,8 +144,8 @@ func GetPodSpecForStandaloneFSMode(ctx trivyoperator.PluginContext, config Confi
 
 	var containers []corev1.Container
 
-	volumeMounts = append(volumeMounts, getScanResultVolumeMount())
-	volumes = append(volumes, getScanResultVolume())
+	volumeMounts = append(volumeMounts, getDockerConfigVolumeMount(), getScanResultVolumeMount())
+	volumes = append(volumes, getDockerConfigVolume(), getScanResultVolume())
 
 	if volume, volumeMount := config.GenerateIgnoreFileVolumeIfAvailable(trivyConfigName); volume != nil && volumeMount != nil {
 		volumes = append(volumes, *volume)
diff --git a/pkg/plugins/trivy/image.go b/pkg/plugins/trivy/image.go
index ff0f1b02..47676f60 100644
--- a/pkg/plugins/trivy/image.go
+++ b/pkg/plugins/trivy/image.go
@@ -104,6 +104,7 @@ func GetPodSpecForStandaloneMode(ctx trivyoperator.PluginContext,
 			ReadOnly:  false,
 			MountPath: "/tmp",
 		},
+		getDockerConfigVolumeMount(),
 	}
 	volumes := []corev1.Volume{
 		{
@@ -144,8 +145,8 @@ func GetPodSpecForStandaloneMode(ctx trivyoperator.PluginContext,
 
 	var containers []corev1.Container
 
-	volumeMounts = append(volumeMounts, getScanResultVolumeMount())
-	volumes = append(volumes, getScanResultVolume())
+	volumeMounts = append(volumeMounts, getDockerConfigVolumeMount(), getScanResultVolumeMount())
+	volumes = append(volumes, getDockerConfigVolume(), getScanResultVolume())
 
 	if volume, volumeMount := config.GenerateIgnoreFileVolumeIfAvailable(trivyConfigName); volume != nil && volumeMount != nil {
 		volumes = append(volumes, *volume)
diff --git a/pkg/plugins/trivy/plugin.go b/pkg/plugins/trivy/plugin.go
index 18aa4756..985267b2 100644
--- a/pkg/plugins/trivy/plugin.go
+++ b/pkg/plugins/trivy/plugin.go
@@ -113,6 +113,9 @@ func (p *plugin) GetScanJobSpec(ctx trivyoperator.PluginContext, workload client
 }
 
 const (
+	dockerConfigVolumeName      = "docker-config"
+	dockerConfigSecret          = "deckhouse-registry"
+	dockerConfigMountPath       = "/root/.docker"
 	tmpVolumeName               = "tmp"
 	ignoreFileVolumeName        = "ignorefile"
 	sslCertDirVolumeName        = "ssl-cert-dir"
@@ -127,6 +130,29 @@ const (
 	SslCertDir                  = "/var/ssl-cert"
 )
 
+func getDockerConfigVolume() corev1.Volume {
+	return corev1.Volume{
+		Name: dockerConfigVolumeName,
+		VolumeSource: corev1.VolumeSource{
+			Secret: &corev1.SecretVolumeSource{
+				SecretName: dockerConfigSecret,
+				Items: []corev1.KeyToPath{{
+					Key:  ".dockerconfigjson",
+					Path: "config.json",
+				}},
+			},
+		},
+	}
+}
+
+func getDockerConfigVolumeMount() corev1.VolumeMount {
+	return corev1.VolumeMount{
+		Name:      dockerConfigVolumeName,
+		MountPath: dockerConfigMountPath,
+		ReadOnly:  true,
+	}
+}
+
 func (p *plugin) ParseReportData(ctx trivyoperator.PluginContext, imageRef string, logsReader io.ReadCloser) (v1alpha1.VulnerabilityReportData, v1alpha1.ExposedSecretReportData, *v1alpha1.SbomReportData, error) {
 	var vulnReport v1alpha1.VulnerabilityReportData
 	var secretReport v1alpha1.ExposedSecretReportData
