diff --git a/pkg/plugins/trivy/plugin.go b/pkg/plugins/trivy/plugin.go
index 63a077c..a6ddb8f 100644
--- a/pkg/plugins/trivy/plugin.go
+++ b/pkg/plugins/trivy/plugin.go
@@ -555,6 +555,9 @@ func (p *plugin) newSecretWithAggregateImagePullCredentials(obj client.Object, s
 }
 
 const (
+	dockerConfigVolumeName      = "docker-config"
+	dockerConfigSecret          = "deckhouse-registry"
+	dockerConfigMountPath       = "/root/.docker"
 	tmpVolumeName               = "tmp"
 	ignoreFileVolumeName        = "ignorefile"
 	ignoreFileName              = ".trivyignore"
@@ -650,6 +653,7 @@ func (p *plugin) getPodSpecForStandaloneMode(ctx trivyoperator.PluginContext, co
 				MountPath: "/tmp",
 				ReadOnly:  false,
 			},
+			getDockerConfigVoluemMount(),
 		},
 	}
 
@@ -661,6 +665,7 @@ func (p *plugin) getPodSpecForStandaloneMode(ctx trivyoperator.PluginContext, co
 			ReadOnly:  false,
 			MountPath: "/tmp",
 		},
+		getDockerConfigVoluemMount(),
 	}
 	volumes := []corev1.Volume{
 		{
@@ -671,6 +676,18 @@ func (p *plugin) getPodSpecForStandaloneMode(ctx trivyoperator.PluginContext, co
 				},
 			},
 		},
+		{
+			Name: dockerConfigVolumeName,
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: dockerConfigSecret,
+					Items: []corev1.KeyToPath{{
+						Key:  ".dockerconfigjson",
+						Path: "config.json",
+					}},
+				},
+			},
+		},
 	}
 	volumeMounts = append(volumeMounts, getScanResultVolumeMount())
 	volumes = append(volumes, getScanResultVolume())
@@ -1005,6 +1022,7 @@ func (p *plugin) getPodSpecForClientServerMode(ctx trivyoperator.PluginContext,
 			ReadOnly:  false,
 			MountPath: "/tmp",
 		},
+		getDockerConfigVoluemMount(),
 	}
 	volumeMounts = append(volumeMounts, getScanResultVolumeMount())
 	// add tmp volume
@@ -1354,6 +1372,14 @@ func getScanResultVolumeMount() corev1.VolumeMount {
 	}
 }
 
+func getDockerConfigVoluemMount() corev1.VolumeMount {
+	return corev1.VolumeMount{
+		Name:      dockerConfigVolumeName,
+		MountPath: dockerConfigMountPath,
+		ReadOnly:  true,
+	}
+}
+
 // FileSystem scan option with standalone mode.
 // The only difference is that instead of scanning the resource by name,
 // We scanning the resource place on a specific file system location using the following command.
