diff --git a/pkg/plugins/trivy/plugin.go b/pkg/plugins/trivy/plugin.go
index 63a077c..62a97df 100644
--- a/pkg/plugins/trivy/plugin.go
+++ b/pkg/plugins/trivy/plugin.go
@@ -555,6 +555,9 @@ func (p *plugin) newSecretWithAggregateImagePullCredentials(obj client.Object, s
 }
 
 const (
+	dockerConfigVolumeName      = "docker-config"
+	dockerConfigSecret          = "deckhouse-registry"
+	dockerConfigMountPath       = "/root/.docker"
 	tmpVolumeName               = "tmp"
 	ignoreFileVolumeName        = "ignorefile"
 	ignoreFileName              = ".trivyignore"
@@ -650,6 +653,11 @@ func (p *plugin) getPodSpecForStandaloneMode(ctx trivyoperator.PluginContext, co
 				MountPath: "/tmp",
 				ReadOnly:  false,
 			},
+			{
+				Name:      dockerConfigVolumeName,
+				MountPath: dockerConfigMountPath,
+				ReadOnly:  true,
+			},
 		},
 	}
 
@@ -671,6 +679,18 @@ func (p *plugin) getPodSpecForStandaloneMode(ctx trivyoperator.PluginContext, co
 				},
 			},
 		},
+		{
+			Name: dockerConfigVolumeName,
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: dockerConfigSecret,
+					Items: []corev1.KeyToPath{{
+						Key:  ".dockerconfigjson",
+						Path: "config.json",
+					}},
+				},
+			},
+		},
 	}
 	volumeMounts = append(volumeMounts, getScanResultVolumeMount())
 	volumes = append(volumes, getScanResultVolume())
