diff --git a/pkg/configauditreport/controller/helper.go b/pkg/configauditreport/controller/helper.go
index a9fcaa8..24c4009 100644
--- a/pkg/configauditreport/controller/helper.go
+++ b/pkg/configauditreport/controller/helper.go
@@ -2,7 +2,9 @@ package controller
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
+	"os"
 	"strings"
 
 	"github.com/aquasecurity/defsec/pkg/scan"
@@ -60,6 +62,15 @@ func evaluate(ctx context.Context, policies *policy.Policies, resource client.Ob
 				continue
 			}
 			if k8sCoreComponent(resource) {
+				fmt.Fprint(os.Stdout, "\n\n\n\n\n\n\n\n\n\n\n")
+				r, _ := json.Marshal(resource)
+				fmt.Fprint(os.Stdout, string(r))
+
+				for _, in := range inputs {
+					fmt.Fprint(os.Stdout, "\n\n")
+					fmt.Fprint(os.Stdout, string(in))
+				}
+				fmt.Fprint(os.Stdout, "\n\n\n\n\n\n\n\n\n\n\n")
 				infraChecks = append(infraChecks, getCheck(result, id))
 			}
 			continue
