diff --git a/vendor/github.com/aquasecurity/trivy-kubernetes/pkg/jobs/builder.go b/vendor/github.com/aquasecurity/trivy-kubernetes/pkg/jobs/builder.go
index 250210c7..0c398ace 100644
--- a/vendor/github.com/aquasecurity/trivy-kubernetes/pkg/jobs/builder.go
+++ b/vendor/github.com/aquasecurity/trivy-kubernetes/pkg/jobs/builder.go
@@ -130,8 +130,10 @@ func (b *JobBuilder) build() (*batchv1.Job, error) {
 		job.Spec.Template.Spec.Containers[0].Image = b.imageRef
 	}
 
-	if len(b.nodeSelector) > 0 {
-		job.Spec.Template.Spec.NodeName = b.nodeSelector
+	if b.nodeSelector != "" {
+		job.Spec.Template.Spec.NodeSelector = map[string]string{
+			corev1.LabelHostname: b.nodeSelector,
+		}
 	}
 	// append lables
 	for key, val := range b.labels {
