diff --git a/pkg/metrics/collector.go b/pkg/metrics/collector.go
index 9a571562..dbdc4101 100644
--- a/pkg/metrics/collector.go
+++ b/pkg/metrics/collector.go
@@ -42,6 +42,7 @@ const (
 	severity           = "severity"
 	vuln_id            = "vuln_id"
 	vuln_title         = "vuln_title"
+	vuln_url           = "vuln_url"
 	vuln_score         = "vuln_score"
 	// compliance
 	title       = "title"
@@ -281,6 +282,7 @@ func buildMetricDescriptors(config trivyoperator.ConfigData) metricDescriptors {
 		vuln_id,
 		vuln_title,
 		vuln_score,
+		vuln_url,
 	}
 	vulnIdLabels = append(vulnIdLabels, dynamicLabels...)
 	exposedSecretLabels := []string{
@@ -668,7 +670,7 @@ func (c ResourcesMetricsCollector) collectVulnerabilityIdReports(ctx context.Con
 				vulnLabelValues[7] = r.Report.Artifact.Tag
 				vulnLabelValues[8] = r.Report.Artifact.Digest
 				for i, label := range c.GetReportResourceLabels() {
-					vulnLabelValues[i+22] = r.Labels[label]
+					vulnLabelValues[i+23] = r.Labels[label]
 				}
 				var vulnList = make(map[string]bool)
 				for _, vuln := range r.Report.Vulnerabilities {
@@ -693,6 +695,12 @@ func (c ResourcesMetricsCollector) collectVulnerabilityIdReports(ctx context.Con
 					if vuln.Score != nil {
 						vulnLabelValues[21] = strconv.FormatFloat(*vuln.Score, 'f', -1, 64)
 					}
+					if strings.HasPrefix(vuln.VulnerabilityID, "BDU") && len(vuln.VulnerabilityID) > 4 {
+						vulnLabelValues[22] = "https://bdu.fstec.ru/vul/" + vuln.VulnerabilityID[4:]
+						vulnLabelValues[20] = vuln.VulnerabilityID
+					} else {
+						vulnLabelValues[22] = vuln.PrimaryLink
+					}
 					metrics <- prometheus.MustNewConstMetric(c.vulnIdDesc, prometheus.GaugeValue, float64(1), vulnLabelValues...)
 				}
 			}

