diff --git a/pkg/operator/ttl_report.go b/pkg/operator/ttl_report.go
index 88c8ba11..88b098cb 100644
--- a/pkg/operator/ttl_report.go
+++ b/pkg/operator/ttl_report.go
@@ -51,6 +51,9 @@ func (r *TTLReportReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	if r.Config.InfraAssessmentScannerEnabled {
 		ttlResources = append(ttlResources, kube.Resource{ForObject: &v1alpha1.InfraAssessmentReport{}})
 	}
+	if r.Config.SbomGenerationEnable {
+		ttlResources = append(ttlResources, kube.Resource{ForObject: &v1alpha1.SbomReport{}})
+	}
 	installModePredicate, err := predicate.InstallModePredicate(r.Config)
 	if err != nil {
 		return err
@@ -114,7 +117,7 @@ func (r *TTLReportReconciler) DeleteReportIfExpired(ctx context.Context, namespa
 
 func (r *TTLReportReconciler) applicableForDeletion(report client.Object, ttlReportAnnotationStr string) bool {
 	reportKind := report.GetObjectKind().GroupVersionKind().Kind
-	if reportKind == "VulnerabilityReport" || reportKind == "ExposedSecretReport" {
+	if reportKind == "VulnerabilityReport" || reportKind == "ExposedSecretReport" || reportKind == "SbomReport" {
 		return true
 	}
 	if ttlReportAnnotationStr == time.Duration(0).String() { // check if it marked as historical report
