diff --git a/pkg/vulnerabilityreport/builder.go b/pkg/vulnerabilityreport/builder.go
index 9913d229..bb9da2ec 100644
--- a/pkg/vulnerabilityreport/builder.go
+++ b/pkg/vulnerabilityreport/builder.go
@@ -166,11 +166,11 @@ func (s *ScanJobBuilder) Get() (*batchv1.Job, []*corev1.Secret, error) {
 		templateSpec.Volumes = append(templateSpec.Volumes, s.customVolumes...)
 	}
 	if len(s.customVolumesMount) > 0 {
-		for _, container := range templateSpec.Containers {
-			container.VolumeMounts = append(container.VolumeMounts, s.customVolumesMount...)
+		for i := range templateSpec.Containers {
+			templateSpec.Containers[i].VolumeMounts = append(templateSpec.Containers[i].VolumeMounts, s.customVolumesMount...)
 		}
-		for _, initContainer := range templateSpec.InitContainers {
-			initContainer.VolumeMounts = append(initContainer.VolumeMounts, s.customVolumesMount...)
+		for i := range templateSpec.InitContainers {
+			templateSpec.InitContainers[i].VolumeMounts = append(templateSpec.InitContainers[i].VolumeMounts, s.customVolumesMount...)
 		}
 	}
 	templateSpec.PriorityClassName = s.podPriorityClassName
