Subject: [PATCH] bugfixes
---
Index: cmd/ovirt-csi-driver/ovirt-csi-driver.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/cmd/ovirt-csi-driver/ovirt-csi-driver.go b/cmd/ovirt-csi-driver/ovirt-csi-driver.go
--- a/cmd/ovirt-csi-driver/ovirt-csi-driver.go	(revision 9811e615e93e6eecd8a6b1ca5821885ed2349cfa)
+++ b/cmd/ovirt-csi-driver/ovirt-csi-driver.go	(revision 15b89feaf875f3d732c2458b71e67f63008fa736)
@@ -7,8 +7,6 @@
 	"os"
 	"time"
 
-	"github.com/ovirt/csi-driver/internal/ovirt"
-	"github.com/ovirt/csi-driver/pkg/service"
 	ovirtclient "github.com/ovirt/go-ovirt-client/v3"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
@@ -16,6 +14,9 @@
 	"sigs.k8s.io/controller-runtime/pkg/client/config"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+
+	"github.com/ovirt/csi-driver/internal/ovirt"
+	"github.com/ovirt/csi-driver/pkg/service"
 )
 
 var (
@@ -26,8 +27,8 @@
 )
 
 func init() {
+	klog.InitFlags(flag.CommandLine)
 	flag.Set("logtostderr", "true")
-	klog.InitFlags(flag.CommandLine)
 }
 
 func main() {
Index: internal/ovirt/ovirt.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/ovirt/ovirt.go b/internal/ovirt/ovirt.go
--- a/internal/ovirt/ovirt.go	(revision 9811e615e93e6eecd8a6b1ca5821885ed2349cfa)
+++ b/internal/ovirt/ovirt.go	(revision 15b89feaf875f3d732c2458b71e67f63008fa736)
@@ -23,6 +23,9 @@
 
 func NewClient() (ovirtclient.Client, error) {
 	ovirtConfig, err := GetOvirtConfig()
+	if err != nil {
+		return nil, err
+	}
 	tls := ovirtclient.TLS()
 	if ovirtConfig.Insecure {
 		tls.Insecure()
