diff --git a/controllers/openstackmachine_controller.go b/controllers/openstackmachine_controller.go
index 5e70dc7..fc8918f 100644
--- a/controllers/openstackmachine_controller.go
+++ b/controllers/openstackmachine_controller.go
@@ -30,7 +30,6 @@ import (
 	"k8s.io/client-go/tools/record"
 	"k8s.io/utils/ptr"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
-	ipamv1 "sigs.k8s.io/cluster-api/exp/ipam/api/v1beta1"
 	"sigs.k8s.io/cluster-api/util"
 	"sigs.k8s.io/cluster-api/util/annotations"
 	"sigs.k8s.io/cluster-api/util/conditions"
@@ -210,10 +209,6 @@ func (r *OpenStackMachineReconciler) SetupWithManager(ctx context.Context, mgr c
 		// with the ClusterUnpausedAndInfrastructureReady predicate. The only
 		// fields in OpenStackCluster we are interested in are dependent on
 		// InfrastructureReady, so we don't need to watch both.
-		Watches(
-			&ipamv1.IPAddressClaim{},
-			handler.EnqueueRequestForOwner(mgr.GetScheme(), mgr.GetRESTMapper(), &infrav1.OpenStackMachine{}),
-		).
 		Watches(
 			&infrav1alpha1.OpenStackServer{},
 			handler.EnqueueRequestForOwner(mgr.GetScheme(), mgr.GetRESTMapper(), &infrav1.OpenStackMachine{}),
diff --git a/main.go b/main.go
index 54dbf00..e368f6f 100644
--- a/main.go
+++ b/main.go
@@ -318,16 +318,6 @@ func setupReconcilers(ctx context.Context, mgr ctrl.Manager, caCerts []byte) {
 		setupLog.Error(err, "unable to create controller", "controller", "OpenStackMachine")
 		os.Exit(1)
 	}
-	if err := (&controllers.OpenStackFloatingIPPoolReconciler{
-		Client:         mgr.GetClient(),
-		Recorder:       mgr.GetEventRecorderFor("floatingippool-controller"),
-		ScopeFactory:   scopeFactory,
-		Scheme:         mgr.GetScheme(),
-		CaCertificates: caCerts,
-	}).SetupWithManager(ctx, mgr); err != nil {
-		setupLog.Error(err, "unable to create controller", "controller", "FloatingIPPool")
-		os.Exit(1)
-	}
 	if err := (&controllers.OpenStackServerReconciler{
 		Client:           mgr.GetClient(),
 		Recorder:         mgr.GetEventRecorderFor("openstackserver-controller"),
