diff --git a/cmd/nginx/main.go b/cmd/nginx/main.go
index b21e1012e..15392f0f2 100644
--- a/cmd/nginx/main.go
+++ b/cmd/nginx/main.go
@@ -325,8 +325,9 @@ func registerProfiler() {
 }

 func startHTTPServer(port int, mux *http.ServeMux) {
+	podIP := os.Getenv("POD_IP")
 	server := &http.Server{
-		Addr:              fmt.Sprintf(":%v", port),
+		Addr:              fmt.Sprintf("%s:%v", podIP, port),
 		Handler:           mux,
 		ReadTimeout:       10 * time.Second,
 		ReadHeaderTimeout: 10 * time.Second,
