--- cluster-autoscaler/core/utils.go	(revision 00e3c3e9bbccdc663dbb2927584cc0df130d9e91)
+++ cluster-autoscaler/core/utils.go	(date 1588150882598)
@@ -429,7 +429,9 @@
 		}
 		if incorrectSize.FirstObserved.Add(context.MaxNodeProvisionTime).Before(currentTime) {
 			delta := incorrectSize.CurrentSize - incorrectSize.ExpectedSize
-			if delta < 0 {
+			if (nodeGroup.MinSize() + delta) < nodeGroup.MinSize() {
+				glog.V(2).Infof("Skipping NodeGroupSize fixing, since it would violate NodeGroups's MinSize %q", nodeGroup.MinSize())
+			} else if delta < 0 {
 				glog.V(0).Infof("Decreasing size of %s, expected=%d current=%d delta=%d", nodeGroup.Id(),
 					incorrectSize.ExpectedSize,
 					incorrectSize.CurrentSize,
