Index: cluster-autoscaler/cloudprovider/mcm/mcm_manager.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/cluster-autoscaler/cloudprovider/mcm/mcm_manager.go b/cluster-autoscaler/cloudprovider/mcm/mcm_manager.go
--- a/cluster-autoscaler/cloudprovider/mcm/mcm_manager.go	(date 1615885420194)
+++ b/cluster-autoscaler/cloudprovider/mcm/mcm_manager.go	(date 1615885420194)
@@ -540,9 +540,6 @@
 	if !machine.GetDeletionTimestamp().IsZero() {
 		klog.Infof("Machine %q is already being terminated, and hence skipping the deletion", machine.Name)
 		return true
-	} else if machine.Annotations != nil && machine.Annotations[machinePriorityAnnotation] == "1" {
-		klog.Infof("Machine %q 's priority is set to 1, we assume it to be triggered for deletion by autoscaler earlier, hence skipping deletion", machine.Name)
-		return true
 	}
 	return false
 }
