diff --git a/cmd/nginx/main.go b/cmd/nginx/main.go
index 781f3a8eb..82bb86fb5 100644
--- a/cmd/nginx/main.go
+++ b/cmd/nginx/main.go
@@ -122,6 +122,7 @@ func main() {
 
 	reg := prometheus.NewRegistry()
 
+	reg.MustRegister(nginx.GeoIPErrors)
 	reg.MustRegister(collectors.NewGoCollector())
 	reg.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{
 		PidFn:        func() (int, error) { return os.Getpid(), nil },
diff --git a/internal/nginx/maxmind.go b/internal/nginx/maxmind.go
index ba0ce42b4..24c1b10a1 100644
--- a/internal/nginx/maxmind.go
+++ b/internal/nginx/maxmind.go
@@ -30,6 +30,7 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/prometheus/client_golang/prometheus"
 	"k8s.io/apimachinery/pkg/util/wait"
 	klog "k8s.io/klog/v2"
 )
@@ -55,6 +56,15 @@ var MaxmindRetriesTimeout = time.Second * 0
 // minimumRetriesCount minimum value of the MaxmindRetriesCount parameter. If MaxmindRetriesCount less than minimumRetriesCount, it will be set to minimumRetriesCount
 const minimumRetriesCount = 1
 
+// GeoIPErrors indicates the current GeoIP download error state.
+var GeoIPErrors = prometheus.NewGaugeVec(
+	prometheus.GaugeOpts{
+		Name: "geoip_download_error",
+		Help: "Whether the last GeoIP download attempt failed (1) or succeeded (0).",
+	},
+	[]string{"reason", "type"},
+)
+
 const (
 	geoIPPath   = "/etc/ingress-controller/geoip"
 	dbExtension = ".mmdb"
@@ -116,6 +126,12 @@ func DownloadGeoLite2DB(attempts int, period time.Duration) error {
 			klog.V(2).Infof("Attempting to download GeoIP DB: %s", dbName)
 			dlError = downloadDatabase(dbName)
 			if dlError != nil {
+				trimmedError := []rune(dlError.Error())
+				if len(trimmedError) > 64 {
+					trimmedError = trimmedError[:64]
+				}
+				stringErr := string(trimmedError)
+				GeoIPErrors.WithLabelValues(stringErr, "download").Set(1)
 				break
 			}
 		}
@@ -123,6 +139,7 @@ func DownloadGeoLite2DB(attempts int, period time.Duration) error {
 		lastErr = dlError
 		if dlError == nil {
 			klog.Infof("GeoLite2 DBs downloaded successfully")
+			GeoIPErrors.Reset()
 			return true, nil
 		}
 
diff --git a/pkg/flags/flags.go b/pkg/flags/flags.go
index ce24160fd..43e293e2c 100644
--- a/pkg/flags/flags.go
+++ b/pkg/flags/flags.go
@@ -411,6 +411,8 @@ https://blog.maxmind.com/2019/12/significant-changes-to-accessing-and-using-geol
 			klog.InfoS("downloading maxmind GeoIP2 databases")
 			if err = nginx.DownloadGeoLite2DB(nginx.MaxmindRetriesCount, nginx.MaxmindRetriesTimeout); err != nil {
 				klog.ErrorS(err, "unexpected error downloading GeoIP2 database")
+				// Do not fail controller startup when GeoIP download fails.
+				err = nil
 			}
 		}
 		config.MaxmindEditionFiles = &nginx.MaxmindEditionFiles
