diff --git a/cmake/modules/grpc.cmake b/cmake/modules/grpc.cmake
index 288cf27f..9e0d36a1 100644
--- a/cmake/modules/grpc.cmake
+++ b/cmake/modules/grpc.cmake
@@ -181,6 +181,10 @@ else()
 			sh
 			-c
 			"sed -i '20s/^/#include <cstdint>/' ${GRPC_SRC}/third_party/abseil-cpp/absl/strings/internal/str_format/extension.h"
+                        &&
+                        sh
+                        -c
+                        "sed -i 's|off64_t|off_t|g' ${GRPC_SRC}/third_party/abseil-cpp/absl/base/internal/direct_mmap.h"
 		)

 		# Zig workaround: Add a PATCH_COMMAND to grpc cmake to fixup emitted -march by abseil-cpp
