Index: pkg/util/fieldindex/register.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/util/fieldindex/register.go b/pkg/util/fieldindex/register.go
--- a/pkg/util/fieldindex/register.go	(revision f41e27d4bfb63f0aa0addb35889bae3d43aede56)
+++ b/pkg/util/fieldindex/register.go	(date 1680215160399)
@@ -23,7 +23,6 @@
 	"sigs.k8s.io/controller-runtime/pkg/client"

 	appsv1alpha1 "github.com/openkruise/kruise/apis/apps/v1alpha1"
-	utildiscovery "github.com/openkruise/kruise/pkg/util/discovery"
 	batchv1 "k8s.io/api/batch/v1"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -70,18 +69,6 @@
 		if err = indexJob(c); err != nil {
 			return
 		}
-		// broadcastjob owner
-		if utildiscovery.DiscoverObject(&appsv1alpha1.BroadcastJob{}) {
-			if err = indexBroadcastCronJob(c); err != nil {
-				return
-			}
-		}
-		// imagepulljob active
-		if utildiscovery.DiscoverObject(&appsv1alpha1.ImagePullJob{}) {
-			if err = indexImagePullJobActive(c); err != nil {
-				return
-			}
-		}
 	})
 	return err
 }
