diff --git a/cluster-autoscaler/utils/daemonset/daemonset.go b/cluster-autoscaler/utils/daemonset/daemonset.go
index cf7e60582..a13ee0ed0 100644
--- a/cluster-autoscaler/utils/daemonset/daemonset.go
+++ b/cluster-autoscaler/utils/daemonset/daemonset.go
@@ -19,8 +19,10 @@ package daemonset
 import (
 	"fmt"
 	"math/rand"
+	"strings"
 
 	"k8s.io/autoscaler/cluster-autoscaler/simulator"
+	"k8s.io/klog"
 
 	appsv1 "k8s.io/api/apps/v1"
 	apiv1 "k8s.io/api/core/v1"
@@ -62,14 +64,36 @@ func GetDaemonSetPodsForNode(nodeInfo *schedulerframework.NodeInfo, daemonsets [
 	return result, nil
 }
 
+func shouldEvictDsPod(pod *apiv1.Pod, evictByDefault bool) bool {
+	if strings.HasPrefix(pod.Namespace, "d8-") {
+		if val, ok := pod.Annotations[EnableDsEvictionKey]; !ok || val != "true" {
+			klog.Infof("DaemonSet Pod %s/%s not eligible for eviction: namespace starts with 'd8-' and eviction not enabled", pod.Namespace, pod.Name)
+			return false
+		}
+	}
+
+	if a, ok := pod.Annotations[EnableDsEvictionKey]; ok {
+		if a == "true" {
+			klog.Infof("DaemonSet Pod %s/%s eligible for eviction: annotation %s=true", pod.Namespace, pod.Name, EnableDsEvictionKey)
+			return true
+		}
+		klog.Infof("DaemonSet Pod %s/%s not eligible for eviction: annotation %s=false", pod.Namespace, pod.Name, EnableDsEvictionKey)
+		return false
+	}
+
+	if evictByDefault {
+		klog.Infof("DaemonSet Pod %s/%s eligible for eviction: no annotation %s but evictByDefault is true", pod.Namespace, pod.Name, EnableDsEvictionKey)
+		return true
+	}
+
+	klog.Infof("DaemonSet Pod %s/%s not eligible for eviction: no annotation %s and evictByDefault is false", pod.Namespace, pod.Name, EnableDsEvictionKey)
+	return false
+}
+
 // PodsToEvict returns a list of DaemonSet pods that should be evicted during scale down.
 func PodsToEvict(pods []*apiv1.Pod, evictByDefault bool) (evictable []*apiv1.Pod) {
 	for _, pod := range pods {
-		if a, ok := pod.Annotations[EnableDsEvictionKey]; ok {
-			if a == "true" {
-				evictable = append(evictable, pod)
-			}
-		} else if evictByDefault {
+		if shouldEvictDsPod(pod, evictByDefault) {
 			evictable = append(evictable, pod)
 		}
 	}
