diff --git a/pkg/controller/linstorcontroller/linstorcontroller_controller.go b/pkg/controller/linstorcontroller/linstorcontroller_controller.go
index 52fd46f..373ade6 100644
--- a/pkg/controller/linstorcontroller/linstorcontroller_controller.go
+++ b/pkg/controller/linstorcontroller/linstorcontroller_controller.go
@@ -873,11 +873,11 @@ func newDeploymentForResource(controllerResource *piraeusv1.LinstorController) *
 	// as well as images without leader election helper
 	livenessProbe := corev1.Probe{
 		ProbeHandler: corev1.ProbeHandler{
-			HTTPGet: &corev1.HTTPGetAction{
-				Path: "/",
-				Port: intstr.FromInt(healthzPort),
+			Exec: &corev1.ExecAction{
+				Command: []string{"./liveness.sh"},
 			},
 		},
+		TimeoutSeconds: 15,
 	}
 
 	meta := getObjectMeta(controllerResource, "%s-controller")
