diff --git a/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go b/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go
index 2c93547..23c2b21 100644
--- a/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go
+++ b/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go
@@ -960,7 +960,8 @@ func newSatelliteDaemonSet(satelliteSet *piraeusv1.LinstorSatelliteSet, satellit
 							}, // Run linstor-satellite.
 							Env:             satelliteSet.Spec.AdditionalEnv,
 							ImagePullPolicy: satelliteSet.Spec.ImagePullPolicy,
-							SecurityContext: &corev1.SecurityContext{Privileged: &kubeSpec.Privileged},
+							SecurityContext: &corev1.SecurityContext{Privileged: &kubeSpec.Privileged,
+								SELinuxOptions: &corev1.SELinuxOptions{Level: kubeSpec.Level, Type: kubeSpec.Type}},
 							Ports: []corev1.ContainerPort{
 								{
 									HostPort:      satelliteSet.Spec.SslConfig.Port(),
@@ -1266,10 +1267,11 @@ func daemonSetWithDRBDKernelModuleInjection(ds *apps.DaemonSet, satelliteSet *pi
 			Name:            "kernel-module-injector",
 			Image:           satelliteSet.Spec.KernelModuleInjectionImage,
 			ImagePullPolicy: satelliteSet.Spec.ImagePullPolicy,
-			SecurityContext: &corev1.SecurityContext{Privileged: &kubeSpec.Privileged},
-			Env:             env,
-			VolumeMounts:    volumeMounts,
-			Resources:       satelliteSet.Spec.KernelModuleInjectionResources,
+			SecurityContext: &corev1.SecurityContext{Privileged: &kubeSpec.Privileged,
+				SELinuxOptions: &corev1.SELinuxOptions{Level: kubeSpec.Level, Type: kubeSpec.Type}},
+			Env:          env,
+			VolumeMounts: volumeMounts,
+			Resources:    satelliteSet.Spec.KernelModuleInjectionResources,
 		},
 	}

diff --git a/pkg/k8s/spec/const.go b/pkg/k8s/spec/const.go
index fe89064..1fb8051 100644
--- a/pkg/k8s/spec/const.go
+++ b/pkg/k8s/spec/const.go
@@ -100,11 +100,13 @@ var (
 	HostPathDirectoryType         = corev1.HostPathDirectory
 	HostPathDirectoryOrCreateType = corev1.HostPathDirectoryOrCreate
 	MountPropagationBidirectional = corev1.MountPropagationBidirectional
-	FileType = corev1.HostPathFile
+	FileType                      = corev1.HostPathFile
 )

 // Shared consts common to container security. These need to be vars, so they
 // are addressible.
 var (
 	Privileged = true
+	Level      = "s0"
+	Type       = "spc_t"
 )
