diff --git a/pkg/controller/linstorcontroller/linstorcontroller_controller.go b/pkg/controller/linstorcontroller/linstorcontroller_controller.go
index babd957..6c9d2ad 100644
--- a/pkg/controller/linstorcontroller/linstorcontroller_controller.go
+++ b/pkg/controller/linstorcontroller/linstorcontroller_controller.go
@@ -176,13 +176,6 @@ func (r *ReconcileLinstorController) reconcileSpec(ctx context.Context, controll
 		return r.finalizeControllerSet(ctx, controllerResource)
 	}
 
-	log.V(DEBUG).Info("reconcile finalizer")
-
-	err = r.addFinalizer(ctx, controllerResource)
-	if err != nil {
-		return fmt.Errorf("failed to add finalizer: %w", err)
-	}
-
 	log.V(DEBUG).Info("reconcile legacy config map name")
 
 	err = reconcileutil.DeleteIfOwned(ctx, r.client, &corev1.ConfigMap{ObjectMeta: getObjectMeta(controllerResource, "%s-config")}, controllerResource)
diff --git a/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go b/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go
index 2c93547..d244c99 100644
--- a/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go
+++ b/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go
@@ -191,12 +191,6 @@ func (r *ReconcileLinstorSatelliteSet) reconcileSpec(ctx context.Context, satell
 		return r.finalizeSatelliteSet(ctx, satelliteSet)
 	}
 
-	log.V(DEBUG).Info("add finalizer")
-
-	if err := r.addFinalizer(ctx, satelliteSet); err != nil {
-		return []error{fmt.Errorf("failed to add finalizer to resource: %w", err)}
-	}
-
 	log.V(DEBUG).Info("reconcile legacy config map name")
 
 	err = reconcileutil.DeleteIfOwned(ctx, r.client, &corev1.ConfigMap{ObjectMeta: getObjectMeta(satelliteSet, "%s-config")}, satelliteSet)
