From 47075f29003135f4cedeab1612e1b1d7ca8f28c8 Mon Sep 17 00:00:00 2001
From: Dmitry Stolyarov <dmitry.stolyarov@flant.com>
Date: Tue, 3 Apr 2018 16:19:35 +0300
Subject: [PATCH] prometheus: Make relabeling field SourceLabel optional

---
 pkg/prometheus/promcfg.go | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/pkg/prometheus/promcfg.go b/pkg/prometheus/promcfg.go
index adbe23c2..349bcad2 100644
--- a/pkg/prometheus/promcfg.go
+++ b/pkg/prometheus/promcfg.go
@@ -392,8 +392,10 @@ func generateServiceMonitorConfig(version semver.Version, m *v1.ServiceMonitor,
 	if ep.MetricRelabelConfigs != nil {
 		var metricRelabelings []yaml.MapSlice
 		for _, c := range ep.MetricRelabelConfigs {
-			relabeling := yaml.MapSlice{
-				{Key: "source_labels", Value: c.SourceLabels},
+			relabeling := yaml.MapSlice{}
+
+			if len(c.SourceLabels) > 0 {
+				relabeling = append(relabeling, yaml.MapItem{Key: "source_labels", Value: c.SourceLabels})
 			}
 
 			if c.Separator != "" {
