diff --git a/api/v1beta1/index/node.go b/api/v1beta1/index/node.go
index 64fa4af29..f8e6d2439 100644
--- a/api/v1beta1/index/node.go
+++ b/api/v1beta1/index/node.go
@@ -40,5 +40,13 @@ func NodeByProviderID(o client.Object) []string {
 		return nil
 	}
 
+	if node.Spec.ProviderID == "static://" {
+		providerID := node.Annotations["node.deckhouse.io/provider-id"]
+
+		if providerID != "" {
+			node.Spec.ProviderID = providerID
+		}
+	}
+
 	return []string{node.Spec.ProviderID}
 }
diff --git a/internal/controllers/machine/machine_controller_noderef.go b/internal/controllers/machine/machine_controller_noderef.go
index 5a74b717c..40fe10fb0 100644
--- a/internal/controllers/machine/machine_controller_noderef.go
+++ b/internal/controllers/machine/machine_controller_noderef.go
@@ -233,6 +233,14 @@ func (r *Reconciler) getNode(ctx context.Context, c client.Reader, providerID st
 		}
 
 		for _, node := range nl.Items {
+				if node.Spec.ProviderID == "static://" {
+					providerID := node.Annotations["node.deckhouse.io/provider-id"]
+
+					if providerID != "" {
+						node.Spec.ProviderID = providerID
+					}
+				}
+
 			if providerID == node.Spec.ProviderID {
 				return &node, nil
 			}
