Index: cmd/kubeadm/app/cmd/join.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/cmd/kubeadm/app/cmd/join.go b/cmd/kubeadm/app/cmd/join.go
--- a/cmd/kubeadm/app/cmd/join.go	(revision 8fa00736d77acc8727f0a501d88db6a01b5591fd)
+++ b/cmd/kubeadm/app/cmd/join.go	(date 1637835709896)
@@ -424,6 +424,11 @@
 		return nil, err
 	}

+	initCfg, err := configutil.LoadInitConfigurationFromFile(opt.cfgPath)
+	if err != nil {
+		return nil, err
+	}
+
 	ignorePreflightErrorsSet, err := validation.ValidateIgnorePreflightErrors(opt.ignorePreflightErrors, cfg.NodeRegistration.IgnorePreflightErrors)
 	if err != nil {
 		return nil, err
@@ -447,6 +452,7 @@

 	return &joinData{
 		cfg:                   cfg,
+		initCfg:               initCfg,
 		tlsBootstrapCfg:       tlsBootstrapCfg,
 		ignorePreflightErrors: ignorePreflightErrorsSet,
 		outputWriter:          out,
