diff --git a/internal/store/pod.go b/internal/store/pod.go
index 5ed054f9..cb641549 100644
--- a/internal/store/pod.go
+++ b/internal/store/pod.go
@@ -36,6 +36,7 @@ var (
 	descPodLabelsDefaultLabels = []string{"namespace", "pod"}
 	containerWaitingReasons    = []string{"ContainerCreating", "CrashLoopBackOff", "CreateContainerConfigError", "ErrImagePull", "ImagePullBackOff", "CreateContainerError", "InvalidImageName"}
 	containerTerminatedReasons = []string{"OOMKilled", "Completed", "Error", "ContainerCannotRun", "DeadlineExceeded"}
+	podStatusReasons           = []string{"NodeLost", "Evicted", "MatchNodeSelector"}
 
 	podMetricFamilies = []metric.FamilyGenerator{
 		{
@@ -353,6 +354,30 @@ var (
 				}
 			}),
 		},
+		{
+			Name: "kube_pod_status_reason",
+			Type: metric.Gauge,
+			Help: "The pod status reasons",
+			GenerateFunc: wrapPodFunc(func(p *v1.Pod) *metric.Family {
+				ms := []*metric.Metric{}
+
+				for _, reason := range podStatusReasons {
+					metric := &metric.Metric{}
+					metric.LabelKeys = []string{"reason"}
+					metric.LabelValues = []string{reason}
+					if p.Status.Reason == reason {
+						metric.Value = boolFloat64(true)
+					} else {
+						metric.Value = boolFloat64(false)
+					}
+					ms = append(ms, metric)
+				}
+
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
 		{
 			Name: "kube_pod_container_info",
 			Type: metric.Gauge,
