diff --git a/grpcutil/health_check.go b/grpcutil/health_check.go
--- a/grpcutil/health_check.go
+++ b/grpcutil/health_check.go
@@ -71,3 +71,17 @@
 func (h *HealthCheck) Watch(_ *grpc_health_v1.HealthCheckRequest, _ grpc_health_v1.Health_WatchServer) error {
 	return status.Error(codes.Unimplemented, "Watching is not supported")
 }
+
+// List implements the grpc healthcheck.
+func (h *HealthCheck) List(ctx context.Context, _ *grpc_health_v1.HealthListRequest) (*grpc_health_v1.HealthListResponse, error) {
+	checkResp, err := h.Check(ctx, nil)
+	if err != nil {
+		return &grpc_health_v1.HealthListResponse{}, err
+	}
+
+	return &grpc_health_v1.HealthListResponse{
+		Statuses: map[string]*grpc_health_v1.HealthCheckResponse{
+			"server": checkResp,
+		},
+	}, nil
+}
