diff --git a/internal/ingress/controller/checker.go b/internal/ingress/controller/checker.go
index d1bf19ddf..7e8ee68e1 100644
--- a/internal/ingress/controller/checker.go
+++ b/internal/ingress/controller/checker.go
@@ -34,8 +34,8 @@ func (n *NGINXController) Name() string {
 }
 
 // Check returns if the nginx healthz endpoint is returning ok (status code 200)
-func (n *NGINXController) Check(_ *http.Request) error {
-	if n.isShuttingDown {
+func (n *NGINXController) Check(r *http.Request) error {
+	if r.Header.Get("D8s-External-Check") == "True" && n.isShuttingDown {
 		return fmt.Errorf("the ingress controller is shutting down")
 	}
 
