diff --git a/cmd/nginx/main.go b/cmd/nginx/main.go
--- a/cmd/nginx/main.go
+++ b/cmd/nginx/main.go
@@ -121,12 +121,13 @@ func main() {
 	reg := prometheus.NewRegistry()

 	reg.MustRegister(collectors.NewGoCollector())
 	reg.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{
 		PidFn:        func() (int, error) { return os.Getpid(), nil },
 		ReportErrors: true,
 	}))
+	reg.MustRegister(nginx.GeoIPErrors)

 	mc := metric.NewDummyCollector()
 	if conf.EnableMetrics {
-		mc, err = metric.NewCollector(conf.MetricsPerHost, conf.ReportStatusClasses, reg, conf.IngressClassConfiguration.Controller, *conf.MetricsBuckets, conf.ExcludeSocketMetrics)
+		mc, err = metric.NewCollector(conf.MetricsPerHost, conf.MetricsPerUndefinedHost, conf.ReportStatusClasses, reg, conf.IngressClassConfiguration.Controller, *conf.MetricsBuckets, conf.MetricsBucketFactor, conf.MetricsMaxBuckets, conf.ExcludeSocketMetrics)
 		if err != nil {
 			klog.Fatalf("Error creating prometheus collector:  %v", err)
 		}
 	}
diff --git a/internal/nginx/maxmind.go b/internal/nginx/maxmind.go
--- a/internal/nginx/maxmind.go
+++ b/internal/nginx/maxmind.go
@@ -27,10 +27,12 @@ import (
 	"os"
 	"os/exec"
 	"syscall"
 	"time"

+	"github.com/prometheus/client_golang/prometheus"
 	"k8s.io/apimachinery/pkg/util/wait"
 	klog "k8s.io/klog/v2"
 )
@@ -49,6 +51,15 @@ var MaxmindRetriesCount = 1
 // MaxmindRetriesTimeout maxmind download retries timeout in seconds, 0 - do not retry to download if something went wrong
 var MaxmindRetriesTimeout = time.Second * 0

+// GeoIPErrors number of errors when loading Geo2 DB
+var GeoIPErrors = prometheus.NewCounterVec(
+	prometheus.CounterOpts{
+		Name: "geoip_errors_total",
+		Help: "Number of errors related to GeoIP",
+	},
+	[]string{"reason", "type"},
+)
+
 // minimumRetriesCount minimum value of the MaxmindRetriesCount parameter. If MaxmindRetriesCount less than minimumRetriesCount, it will be set to minimumRetriesCount
 const minimumRetriesCount = 1

@@ -113,6 +124,13 @@ func DownloadGeoLite2DB(attempts int, period time.Duration) error {
 			klog.V(2).Infof("Attempting to download GeoIP DB: %s", dbName)
 			dlError = downloadDatabase(dbName)
 			if dlError != nil {
+				trimmedError := []rune(dlError.Error())
+				if len(trimmedError) > 64 {
+					trimmedError = trimmedError[:64]
+				}
+				stringErr := string(trimmedError)
+				GeoIPErrors.WithLabelValues(stringErr, "download").Inc()
+
 				break
 			}
 		}
