diff --git a/web/federate.go b/web/federate.go
index b0f4c0610..dc38e845f 100644
--- a/web/federate.go
+++ b/web/federate.go
@@ -71,6 +71,8 @@ func (h *Handler) federation(w http.ResponseWriter, req *http.Request) {
 		matcherSets = append(matcherSets, matchers)
 	}
 
+	dropExternalLabels := (req.Form.Get("drop_external_labels") == "1")
+
 	var (
 		mint   = timestamp.FromTime(h.now().Time().Add(-h.lookbackDelta))
 		maxt   = timestamp.FromTime(h.now().Time())
@@ -256,13 +258,15 @@ Loop:
 			continue
 		}
 		// Attach global labels if they do not exist yet.
-		for _, ln := range externalLabelNames {
-			lv := externalLabels[ln]
-			if _, ok := globalUsed[ln]; !ok {
-				protMetric.Label = append(protMetric.Label, &dto.LabelPair{
-					Name:  proto.String(ln),
-					Value: proto.String(lv),
-				})
+		if !dropExternalLabels {
+			for _, ln := range externalLabelNames {
+				lv := externalLabels[ln]
+				if _, ok := globalUsed[ln]; !ok {
+					protMetric.Label = append(protMetric.Label, &dto.LabelPair{
+						Name:  proto.String(ln),
+						Value: proto.String(lv),
+					})
+				}
 			}
 		}
 
