diff --git a/storage/remote/queue_manager.go b/storage/remote/queue_manager.go
index 3edd31b91..3220c34ab 100644
--- a/storage/remote/queue_manager.go
+++ b/storage/remote/queue_manager.go
@@ -581,7 +581,7 @@ outer:
 			t.metrics.droppedSamplesTotal.Inc()
 			t.dataDropped.incr(1)
 			if _, ok := t.droppedSeries[s.Ref]; !ok {
-				level.Info(t.logger).Log("msg", "Dropped sample for series that was not explicitly dropped via relabelling", "ref", s.Ref)
+				level.Info(t.logger).Log("msg", "Dropped sample for series that was not explicitly dropped via relabelling", "ref", s.Ref, s.T, s.V)
 			}
 			t.seriesMtx.Unlock()
 			continue
