diff --git a/storage/remote/queue_manager.go b/storage/remote/queue_manager.go
index 3edd31b91..3220c34ab 100644
--- a/storage/remote/queue_manager.go
+++ b/storage/remote/queue_manager.go
@@ -581,7 +581,7 @@ outer:
 			t.metrics.droppedSamplesTotal.Inc()
 			t.dataDropped.incr(1)
 			if _, ok := t.droppedSeries[s.Ref]; !ok {
-				level.Info(t.logger).Log("msg", "Dropped sample for series that was not explicitly dropped via relabelling", "ref", s.Ref)
+				level.Info(t.logger).Log("msg", "Dropped sample for series that was not explicitly dropped via relabelling", "ref", s.Ref, s.T, s.V)
 			}
 			t.seriesMtx.Unlock()
 			continue
diff --git a/tsdb/wlog/checkpoint.go b/tsdb/wlog/checkpoint.go
index 42b03f48f..28a18c041 100644
--- a/tsdb/wlog/checkpoint.go
+++ b/tsdb/wlog/checkpoint.go
@@ -271,6 +271,9 @@ func Checkpoint(logger log.Logger, w *WL, from, to int, keep func(id chunks.Head
 			// Only keep reference to the latest found metadata for each refID.
 			repl := 0
 			for _, m := range metadata {
+				if m.Ref == 12756107 || m.Ref == 12756108 || m.Ref == 12756109 {
+					 level.Info(logger).Log("msg", "debug", "ref", m.Ref, "help", m.Help)
+				}
 				if keep(m.Ref) {
 					if _, ok := latestMetadataMap[m.Ref]; !ok {
 						repl++
diff --git a/tsdb/wlog/watcher.go b/tsdb/wlog/watcher.go
index b0c17dcba..82c9c2c6e 100644
--- a/tsdb/wlog/watcher.go
+++ b/tsdb/wlog/watcher.go
@@ -500,6 +500,11 @@ func (w *Watcher) readSegment(r *LiveReader, segmentNum int, tail bool) error {
 				w.recordDecodeFailsMetric.Inc()
 				return err
 			}
+			for _, v := range series {
+				if v.Ref == 12756108 || v.Ref == 12756109 || v.Ref ==  12756107 {
+					level.Info(w.logger).Log("msg", "debug", "ref", v.Ref, "lbls", v.Labels)
+				}
+			}
 			w.writer.StoreSeries(series, segmentNum)
 
 		case record.Samples:
