diff --git a/satellite/src/main/java/com/linbit/linstor/core/Satellite.java b/satellite/src/main/java/com/linbit/linstor/core/Satellite.java
index 00b7b1fce..3c026bd8a 100644
--- a/satellite/src/main/java/com/linbit/linstor/core/Satellite.java
+++ b/satellite/src/main/java/com/linbit/linstor/core/Satellite.java
@@ -377,6 +377,7 @@ public final class Satellite
     {
         System.out.printf("%s, Module %s\n", LinStor.PROGRAM, LinStor.SATELLITE_MODULE);
         LinStor.printStartupInfo();
+        System.out.println("TEST FIX-LOGS");
 
         StltConfig cfg = new StltConfig(args);
 
diff --git a/satellite/src/main/java/com/linbit/linstor/layer/storage/utils/Commands.java b/satellite/src/main/java/com/linbit/linstor/layer/storage/utils/Commands.java
index a63c78cba..2c88ea9de 100644
--- a/satellite/src/main/java/com/linbit/linstor/layer/storage/utils/Commands.java
+++ b/satellite/src/main/java/com/linbit/linstor/layer/storage/utils/Commands.java
@@ -12,10 +12,7 @@ import com.linbit.linstor.storage.StorageException;
 import com.linbit.utils.StringUtils;
 
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.List;
+import java.util.*;
 
 public class Commands
 {
@@ -77,6 +74,16 @@ public class Commands
         try
         {
             outData = extCmd.exec(command);
+            System.out.println("------- command------------------------");
+            System.out.println(Arrays.toString(command));
+            System.out.println("---------------------------------------");
+            System.out.println("------- extCmd.exec outData-------------------------");
+            System.out.println("outData.exitCode: " + outData.exitCode);
+            System.out.println("outData.stderrData: " + Arrays.toString(outData.stderrData));
+            System.out.println("outData.stdoutData: " + Arrays.toString(outData.stdoutData));
+            System.out.println("outData.executedCommand: " + Arrays.toString(outData.executedCommand));
+            System.out.println("----------------------------------------------------");
+
 
             boolean skipExitCodeCheck = false;
             while (!(outData.exitCode == ExtCmdUtils.DEFAULT_RET_CODE_OK || allowExitCodes.contains(outData.exitCode)))
@@ -121,7 +128,6 @@ public class Commands
                 exc
             );
         }
-
         return outData;
     }

