diff --git a/satellite/src/main/java/com/linbit/linstor/core/Satellite.java b/satellite/src/main/java/com/linbit/linstor/core/Satellite.java
index 00b7b1fce..3c026bd8a 100644
--- a/satellite/src/main/java/com/linbit/linstor/core/Satellite.java
+++ b/satellite/src/main/java/com/linbit/linstor/core/Satellite.java
@@ -377,6 +377,7 @@ public final class Satellite
     {
         System.out.printf("%s, Module %s\n", LinStor.PROGRAM, LinStor.SATELLITE_MODULE);
         LinStor.printStartupInfo();
+        System.out.println("TEST FIX-LOGS");
 
         StltConfig cfg = new StltConfig(args);
 
diff --git a/satellite/src/main/java/com/linbit/linstor/layer/drbd/utils/DrbdAdm.java b/satellite/src/main/java/com/linbit/linstor/layer/drbd/utils/DrbdAdm.java
index efa919929..241a4dad8 100644
--- a/satellite/src/main/java/com/linbit/linstor/layer/drbd/utils/DrbdAdm.java
+++ b/satellite/src/main/java/com/linbit/linstor/layer/drbd/utils/DrbdAdm.java
@@ -23,6 +23,8 @@ import javax.inject.Singleton;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Path;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
@@ -89,6 +91,7 @@ public class DrbdAdm
         // command.add(resName);
         command.add(drbdRscData.getSuffixedResourceName());
         // execute(Arrays.asList("drbdsetup", "show", drbdRscData.getSuffixedResourceName()));
+        System.out.println(" >>> DRBD adjust <<<");
         execute(command);
         // execute(Arrays.asList("drbdsetup", "show", drbdRscData.getSuffixedResourceName()));
     }
@@ -121,6 +124,7 @@ public class DrbdAdm
         // command.addAll(asConfigParameter(resName)); // basically just adds the -c <rscName.conf> as parameter
         command.add("resize");
         command.add(resName + "/" + drbdVlmData.getVlmNr().value);
+        System.out.println(" >>> DRBD resize <<<");
         execute(command);
     }
 
@@ -259,6 +263,7 @@ public class DrbdAdm
         {
             commands.add("--diskless");
         }
+        System.out.println(" >>> DRBD detach <<<");
         execute(commands);
     }
 
@@ -289,10 +294,24 @@ public class DrbdAdm
 
         String[] params = getGetGiCommand(minorNr, blockDevPath, mdTypeParam);
         ExtCmd utilsCmd = extCmdFactory.create();
+        System.out.println(" >>> DRBD hasMetaData <<<");
         File nullDevice = new File(Platform.nullDevice());
         try
         {
             OutputData outputData = utilsCmd.pipeExec(ProcessBuilder.Redirect.from(nullDevice), params);
+
+            String cmd = String.join(" ", outputData.executedCommand);
+            System.out.println("executedCommand: " + cmd);
+
+            System.out.println("exitCode: " + outputData.exitCode);
+            String stdoutStr = new String(outputData.stdoutData);
+
+            System.out.println("stdoutData: " + stdoutStr);
+            String stderrStr = new String(outputData.stderrData);
+
+            System.out.println("stderrData: " + stderrStr);
+            System.out.println(" >>> <<<");
+
             if (outputData.exitCode == DRBDMETA_NO_VALID_MD_RC || outputData.exitCode == DRBDMETA_STRANGE_BM_OFFSET)
             {
                 mdFlag = false;
@@ -314,10 +333,25 @@ public class DrbdAdm
         String ret;
 
         String[] params = getGetGiCommand(minorNr, blockDevPath, mdTypeParam);
+        System.out.println(" >>> DRBD getCurrentGID <<<");
         ExtCmd utilsCmd = extCmdFactory.create();
         try
         {
             OutputData outputData = utilsCmd.exec(params);
+
+            String cmd = String.join(" ", outputData.executedCommand);
+            System.out.println("executedCommand: " + cmd);
+
+            System.out.println("exitCode: " + outputData.exitCode);
+            String stdoutStr = new String(outputData.stdoutData);
+
+            System.out.println("stdoutData: " + stdoutStr);
+            String stderrStr = new String(outputData.stderrData);
+
+            System.out.println("stderrData: " + stderrStr);
+            System.out.println(" >>> <<<");
+
+
             ret = new String(outputData.stdoutData);
         }
         catch (ChildProcessTimeoutException timeoutExc)
@@ -411,6 +445,7 @@ public class DrbdAdm
     public void resumeIo(DrbdRscData<Resource> drbdRscData)
         throws ExtCmdFailedException
     {
+        System.out.println(" >>> DRBD resumeIo <<<");
         if (drbdRscData.resFileExists())
         {
             execute(DRBDADM_UTIL, "resume-io", drbdRscData.getSuffixedResourceName());
@@ -458,6 +493,9 @@ public class DrbdAdm
             "sh-nop"
         );
 
+        System.out.println(DRBDADM_UTIL + " --config-to-test " + tmpResPathStr + " --config-to-exclude " + resPathStr + " sh-nop ");
+        System.out.println(" >>> DRBD checkResFile <<<");
+
         // Using -c disables /etc/drbd.d/global_common.conf
         // execute(DRBDADM_UTIL, "-c", tmpResPathStr, "-d", "up", resourceName.value);
         // execute(DRBDADM_UTIL, "-d", "up", resourceName.value);
@@ -477,6 +515,8 @@ public class DrbdAdm
             suffixedRscName,
             Integer.toString(nodeId)
         );
+
+        System.out.println(" >>> DRBD deletePeer <<<");
     }
 
     public void forgetPeer(DrbdRscData<Resource> rscData) throws ExtCmdFailedException
@@ -492,6 +532,8 @@ public class DrbdAdm
             suffixedRscName,
             Integer.toString(nodeId)
         );
+
+        System.out.println(" >>> DRBD forgetPeer <<<");
     }
 
     private void simpleSetupCommand(
@@ -513,6 +555,8 @@ public class DrbdAdm
         command.add(drbdObj);
 
         execute(command);
+
+        System.out.println(" >>> DRBD simpleSetupCommand <<<");
     }
 
     private void simpleAdmCommand(DrbdRscData<Resource> drbdRscData, String subcommand) throws ExtCmdFailedException
@@ -544,6 +588,7 @@ public class DrbdAdm
         }
         command.add(resName);
 
+        System.out.println(" >>> DRBD simpleAdmCommand <<<");
         execute(command);
     }
 
@@ -560,7 +605,9 @@ public class DrbdAdm
             "--wait-after-sb=yes",
             "--wfc-timeout=" + timeout,
             drbdRscData.getSuffixedResourceName()
+
         );
+        System.out.println(" >>> DRBD waitForFamily <<<");
     }
 
     // Using -c disables /etc/drbd.d/global_common.conf
@@ -600,6 +647,25 @@ public class DrbdAdm
                 extCmd.setTimeout(TimeoutType.WAIT, 5*60*1000);
             }
             OutputData outputData = extCmd.pipeExec(ProcessBuilder.Redirect.from(nullDevice), command);
+
+            DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy/MM/dd/HH:mm:ss:SSS");
+            LocalDateTime now = LocalDateTime.now();
+            System.out.println(dtf.format(now) + "  <<<<< DRBD >>>>>");
+
+            String cmd = String.join(" ", outputData.executedCommand);
+            System.out.println("executedCommand: " + cmd);
+
+
+            System.out.println("outputData.exitCode: " +  outputData.exitCode);
+
+            String stdoutStr = new String(outputData.stdoutData);
+            System.out.println("outputData.stdoutData: " +  stdoutStr);
+
+            String stdouterrStr = new String(outputData.stderrData);
+            System.out.println("outputData.stderrData: " +  stdouterrStr);
+
+            System.out.println("<<<<<  >>>>>");
+
             if (outputData.exitCode != 0)
             {
                 throw new ExtCmdFailedException(command, outputData);
diff --git a/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/LvmProvider.java b/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/LvmProvider.java
index e06e93901..580aebb28 100644
--- a/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/LvmProvider.java
+++ b/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/LvmProvider.java
@@ -452,6 +452,8 @@ public class LvmProvider extends AbsStorageProvider<LvsInfo, LvmData<Resource>,
                 config
             )
         );
+        LvmCommands.vgscanmknodes(extCmdFactory.create());
+        System.out.println("point vgscan --mknodes after resizeLvImpl");
     }
 
     @Override
@@ -567,6 +569,7 @@ public class LvmProvider extends AbsStorageProvider<LvsInfo, LvmData<Resource>,
         if (hasSharedVolumeGroups(vlmDataList, snapVlms))
         {
             LvmCommands.vgscan(extCmdFactory.create(), true);
+            LvmCommands.vgscanmknodes(extCmdFactory.create());
         }
 
         return LvmUtils.getLvsInfo(
diff --git a/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/utils/LvmCommands.java b/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/utils/LvmCommands.java
index 95ad756ff..88e1992e0 100644
--- a/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/utils/LvmCommands.java
+++ b/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/utils/LvmCommands.java
@@ -694,6 +694,23 @@ public class LvmCommands
         );
     }
 
+    public static OutputData vgscanmknodes(ExtCmd extCmd) throws StorageException
+    {
+        String[] cmd;
+            cmd = new String[]
+                    {
+                            "vgscan", "--mknodes"
+                    };
+
+
+        return genericExecutor(
+                extCmd.setSaveWithoutSharedLocks(true),
+                cmd,
+                "'vgscanmknodes' failed",
+                "'vgscanmknodes' failed"
+        );
+    }
+
     public enum LvmVolumeType
     {
         VOLUME("volume"), SNAPSHOT("snapshot"), THIN_POOL("thin pool");
diff --git a/satellite/src/main/java/com/linbit/linstor/layer/storage/utils/Commands.java b/satellite/src/main/java/com/linbit/linstor/layer/storage/utils/Commands.java
index a63c78cba..ca5df3167 100644
--- a/satellite/src/main/java/com/linbit/linstor/layer/storage/utils/Commands.java
+++ b/satellite/src/main/java/com/linbit/linstor/layer/storage/utils/Commands.java
@@ -11,11 +11,13 @@ import com.linbit.extproc.ExtCmdUtils;
 import com.linbit.linstor.storage.StorageException;
 import com.linbit.utils.StringUtils;
 
+import java.io.BufferedReader;
+import java.io.File;
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.List;
+import java.io.InputStreamReader;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.*;
 
 public class Commands
 {
@@ -77,6 +79,23 @@ public class Commands
         try
         {
             outData = extCmd.exec(command);
+            DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy/MM/dd/HH:mm:ss:SSS");
+            LocalDateTime now = LocalDateTime.now();
+            System.out.println( dtf.format(now) +  "  ------- LVM ------------------------------------");
+
+            String cmd = String.join(" ", outData.executedCommand);
+            System.out.println("executedCommand: " + cmd);
+
+            System.out.println("exitCode: " + outData.exitCode);
+
+            String stdoutStr = new String(outData.stdoutData);
+            System.out.println("stdoutData: " + stdoutStr);
+
+            String stderrStr = new String(outData.stderrData);
+            System.out.println("stderrData: " + stderrStr);
+
+            System.out.println("-------------------------------------------------------------------------");
+
 
             boolean skipExitCodeCheck = false;
             while (!(outData.exitCode == ExtCmdUtils.DEFAULT_RET_CODE_OK || allowExitCodes.contains(outData.exitCode)))
@@ -121,7 +140,6 @@ public class Commands
                 exc
             );
         }
-
         return outData;
     }
 
@@ -191,12 +209,24 @@ public class Commands
     )
         throws StorageException
     {
+
+        File f = new File(devicePath);
+        if (!f.exists()) {
+            System.out.println("[error symlink] " + devicePath);
+        }
+
         OutputData output = genericExecutor(
             extCmd,
             new String[] {"blockdev", "--getsize64", devicePath},
             "Failed to get block size of " + devicePath,
             "Failed to get block size of " + devicePath
         );
+        if (output.exitCode != 0)
+        {
+            DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy/MM/dd/HH:mm:ss:SSS");
+            LocalDateTime now = LocalDateTime.now();
+            System.out.println(dtf.format(now) + " [lvresize-faceslapper activated]");
+        }
         String outRaw = new String(output.stdoutData);
         return SizeConv.convert(
             Long.parseLong(outRaw.trim()),
