diff --git a/satellite/src/main/java/com/linbit/linstor/core/Satellite.java b/satellite/src/main/java/com/linbit/linstor/core/Satellite.java
index 00b7b1fce..3c026bd8a 100644
--- a/satellite/src/main/java/com/linbit/linstor/core/Satellite.java
+++ b/satellite/src/main/java/com/linbit/linstor/core/Satellite.java
@@ -377,6 +377,7 @@ public final class Satellite
     {
         System.out.printf("%s, Module %s\n", LinStor.PROGRAM, LinStor.SATELLITE_MODULE);
         LinStor.printStartupInfo();
+        System.out.println("TEST FIX-LOGS");
 
         StltConfig cfg = new StltConfig(args);
 
diff --git a/satellite/src/main/java/com/linbit/linstor/layer/drbd/utils/DrbdAdm.java b/satellite/src/main/java/com/linbit/linstor/layer/drbd/utils/DrbdAdm.java
index efa919929..fdbb15a8b 100644
--- a/satellite/src/main/java/com/linbit/linstor/layer/drbd/utils/DrbdAdm.java
+++ b/satellite/src/main/java/com/linbit/linstor/layer/drbd/utils/DrbdAdm.java
@@ -23,6 +23,8 @@ import javax.inject.Singleton;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Path;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
@@ -89,6 +91,7 @@ public class DrbdAdm
         // command.add(resName);
         command.add(drbdRscData.getSuffixedResourceName());
         // execute(Arrays.asList("drbdsetup", "show", drbdRscData.getSuffixedResourceName()));
+        System.out.println(" >>> DRBD adjust <<<");
         execute(command);
         // execute(Arrays.asList("drbdsetup", "show", drbdRscData.getSuffixedResourceName()));
     }
@@ -121,6 +124,7 @@ public class DrbdAdm
         // command.addAll(asConfigParameter(resName)); // basically just adds the -c <rscName.conf> as parameter
         command.add("resize");
         command.add(resName + "/" + drbdVlmData.getVlmNr().value);
+        System.out.println(" >>> DRBD resize <<<");
         execute(command);
     }
 
@@ -259,6 +263,7 @@ public class DrbdAdm
         {
             commands.add("--diskless");
         }
+        System.out.println(" >>> DRBD detach <<<");
         execute(commands);
     }
 
@@ -289,6 +294,7 @@ public class DrbdAdm
 
         String[] params = getGetGiCommand(minorNr, blockDevPath, mdTypeParam);
         ExtCmd utilsCmd = extCmdFactory.create();
+        System.out.println(" >>> DRBD hasMetaData <<<");
         File nullDevice = new File(Platform.nullDevice());
         try
         {
@@ -314,6 +320,7 @@ public class DrbdAdm
         String ret;
 
         String[] params = getGetGiCommand(minorNr, blockDevPath, mdTypeParam);
+        System.out.println(" >>> DRBD getCurrentGID <<<");
         ExtCmd utilsCmd = extCmdFactory.create();
         try
         {
@@ -411,6 +418,7 @@ public class DrbdAdm
     public void resumeIo(DrbdRscData<Resource> drbdRscData)
         throws ExtCmdFailedException
     {
+        System.out.println(" >>> DRBD resumeIo <<<");
         if (drbdRscData.resFileExists())
         {
             execute(DRBDADM_UTIL, "resume-io", drbdRscData.getSuffixedResourceName());
@@ -458,6 +466,8 @@ public class DrbdAdm
             "sh-nop"
         );
 
+        System.out.println(" >>> DRBD checkResFile <<<");
+
         // Using -c disables /etc/drbd.d/global_common.conf
         // execute(DRBDADM_UTIL, "-c", tmpResPathStr, "-d", "up", resourceName.value);
         // execute(DRBDADM_UTIL, "-d", "up", resourceName.value);
@@ -477,6 +487,8 @@ public class DrbdAdm
             suffixedRscName,
             Integer.toString(nodeId)
         );
+
+        System.out.println(" >>> DRBD deletePeer <<<");
     }
 
     public void forgetPeer(DrbdRscData<Resource> rscData) throws ExtCmdFailedException
@@ -492,6 +504,8 @@ public class DrbdAdm
             suffixedRscName,
             Integer.toString(nodeId)
         );
+
+        System.out.println(" >>> DRBD forgetPeer <<<");
     }
 
     private void simpleSetupCommand(
@@ -513,6 +527,8 @@ public class DrbdAdm
         command.add(drbdObj);
 
         execute(command);
+
+        System.out.println(" >>> DRBD simpleSetupCommand <<<");
     }
 
     private void simpleAdmCommand(DrbdRscData<Resource> drbdRscData, String subcommand) throws ExtCmdFailedException
@@ -544,6 +560,7 @@ public class DrbdAdm
         }
         command.add(resName);
 
+        System.out.println(" >>> DRBD simpleAdmCommand <<<");
         execute(command);
     }
 
@@ -560,7 +577,9 @@ public class DrbdAdm
             "--wait-after-sb=yes",
             "--wfc-timeout=" + timeout,
             drbdRscData.getSuffixedResourceName()
+
         );
+        System.out.println(" >>> DRBD waitForFamily <<<");
     }
 
     // Using -c disables /etc/drbd.d/global_common.conf
@@ -600,6 +619,25 @@ public class DrbdAdm
                 extCmd.setTimeout(TimeoutType.WAIT, 5*60*1000);
             }
             OutputData outputData = extCmd.pipeExec(ProcessBuilder.Redirect.from(nullDevice), command);
+
+            DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy/MM/dd/HH:mm:ss:SSS");
+            LocalDateTime now = LocalDateTime.now();
+            System.out.println(dtf.format(now) + "  <<<<< DRBD >>>>>");
+
+            String cmd = String.join(" ", outputData.executedCommand);
+            System.out.println("executedCommand: " + cmd);
+
+
+            System.out.println("outputData.exitCode: " +  outputData.exitCode);
+
+            String stdoutStr = new String(outputData.stdoutData);
+            System.out.println("outputData.stdoutData: " +  stdoutStr);
+
+            String stdouterrStr = new String(outputData.stderrData);
+            System.out.println("outputData.stderrData: " +  stdouterrStr);
+
+            System.out.println("<<<<<  >>>>>");
+
             if (outputData.exitCode != 0)
             {
                 throw new ExtCmdFailedException(command, outputData);
diff --git a/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/LvmProvider.java b/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/LvmProvider.java
index e06e93901..244262367 100644
--- a/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/LvmProvider.java
+++ b/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/LvmProvider.java
@@ -452,6 +452,7 @@ public class LvmProvider extends AbsStorageProvider<LvsInfo, LvmData<Resource>,
                 config
             )
         );
+        System.out.println("LvmCommands.vgscan(extCmdFactory.create(), true);");
     }
 
     @Override
diff --git a/satellite/src/main/java/com/linbit/linstor/layer/storage/utils/Commands.java b/satellite/src/main/java/com/linbit/linstor/layer/storage/utils/Commands.java
index a63c78cba..7549f6f28 100644
--- a/satellite/src/main/java/com/linbit/linstor/layer/storage/utils/Commands.java
+++ b/satellite/src/main/java/com/linbit/linstor/layer/storage/utils/Commands.java
@@ -12,10 +12,10 @@ import com.linbit.linstor.storage.StorageException;
 import com.linbit.utils.StringUtils;
 
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.List;
+import java.nio.charset.StandardCharsets;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.*;
 
 public class Commands
 {
@@ -77,6 +77,23 @@ public class Commands
         try
         {
             outData = extCmd.exec(command);
+            DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy/MM/dd/HH:mm:ss:SSS");
+            LocalDateTime now = LocalDateTime.now();
+            System.out.println( dtf.format(now) +  "  ------- LVM ------------------------------------");
+
+            String cmd = String.join(" ", outData.executedCommand);
+            System.out.println("executedCommand: " + cmd);
+
+            System.out.println("exitCode: " + outData.exitCode);
+
+            String stdoutStr = new String(outData.stdoutData);
+            System.out.println("stdoutData: " + stdoutStr);
+
+            String stderrStr = new String(outData.stderrData);
+            System.out.println("stderrData: " + stderrStr);
+
+            System.out.println("-------------------------------------------------------------------------");
+
 
             boolean skipExitCodeCheck = false;
             while (!(outData.exitCode == ExtCmdUtils.DEFAULT_RET_CODE_OK || allowExitCodes.contains(outData.exitCode)))
@@ -121,7 +138,6 @@ public class Commands
                 exc
             );
         }
-
         return outData;
     }
 
