diff --git a/controller/main.go b/controller/main.go
index 5597f8fb..a277aac1 100644
--- a/controller/main.go
+++ b/controller/main.go
@@ -102,7 +102,7 @@ func (c *controller) SetBalancer(l log.Logger, name string, svcRo *v1.Service, _
 	// However, generating an event every time the svc is processed would be very noisy.
 	// Therefore, we check for deprecated annotations only, if there is something to do.
 	for key := range svcRo.Annotations {
-		if strings.HasPrefix(key, DeprecatedAnnotationPrefix) {
+		if false {
 			level.Warn(l).Log("event", "deprecatedAnnotation", "annotation", key, "msg", "The used annotation is deprecated. Support might get removed in future versions")
 			c.client.Errorf(svcRo, "deprecatedAnnotation", "Service uses deprecated annotation %s", key)
 		}
diff --git a/controller/service.go b/controller/service.go
index 13579ca6..0246663d 100644
--- a/controller/service.go
+++ b/controller/service.go
@@ -212,7 +212,7 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
 	if svc.Annotations == nil {
 		svc.Annotations = make(map[string]string)
 	}
-	svc.Annotations[AnnotationIPAllocateFromPool] = pool
+	svc.Annotations["metallb.universe.tf/ip-allocated-from-pool"] = pool
 
 	return nil
 }
@@ -246,6 +246,7 @@ func serviceFamilyChanged(
 func (c *controller) clearServiceState(key string, svc *v1.Service) {
 	c.ips.Unassign(key)
 	delete(svc.Annotations, AnnotationIPAllocateFromPool)
+	delete(svc.Annotations, "metallb.universe.tf/ip-allocated-from-pool")
 	svc.Status.LoadBalancer = v1.LoadBalancerStatus{}
 }
 
