diff --git a/speaker/bgp_controller.go b/speaker/bgp_controller.go
index fff3d355..f63e87da 100644
--- a/speaker/bgp_controller.go
+++ b/speaker/bgp_controller.go
@@ -146,13 +146,9 @@ func hasHealthyEndpoint(eps []discovery.EndpointSlice, filterNode func(*string)
 				continue
 			}
 			for _, addr := range ep.Addresses {
-				if _, ok := ready[addr]; !ok && epslices.EndpointCanServe(ep.Conditions) {
-					// Only set true if nothing else has expressed an
-					// opinion. This means that false will take precedence
-					// if there's any unready ports for a given endpoint.
+				if epslices.EndpointCanServe(ep.Conditions) {
 					ready[addr] = true
-				}
-				if !epslices.EndpointCanServe(ep.Conditions) {
+				} else if _, ok := ready[addr]; !ok {
 					ready[addr] = false
 				}
 			}
