Index: pkg/cloudprovider/vsphere/instances.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/cloudprovider/vsphere/instances.go b/pkg/cloudprovider/vsphere/instances.go
--- a/pkg/cloudprovider/vsphere/instances.go	(revision 2461f899209081458fe3a0a7a109dad80f6ad02d)
+++ b/pkg/cloudprovider/vsphere/instances.go	(revision 6b6f23f21569d43b84721011068fa2539b9ab7f8)
@@ -164,6 +164,10 @@
 func (i *instances) InstanceExistsByProviderID(ctx context.Context, providerID string) (bool, error) {
 	klog.V(4).Info("instances.InstanceExistsByProviderID() called with ", providerID)

+	if providerID == "static://" {
+		return true, nil
+	}
+
 	// Check if node has been discovered already
 	uid := GetUUIDFromProviderID(providerID)
 	err := i.nodeManager.DiscoverNode(uid, cm.FindVMByUUID)
