diff --git a/pkg/validation/sessions.go b/pkg/validation/sessions.go
index 96ea6d4f..8614a60c 100644
--- a/pkg/validation/sessions.go
+++ b/pkg/validation/sessions.go
@@ -64,7 +64,13 @@ func validateRedisSessionStore(o *options.Options) []string {
 func sendRedisConnectionTest(client redis.Client, key string, val string) []string {
        msgs := []string{}
        ctx := context.Background()
-
+       for i := 0; i < 5; i++ {
+               err := client.Set(ctx, key, []byte(val), time.Duration(60)*time.Second)
+               if err == nil {
+                       break
+               }
+               time.Sleep(2 * time.Second)
+       }
        err := client.Set(ctx, key, []byte(val), time.Duration(60)*time.Second)
        if err != nil {
                msgs = append(msgs, fmt.Sprintf("unable to set a redis initialization key: %v", err))
