diff --git a/pkg/validation/sessions.go b/pkg/validation/sessions.go
index 96ea6d4f..294c1a00 100644
--- a/pkg/validation/sessions.go
+++ b/pkg/validation/sessions.go
@@ -8,6 +8,7 @@ import (

        "github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
        "github.com/oauth2-proxy/oauth2-proxy/v7/pkg/encryption"
+       "github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
        "github.com/oauth2-proxy/oauth2-proxy/v7/pkg/sessions/redis"
 )

@@ -64,7 +65,14 @@ func validateRedisSessionStore(o *options.Options) []string {
 func sendRedisConnectionTest(client redis.Client, key string, val string) []string {
        msgs := []string{}
        ctx := context.Background()
-
+       for i := 0; i < 5; i++ {
+               logger.Printf("Testing connection to redis...")
+               err := client.Set(ctx, key, []byte(val), time.Duration(60)*time.Second)
+               if err == nil {
+                       break
+               }
+               time.Sleep(2 * time.Second)
+       }
        err := client.Set(ctx, key, []byte(val), time.Duration(60)*time.Second)
        if err != nil {
                msgs = append(msgs, fmt.Sprintf("unable to set a redis initialization key: %v", err))
