diff --git a/oauthproxy.go b/oauthproxy.go
index 10a69b31..cfe0f0ec 100644
--- a/oauthproxy.go
+++ b/oauthproxy.go
@@ -931,6 +931,7 @@ func (p *OAuthProxy) enrichSessionState(ctx context.Context, s *sessionsapi.Sess
 func (p *OAuthProxy) AuthOnly(rw http.ResponseWriter, req *http.Request) {
 	session, err := p.getAuthenticatedSession(rw, req)
 	if err != nil {
+		rw.Header().Set("X-Auth-Request-Result", "Denied")
 		http.Error(rw, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
 		return
 	}
@@ -945,7 +946,7 @@ func (p *OAuthProxy) AuthOnly(rw http.ResponseWriter, req *http.Request) {
 	// we are authenticated
 	p.addHeadersForProxying(rw, session)
 	p.headersChain.Then(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
-		rw.WriteHeader(http.StatusAccepted)
+		rw.WriteHeader(http.StatusOK)
 	})).ServeHTTP(rw, req)
 }
 
diff --git a/oauthproxy_test.go b/oauthproxy_test.go
index 0d8bc91..faa0d60 100644
--- a/oauthproxy_test.go
+++ b/oauthproxy_test.go
@@ -1092,7 +1092,7 @@ func TestAuthOnlyEndpointAccepted(t *testing.T) {
 	assert.NoError(t, err)
 
 	test.proxy.ServeHTTP(test.rw, test.req)
-	assert.Equal(t, http.StatusAccepted, test.rw.Code)
+	assert.Equal(t, http.StatusOK, test.rw.Code)
 	bodyBytes, _ := io.ReadAll(test.rw.Body)
 	assert.Equal(t, "", string(bodyBytes))
 }
@@ -1105,6 +1105,7 @@ func TestAuthOnlyEndpointUnauthorizedOnNoCookieSetError(t *testing.T) {
 
 	test.proxy.ServeHTTP(test.rw, test.req)
 	assert.Equal(t, http.StatusUnauthorized, test.rw.Code)
+	assert.Equal(t, "Denied", test.rw.Header().Get("X-Auth-Request-Result"))
 	bodyBytes, _ := io.ReadAll(test.rw.Body)
 	assert.Equal(t, "Unauthorized\n", string(bodyBytes))
 }
@@ -1125,6 +1126,7 @@ func TestAuthOnlyEndpointUnauthorizedOnExpiration(t *testing.T) {
 
 	test.proxy.ServeHTTP(test.rw, test.req)
 	assert.Equal(t, http.StatusUnauthorized, test.rw.Code)
+	assert.Equal(t, "Denied", test.rw.Header().Get("X-Auth-Request-Result"))
 	bodyBytes, _ := io.ReadAll(test.rw.Body)
 	assert.Equal(t, "Unauthorized\n", string(bodyBytes))
 }
@@ -1144,6 +1146,7 @@ func TestAuthOnlyEndpointUnauthorizedOnEmailValidationFailure(t *testing.T) {
 
 	test.proxy.ServeHTTP(test.rw, test.req)
 	assert.Equal(t, http.StatusUnauthorized, test.rw.Code)
+	assert.Equal(t, "Denied", test.rw.Header().Get("X-Auth-Request-Result"))
 	bodyBytes, _ := io.ReadAll(test.rw.Body)
 	assert.Equal(t, "Unauthorized\n", string(bodyBytes))
 }
@@ -1222,7 +1225,7 @@ func TestAuthOnlyEndpointSetXAuthRequestHeaders(t *testing.T) {
 	assert.NoError(t, err)
 
 	pcTest.proxy.ServeHTTP(pcTest.rw, pcTest.req)
-	assert.Equal(t, http.StatusAccepted, pcTest.rw.Code)
+	assert.Equal(t, http.StatusOK, pcTest.rw.Code)
 	assert.Equal(t, "oauth_user", pcTest.rw.Header().Get("X-Auth-Request-User"))
 	assert.Equal(t, startSession.Groups, pcTest.rw.Header().Values("X-Auth-Request-Groups"))
 	assert.Equal(t, "oauth_user@example.com", pcTest.rw.Header().Get("X-Auth-Request-Email"))
@@ -1315,7 +1318,7 @@ func TestAuthOnlyEndpointSetBasicAuthTrueRequestHeaders(t *testing.T) {
 	assert.NoError(t, err)
 
 	pcTest.proxy.ServeHTTP(pcTest.rw, pcTest.req)
-	assert.Equal(t, http.StatusAccepted, pcTest.rw.Code)
+	assert.Equal(t, http.StatusOK, pcTest.rw.Code)
 	assert.Equal(t, "oauth_user", pcTest.rw.Header().Values("X-Auth-Request-User")[0])
 	assert.Equal(t, "oauth_user@example.com", pcTest.rw.Header().Values("X-Auth-Request-Email")[0])
 	expectedHeader := "Basic " + base64.StdEncoding.EncodeToString([]byte("oauth_user:This is a secure password"))
@@ -1395,7 +1398,7 @@ func TestAuthOnlyEndpointSetBasicAuthFalseRequestHeaders(t *testing.T) {
 	assert.NoError(t, err)
 
 	pcTest.proxy.ServeHTTP(pcTest.rw, pcTest.req)
-	assert.Equal(t, http.StatusAccepted, pcTest.rw.Code)
+	assert.Equal(t, http.StatusOK, pcTest.rw.Code)
 	assert.Equal(t, "oauth_user", pcTest.rw.Header().Values("X-Auth-Request-User")[0])
 	assert.Equal(t, "oauth_user@example.com", pcTest.rw.Header().Values("X-Auth-Request-Email")[0])
 	assert.Equal(t, 0, len(pcTest.rw.Header().Values("Authorization")), "should not have Authorization header entries")
@@ -2908,21 +2911,21 @@ func TestAuthOnlyAllowedGroups(t *testing.T) {
 			allowedGroups:      []string{},
 			groups:             []string{},
 			querystring:        "",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "NoAllowedGroupsUserHasGroups",
 			allowedGroups:      []string{},
 			groups:             []string{"a", "b"},
 			querystring:        "",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserInAllowedGroup",
 			allowedGroups:      []string{"a"},
 			groups:             []string{"a", "b"},
 			querystring:        "",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserNotInAllowedGroup",
@@ -2936,28 +2939,28 @@ func TestAuthOnlyAllowedGroups(t *testing.T) {
 			allowedGroups:      []string{"a", "b"},
 			groups:             []string{"a", "c"},
 			querystring:        "?allowed_groups=a",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserInMultiParamQuerystringGroup",
 			allowedGroups:      []string{"a", "b"},
 			groups:             []string{"b"},
 			querystring:        "?allowed_groups=a&allowed_groups=b,d",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserInOnlyQuerystringGroup",
 			allowedGroups:      []string{},
 			groups:             []string{"a", "c"},
 			querystring:        "?allowed_groups=a,b",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserInDelimitedQuerystringGroup",
 			allowedGroups:      []string{"a", "b", "c"},
 			groups:             []string{"c"},
 			querystring:        "?allowed_groups=a,c",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserNotInQuerystringGroup",
@@ -3026,7 +3029,7 @@ func TestAuthOnlyAllowedGroupsWithSkipMethods(t *testing.T) {
 			method:             "OPTIONS",
 			ip:                 "1.2.3.5:43670",
 			withSession:        true,
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserWithGroupTrustedIp",
@@ -3034,7 +3037,7 @@ func TestAuthOnlyAllowedGroupsWithSkipMethods(t *testing.T) {
 			method:             "GET",
 			ip:                 "1.2.3.4:43670",
 			withSession:        true,
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserWithoutGroupSkipAuthPreflight",
@@ -3057,14 +3060,14 @@ func TestAuthOnlyAllowedGroupsWithSkipMethods(t *testing.T) {
 			method:             "OPTIONS",
 			ip:                 "1.2.3.5:43670",
 			withSession:        false,
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserWithoutSessionTrustedIp",
 			method:             "GET",
 			ip:                 "1.2.3.4:43670",
 			withSession:        false,
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 	}
 
@@ -3111,13 +3114,13 @@ func TestAuthOnlyAllowedEmailDomains(t *testing.T) {
 			name:               "NotEmailRestriction",
 			email:              "toto@example.com",
 			querystring:        "",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserInAllowedEmailDomain",
 			email:              "toto@example.com",
 			querystring:        "?allowed_email_domains=example.com",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserNotInAllowedEmailDomain",
@@ -3135,13 +3138,13 @@ func TestAuthOnlyAllowedEmailDomains(t *testing.T) {
 			name:               "UserInAllowedEmailDomains",
 			email:              "toto@example.com",
 			querystring:        "?allowed_email_domains=a.example.com,example.com",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserInAllowedEmailDomainWildcard",
 			email:              "toto@foo.example.com",
 			querystring:        "?allowed_email_domains=*.example.com",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserNotInAllowedEmailDomainWildcard",
@@ -3159,7 +3162,7 @@ func TestAuthOnlyAllowedEmailDomains(t *testing.T) {
 			name:               "UserInAllowedEmailDomainsWildcard",
 			email:              "toto@c.example.com",
 			querystring:        "?allowed_email_domains=a.b.c.example.com,*.c.example.com",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 	}
 
@@ -3202,13 +3205,13 @@ func TestAuthOnlyAllowedEmails(t *testing.T) {
 			name:               "NotEmailRestriction",
 			email:              "toto@example.com",
 			querystring:        "",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserInAllowedEmail",
 			email:              "toto@example.com",
 			querystring:        "?allowed_emails=toto@example.com",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 		{
 			name:               "UserNotInAllowedEmail",
@@ -3226,7 +3229,7 @@ func TestAuthOnlyAllowedEmails(t *testing.T) {
 			name:               "UserInAllowedEmails",
 			email:              "toto@example.com",
 			querystring:        "?allowed_emails=tete@example.com,toto@example.com",
-			expectedStatusCode: http.StatusAccepted,
+			expectedStatusCode: http.StatusOK,
 		},
 	}
 
