diff --git a/plugins/snapshots/erofs/erofs_linux.go b/plugins/snapshots/erofs/erofs_linux.go
index 02f6a10d2..25656951f 100644
--- a/plugins/snapshots/erofs/erofs_linux.go
+++ b/plugins/snapshots/erofs/erofs_linux.go
@@ -397,6 +397,10 @@ func (s *snapshotter) mounts(snap storage.Snapshot, info snapshots.Info, opts []
 		// If the lowerdir is actually an EROFS committed layer but
 		// doesn't have an EROFS mount.  Let's recover now.
 		if mntpoint != m.Source && !isErofs(mntpoint) {
+			// Prevents "Can't open blockdev" dmesg line before fallback.
+			if strings.HasSuffix(m.Source, ".erofs") {
+				m.Options = append(m.Options, "loop")
+			}
 			err := m.Mount(mntpoint)
 			// Use loop if the current kernel (6.12+) doesn't support file-backed mount
 			if errors.Is(err, unix.ENOTBLK) {
