diff --git a/etc/nginx/template/nginx.tmpl b/etc/nginx/template/nginx.tmpl
index b4c91ef54..226dee591 100755
--- a/etc/nginx/template/nginx.tmpl
+++ b/etc/nginx/template/nginx.tmpl
@@ -1177,7 +1177,7 @@ stream {
             # this location requires authentication
             auth_request        {{ $authPath }};
             auth_request_set    $auth_cookie $upstream_http_set_cookie;
-            add_header          Set-Cookie $auth_cookie;
+            add_header          Set-Cookie $auth_cookie always;
             {{- range $line := buildAuthResponseHeaders $proxySetHeader $externalAuth.ResponseHeaders }}
             {{ $line }}
             {{- end }}
