diff --git a/pkg/loki/modules.go b/pkg/loki/modules.go
--- a/pkg/loki/modules.go
+++ b/pkg/loki/modules.go
@@ -1678,7 +1678,7 @@
 		t.InternalServer.HTTP.PathPrefix(prefix).Handler(compactorHandler)
 	}
 
-	if t.Cfg.CompactorConfig.RetentionEnabled {
+	if t.Cfg.CompactorConfig.RetentionEnabled && isDeletingLogsAllowed() {
 		t.Server.HTTP.Path("/loki/api/v1/delete").Methods("PUT", "POST").Handler(t.addCompactorMiddleware(t.compactor.DeleteRequestsHandler.AddDeleteRequestHandler))
 		t.Server.HTTP.Path("/loki/api/v1/delete").Methods("GET").Handler(t.addCompactorMiddleware(t.compactor.DeleteRequestsHandler.GetAllDeleteRequestsHandler))
 		t.Server.HTTP.Path("/loki/api/v1/delete").Methods("DELETE").Handler(t.addCompactorMiddleware(t.compactor.DeleteRequestsHandler.CancelDeleteRequestHandler))
@@ -1689,6 +1689,12 @@
 	return t.compactor, nil
 }
 
+func isDeletingLogsAllowed() bool {
+	rawVal := os.Getenv("ALLOW_DELETE_LOGS")
+	val, _ := strconv.ParseBool(rawVal)
+	return val
+}
+
 func (t *Loki) addCompactorMiddleware(h http.HandlerFunc) http.Handler {
 	return middleware.Merge(t.HTTPAuthMiddleware, deletion.TenantMiddleware(t.Overrides)).Wrap(h)
 }
