diff --git a/pilot/pkg/credentials/kube/secrets.go b/pilot/pkg/credentials/kube/secrets.go
index fbde05a819..155ad0009f 100644
--- a/pilot/pkg/credentials/kube/secrets.go
+++ b/pilot/pkg/credentials/kube/secrets.go
@@ -76,6 +76,19 @@ type authorizationResponse struct {
 
 var _ credentials.Controller = &CredentialsController{}
 
+const (
+	istioNamespace = "d8-istio"
+)
+
+var (
+	allowedSecrets = map[string]bool{
+		"cacerts":                            true,
+		"d8-remote-clusters-public-metadata": true,
+		"d8-remote-authn-keypair":            true,
+		"d8-istio-sidecar-registry":          true,
+	}
+)
+
 func NewCredentialsController(kc kube.Client) *CredentialsController {
 	// We only care about TLS certificates and docker config for Wasm image pulling.
 	// Unfortunately, it is not as simple as selecting type=kubernetes.io/tls and type=kubernetes.io/dockerconfigjson.
@@ -87,9 +100,31 @@ func NewCredentialsController(kc kube.Client) *CredentialsController {
 	fieldSelector := fields.AndSelectors(
 		fields.OneTermNotEqualSelector("type", "helm.sh/release.v1"),
 		fields.OneTermNotEqualSelector("type", string(v1.SecretTypeServiceAccountToken))).String()
-	secrets := kclient.NewFiltered[*v1.Secret](kc, kclient.Filter{
+
+	objectFilter := func(obj any) bool {
+		secret, ok := obj.(*v1.Secret)
+		if !ok {
+			return true
+		}
+
+		secretName := secret.GetName()
+		secretNamespace := secret.GetNamespace()
+
+		if secretNamespace != istioNamespace {
+			return false
+		}
+
+		// Only allow specific secret names
+		return allowedSecrets[secretName]
+	}
+
+	filter := kclient.Filter{
+		Namespace:     istioNamespace,
 		FieldSelector: fieldSelector,
-	})
+		ObjectFilter:  objectFilter,
+	}
+	log.Infof("Creating secrets informer with namespace=%s, fieldSelector=%s", filter.Namespace, filter.FieldSelector)
+	secrets := kclient.NewFiltered[*v1.Secret](kc, filter)
 
 	return &CredentialsController{
 		secrets:            secrets,
