diff --git a/pilot/pkg/credentials/kube/secrets.go b/pilot/pkg/credentials/kube/secrets.go
index fbde05a819..b78b4cc956 100644
--- a/pilot/pkg/credentials/kube/secrets.go
+++ b/pilot/pkg/credentials/kube/secrets.go
@@ -88,7 +88,9 @@ func NewCredentialsController(kc kube.Client) *CredentialsController {
 		fields.OneTermNotEqualSelector("type", "helm.sh/release.v1"),
 		fields.OneTermNotEqualSelector("type", string(v1.SecretTypeServiceAccountToken))).String()
 	secrets := kclient.NewFiltered[*v1.Secret](kc, kclient.Filter{
+		Namespace:     "d8-istio",
 		FieldSelector: fieldSelector,
+		LabelSelector: "module=istio",
 	})
 
 	return &CredentialsController{
