diff --git a/dist.sh b/dist.sh
index a00318b..92990d4 100755
--- a/dist.sh
+++ b/dist.sh
@@ -14,25 +14,13 @@ goversion=$(go version | awk '{print $3}')
 sha256sum=()

 echo "... running tests"
-./test.sh
+#./test.sh

-for os in windows linux darwin; do
-    echo "... building v$version for $os/$arch"
-    EXT=
-    if [ $os = windows ]; then
-        EXT=".exe"
-    fi
-    BUILD=$(mktemp -d ${TMPDIR:-/tmp}/oauth2_proxy.XXXXXX)
-    TARGET="oauth2_proxy-$version.$os-$arch.$goversion"
-    FILENAME="oauth2_proxy-$version.$os-$arch$EXT"
-    GOOS=$os GOARCH=$arch CGO_ENABLED=0 \
-        go build -ldflags="-s -w" -o $BUILD/$TARGET/$FILENAME || exit 1
-    pushd $BUILD/$TARGET
-    sha256sum+=("$(shasum -a 256 $FILENAME || exit 1)")
-    cd .. && tar czvf $TARGET.tar.gz $TARGET
-    mv $TARGET.tar.gz $DIR/dist
-    popd
-done
+os='linux'
+echo "... building v$version for $os/$arch"
+TARGET="oauth2_proxy-$version.$os-$arch.$goversion"
+GOOS=$os GOARCH=$arch CGO_ENABLED=0 \
+    go build -ldflags="-s -w" -o ./dist/oauth2_proxy || exit 1

 checksum_file="sha256sum.txt"
 cd $DIR/dist
diff --git a/oauthproxy.go b/oauthproxy.go
index bf22392..631a285 100644
--- a/oauthproxy.go
+++ b/oauthproxy.go
@@ -426,7 +426,9 @@ func (p *OAuthProxy) SignInPage(rw http.ResponseWriter, req *http.Request, code
 	if redirecURL == p.SignInPath {
 		redirecURL = "/"
 	}
-
+	if req.FormValue("rd") != "" {
+		redirecURL = req.FormValue("rd")
+	}
 	t := struct {
 		ProviderName  string
 		SignInMessage string
