Subject: [PATCH] Added route table API locking
---
Index: pkg/cloudprovider/yandex/routes.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/cloudprovider/yandex/routes.go b/pkg/cloudprovider/yandex/routes.go
--- a/pkg/cloudprovider/yandex/routes.go	(revision 793ac2bd44d4c2eb02e6f62757db215518ef4ff9)
+++ b/pkg/cloudprovider/yandex/routes.go	(date 1674043080967)
@@ -2,9 +2,12 @@

 import (
 	"context"
+	"errors"
 	"fmt"
+	"sync"

 	"github.com/prometheus/common/log"
+	"k8s.io/klog/v2"

 	v1 "k8s.io/api/core/v1"

@@ -24,7 +27,18 @@
 	cpiNodeRoleLabel     = cpiRouteLabelsPrefix + "node-role" // we store Node's name here. The reason for this is lost in time (like tears in rain).
 )

+// these may get called in parallel, but since we have to modify the whole Route Table, we'll synchronize operations
+var routeAPILock sync.Mutex
+
 func (yc *Cloud) ListRoutes(ctx context.Context, _ string) ([]*cloudprovider.Route, error) {
+	klog.Info("ListRoutes called")
+
+	if routeAPILock.TryLock() {
+		defer routeAPILock.Unlock()
+	} else {
+		return nil, errors.New("VPC route API locked")
+	}
+
 	req := &vpc.GetRouteTableRequest{
 		RouteTableId: yc.config.RouteTableID,
 	}
@@ -45,35 +59,6 @@
 			continue
 		}

-		// let's verify NextHop relevance
-		currentNextHop := staticRoute.NextHop.(*vpc.StaticRoute_NextHopAddress).NextHopAddress
-		internalIP, err := yc.getInternalIpByNodeName(nodeName)
-		if err != nil {
-			log.Infof("Failed to verify NextHop relevance: %s", err)
-		} else if currentNextHop != internalIP {
-			log.Warnf("Changing %q's NextHop from %s to %s", nodeName, currentNextHop, internalIP)
-
-			filteredStaticRoutes := filterStaticRoutes(routeTable.StaticRoutes, routeFilterTerm{
-				termType:        routeFilterAddOrUpdate,
-				nodeName:        nodeName,
-				destinationCIDR: staticRoute.Destination.(*vpc.StaticRoute_DestinationPrefix).DestinationPrefix,
-				nextHop:         internalIP,
-			})
-
-			req := &vpc.UpdateRouteTableRequest{
-				RouteTableId: yc.config.RouteTableID,
-				UpdateMask: &field_mask.FieldMask{
-					Paths: []string{"static_routes"},
-				},
-				StaticRoutes: filteredStaticRoutes,
-			}
-
-			_, _, err := yc.yandexService.OperationWaiter(ctx, func() (*operation.Operation, error) { return yc.yandexService.VPCSvc.RouteTableSvc.Update(ctx, req) })
-			if err != nil {
-				return nil, err
-			}
-		}
-
 		cpiRoutes = append(cpiRoutes, &cloudprovider.Route{
 			Name:            nodeName,
 			TargetNode:      types.NodeName(nodeName),
@@ -87,6 +72,12 @@
 func (yc *Cloud) CreateRoute(ctx context.Context, _ string, _ string, route *cloudprovider.Route) error {
 	log.Infof("CreateRoute called with %+v", *route)

+	if routeAPILock.TryLock() {
+		defer routeAPILock.Unlock()
+	} else {
+		return errors.New("VPC route API locked")
+	}
+
 	rt, err := yc.yandexService.VPCSvc.RouteTableSvc.Get(ctx, &vpc.GetRouteTableRequest{RouteTableId: yc.config.RouteTableID})
 	if err != nil {
 		return err
@@ -120,6 +111,12 @@
 func (yc *Cloud) DeleteRoute(ctx context.Context, _ string, route *cloudprovider.Route) error {
 	log.Infof("DeleteRoute called with %+v", *route)

+	if routeAPILock.TryLock() {
+		defer routeAPILock.Unlock()
+	} else {
+		return errors.New("VPC route API locked")
+	}
+
 	rt, err := yc.yandexService.VPCSvc.RouteTableSvc.Get(ctx, &vpc.GetRouteTableRequest{RouteTableId: yc.config.RouteTableID})
 	if err != nil {
 		return err
