diff --git a/internal/ingress/controller/template/template.go b/internal/ingress/controller/template/template.go
index 7410ce6e0..aca7d3391 100644
--- a/internal/ingress/controller/template/template.go
+++ b/internal/ingress/controller/template/template.go
@@ -377,6 +377,12 @@ func configForLua(input interface{}) string {
 		return "{}"
 	}
 
+	proxyRealIPCIDRs, err := convertGoSliceIntoLuaTable(all.Cfg.ProxyRealIPCIDR, false)
+	if err != nil {
+		klog.Errorf("failed to convert %v into Lua table: %q", all.Cfg.ProxyRealIPCIDR, err)
+		proxyRealIPCIDRs = "{}"
+	}
+
 	return fmt.Sprintf(`{
 		use_forwarded_headers = %t,
 		use_proxy_protocol = %t,
@@ -389,6 +395,8 @@ func configForLua(input interface{}) string {
 		hsts_include_subdomains = %t,
 		hsts_preload = %t,
 
+		proxy_real_ip_cidr = %v,
+
 		global_throttle = {
 			memcached = {
 				host = "%v", port = %d, connect_timeout = %d, max_idle_timeout = %d, pool_size = %d,
@@ -408,6 +416,8 @@ func configForLua(input interface{}) string {
 		all.Cfg.HSTSIncludeSubdomains,
 		all.Cfg.HSTSPreload,
 
+		proxyRealIPCIDRs,
+
 		all.Cfg.GlobalRateLimitMemcachedHost,
 		all.Cfg.GlobalRateLimitMemcachedPort,
 		all.Cfg.GlobalRateLimitMemcachedConnectTimeout,
diff --git a/rootfs/etc/nginx/lua/lua_ingress.lua b/rootfs/etc/nginx/lua/lua_ingress.lua
index 49e0f5b05..998f73cc7 100644
--- a/rootfs/etc/nginx/lua/lua_ingress.lua
+++ b/rootfs/etc/nginx/lua/lua_ingress.lua
@@ -11,6 +11,7 @@ local string = string
 local original_randomseed = math.randomseed
 local string_format = string.format
 local ngx_redirect = ngx.redirect
+local iputils = require("resty.iputils")
 
 local _M = {}
 
@@ -103,6 +104,11 @@ function _M.init_worker()
 end
 
 function _M.set_config(new_config)
+  if new_config.use_forwarded_headers then
+    iputils.enable_lrucache()
+    new_config.real_ip_from = iputils.parse_cidrs(new_config.proxy_real_ip_cidr)
+  end
+
   config = new_config
 end
 
@@ -114,7 +120,7 @@ function _M.rewrite(location_config)
 
   ngx.var.best_http_host = ngx.var.http_host or ngx.var.host
 
-  if config.use_forwarded_headers then
+  if config.use_forwarded_headers and iputils.ip_in_cidrs(ngx.var.realip_remote_addr, config.real_ip_from) then
     -- trust http_x_forwarded_proto headers correctly indicate ssl offloading
     if ngx.var.http_x_forwarded_proto then
       ngx.var.pass_access_scheme = ngx.var.http_x_forwarded_proto
