diff --git a/go.mod b/go.mod
index c97bc9f1..83e0e53e 100644
--- a/go.mod
+++ b/go.mod
@@ -68,4 +68,11 @@ require (
 	sigs.k8s.io/kubetest2 v0.0.0-20210309183806-9230b4e73d8d // indirect
 )
 
+replace (
+	github.com/coreos/etcd => github.com/coreos/etcd v3.3.13+incompatible
+	go.etcd.io/bbolt => go.etcd.io/bbolt v1.3.5
+	go.etcd.io/etcd => go.etcd.io/etcd v0.5.0-alpha.5.0.20200910180754-dd1b699fc489 // ae9734ed278b is the SHA for git tag v3.4.13
+	google.golang.org/grpc => google.golang.org/grpc v1.27.1
+)
+
 go 1.13
