diff --git a/pkg/redirectpolicy/manager.go b/pkg/redirectpolicy/manager.go
index 9c12fd38cc..22ceb311e9 100644
--- a/pkg/redirectpolicy/manager.go
+++ b/pkg/redirectpolicy/manager.go
@@ -763,11 +763,12 @@ func (rpm *Manager) upsertService(config *LRPConfig, frontendMapping *feMapping)
 			Name:      config.id.Name + localRedirectSvcStr,
 			Namespace: config.id.Namespace,
 		},
-		Type:             lb.SVCTypeLocalRedirect,
-		Frontend:         frontendAddr,
-		Backends:         backendAddrs,
-		ExtTrafficPolicy: lb.SVCTrafficPolicyCluster,
-		IntTrafficPolicy: lb.SVCTrafficPolicyCluster,
+		Type:                  lb.SVCTypeLocalRedirect,
+		Frontend:              frontendAddr,
+		Backends:              backendAddrs,
+		ExtTrafficPolicy:      lb.SVCTrafficPolicyCluster,
+		IntTrafficPolicy:      lb.SVCTrafficPolicyCluster,
+		LoadBalancerAlgorithm: lb.SVCLoadBalancingAlgorithmRandom,
 	}
 
 	if _, _, err := rpm.svcManager.UpsertService(p); err != nil {
