From 865d11495cf32f0717cb0283108d3a9432ea32dc Mon Sep 17 00:00:00 2001
From: Dmitriy Andreychenko <dmitriy.andreychenko@flant.com>
Date: Wed, 20 Aug 2025 14:29:48 +0300
Subject: [PATCH] add debug logs for egress

Signed-off-by: Dmitriy Andreychenko <dmitriy.andreychenko@flant.com>
---
 pkg/egressgateway/manager.go | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/pkg/egressgateway/manager.go b/pkg/egressgateway/manager.go
index 971d84e3e7..da49cdc9da 100644
--- a/pkg/egressgateway/manager.go
+++ b/pkg/egressgateway/manager.go
@@ -722,7 +722,9 @@ func (manager *Manager) removeUnusedEgressRules() {
 // Whenever it encounters an error, it will just log it and move to the next
 // item, in order to reconcile as many states as possible.
 func (manager *Manager) reconcileLocked() {
+	fmt.Printf("DEBUG reconcileLocked START\n")
 	if !manager.allCachesSynced {
+		fmt.Printf("DEBUG reconcileLocked RETURN 111\n")
 		return
 	}
 
@@ -731,13 +733,19 @@ func (manager *Manager) reconcileLocked() {
 	// we don't know which k8s events/resources were received during the
 	// initial k8s sync
 	case manager.eventBitmapIsSet(eventUpdateEndpoint, eventDeleteEndpoint, eventUpdateNode, eventDeleteNode, eventK8sSyncDone):
+		fmt.Printf("DEBUG reconcileLocked eventUpdateEndpoint\n")
 		manager.updatePoliciesMatchedEndpointIDs()
 		fallthrough
 	case manager.eventBitmapIsSet(eventAddPolicy, eventDeletePolicy):
+		fmt.Printf("DEBUG reconcileLocked eventAddPolicy\n")
 		manager.updatePoliciesBySourceIP()
+	default:
+		fmt.Printf("DEBUG reconcileLocked default - no actions\n")
 	}
 
+	fmt.Printf("DEBUG reconcileLocked 222\n")
 	if manager.eventBitmapIsSet(eventK8sSyncDone, eventAddPolicy, eventDeletePolicy, eventUpdateNode, eventDeleteNode) {
+		fmt.Printf("DEBUG reconcileLocked 333\n")
 		manager.regenerateGatewayConfigs()
 
 		// Sysctl updates are handled by a reconciler, with the initial update attempting to wait some time
@@ -757,7 +765,9 @@ func (manager *Manager) reconcileLocked() {
 
 	// The order of the next 2 function calls matters, as by first adding missing policies and
 	// only then removing obsolete ones we make sure there will be no connectivity disruption
+	fmt.Printf("DEBUG reconcileLocked 444\n")
 	manager.addMissingEgressRules()
+	fmt.Printf("DEBUG reconcileLocked 555\n")
 	manager.removeUnusedEgressRules()
 
 	// clear the events bitmap
-- 
2.34.1

