diff --git a/pkg/endpoint/bpf.go b/pkg/endpoint/bpf.go
index 59e81f5bd1..98791af572 100644
--- a/pkg/endpoint/bpf.go
+++ b/pkg/endpoint/bpf.go
@@ -566,7 +566,7 @@ func (e *Endpoint) realizeBPFState(regenContext *regenerationContext) (err error
 		templateHash, err := e.owner.Orchestrator().ReloadDatapath(datapathRegenCtxt.completionCtx, datapathRegenCtxt.epInfoCache, &stats.datapathRealization)
 		if err != nil {
 			if !errors.Is(err, context.Canceled) {
-				e.getLogger().WithError(err).Error("Error while reloading endpoint BPF program")
+				e.getLogger().WithError(err).Fatal("Error while reloading endpoint BPF program")
 			}
 			return err
 		}
