diff --git a/pkg/egressgateway/manager.go b/pkg/egressgateway/manager.go
index 2911f396..821a685b 100644
--- a/pkg/egressgateway/manager.go
+++ b/pkg/egressgateway/manager.go
@@ -618,6 +618,9 @@ func (manager *Manager) relaxRPFilter() error {
 		}
 
 		ifaceName := pc.gatewayConfig.ifaceName
+		if ifaceName == "" {
+			continue
+		}
 		if _, ok := ifSet[ifaceName]; !ok {
 			ifSet[ifaceName] = struct{}{}
 			sysSettings = append(sysSettings, tables.Sysctl{
diff --git a/pkg/egressgateway/policy.go b/pkg/egressgateway/policy.go
index 36e8e871..dedd25f7 100644
--- a/pkg/egressgateway/policy.go
+++ b/pkg/egressgateway/policy.go
@@ -171,7 +171,10 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
 		gwc.egressIP = gc.egressIP
 		gwc.ifaceName, err = netdevice.GetIfaceWithIPv4Address(gc.egressIP)
 		if err != nil {
-			return fmt.Errorf("failed to retrieve interface with egress IP: %w", err)
+			expectedErr := fmt.Sprintf("no interface with %s IPv4 assigned to", gc.egressIP)
+			if err.Error() != expectedErr {
+				return fmt.Errorf("failed to retrieve interface with egress IP: %w", err)
+			}
 		}
 	default:
 		// If the gateway config doesn't specify any egress IP or interface, use the
