From 4cbd99a0e84f347757e6dd98ed45b9f29d278a49 Mon Sep 17 00:00:00 2001
From: Dmitriy Andreychenko <dmitriy.andreychenko@flant.com>
Date: Thu, 4 Sep 2025 13:18:30 +0300
Subject: [PATCH] ignore egress gateway inactual warning

Signed-off-by: Dmitriy Andreychenko <dmitriy.andreychenko@flant.com>
---
 pkg/egressgateway/policy.go | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/pkg/egressgateway/policy.go b/pkg/egressgateway/policy.go
index 36e8e871ad..fcce9c6a03 100644
--- a/pkg/egressgateway/policy.go
+++ b/pkg/egressgateway/policy.go
@@ -171,6 +171,10 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
 		gwc.egressIP = gc.egressIP
 		gwc.ifaceName, err = netdevice.GetIfaceWithIPv4Address(gc.egressIP)
 		if err != nil {
+			expectedErr := fmt.Sprintf("no interface with %s IPv4 assigned to", gc.egressIP)
+			if err.Error() != expectedErr {
+				return fmt.Errorf("failed to retrieve interface with egress IP: %w", err)
+			}
 			return fmt.Errorf("failed to retrieve interface with egress IP: %w", err)
 		}
 	default:
-- 
2.34.1

