diff --git a/bpf/lib/lb.h b/bpf/lib/lb.h
index 15dacfa4df..28cff178ee 100644
--- a/bpf/lib/lb.h
+++ b/bpf/lib/lb.h
@@ -1300,7 +1300,8 @@ lb4_extract_tuple(struct __ctx_buff *ctx, struct iphdr *ip4, int l3_off, int *l4
 			return ret;
 		return 0;
 	case IPPROTO_ICMP:
-		return DROP_UNSUPP_SERVICE_PROTO;
+		//return DROP_UNSUPP_SERVICE_PROTO;
+		return 0;
 	default:
 		return DROP_UNKNOWN_L4;
 	}
diff --git a/bpf/lib/nodeport.h b/bpf/lib/nodeport.h
index 0142716855..c67d3b6265 100644
--- a/bpf/lib/nodeport.h
+++ b/bpf/lib/nodeport.h
@@ -2679,6 +2679,72 @@ static __always_inline int nodeport_svc_lb4(struct __ctx_buff *ctx,
 				  ext_err);
 }
 
+static __always_inline 
+int make_icmp_responce_lb4(struct __ctx_buff *ctx,
+					struct iphdr *ip4,
+					int * l3_off) {
+	void *data, *data_end;
+	struct ethhdr *ethhdr;
+	struct icmphdr *icmphdr;
+	union macaddr smac = {};
+	union macaddr dmac = {};
+	__be32 tmp_addr;
+	__u8	tos;
+	__wsum csum;
+	/* changing size invalidates pointers, so we need to re-fetch them. */
+	data = ctx_data(ctx);
+	data_end = ctx_data_end(ctx);
+	printk("MEG: GOT A NEW ICMP HANDLER ENTER!\n");
+
+	if (data + sizeof(struct ethhdr) + sizeof(struct iphdr) + 
+		sizeof(struct icmphdr) > data_end)
+		return DROP_INVALID;
+
+	if (eth_load_saddr(ctx, smac.addr, 0) < 0)
+		return DROP_INVALID;
+
+	if (eth_load_daddr(ctx, dmac.addr, 0) < 0)
+		return DROP_INVALID;
+
+	/* Write reversed eth header, ready for egress */
+	ethhdr = data;
+	memcpy(ethhdr->h_dest, smac.addr, sizeof(smac.addr));
+	memcpy(ethhdr->h_source, dmac.addr, sizeof(dmac.addr));
+	ethhdr->h_proto = bpf_htons(ETH_P_IP);
+
+	/* Write reversed ip header, ready for egress */
+	//ip4 = data + sizeof(struct ethhdr);
+	ip4->version = 4;
+	ip4->ihl = sizeof(struct iphdr) >> 2;
+	//ip4->tos = tos;
+	//ip4->tot_len = bpf_htons(sizeof(struct iphdr) + sizeof(struct icmphdr) +
+	//	       (__u16)sample_len);
+	ip4->id = 0;
+	//ip4->frag_off = 0;
+	ip4->ttl = IPDEFTTL;
+	ip4->protocol = IPPROTO_ICMP;
+	ip4->check = 0;
+	tmp_addr = ip4->daddr;
+	ip4->daddr = ip4->saddr;
+	ip4->saddr = tmp_addr;
+	ip4->check = csum_fold(csum_diff(ip4, 0, ip4, sizeof(struct iphdr), 0));
+
+	/* Write reversed icmp header */
+	icmphdr = data + sizeof(struct ethhdr) + sizeof(struct iphdr);
+	icmphdr->type = ICMP_ECHOREPLY;
+	icmphdr->code = 0;
+	icmphdr->checksum = 0;
+	icmphdr->un.gateway = 0;
+
+	/* Add ICMP header checksum to sum of its body */
+	csum += csum_diff(icmphdr, 0, icmphdr, sizeof(struct icmphdr), 0);
+	icmphdr->checksum = csum_fold(csum);
+
+	/* Redirect ICMP to the interface we received it on. */
+	printk("MEG: GOT A NEW ICMP HANDLER DONE!\n");
+	return ctx_redirect(ctx, ctx_get_ifindex(ctx), 0);
+}
+
 /* Main node-port entry point for host-external ingressing node-port traffic
  * which handles the case of: i) backend is local EP, ii) backend is remote EP,
  * iii) reply from remote backend EP.
@@ -2716,6 +2782,11 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
 	lb4_fill_key(&key, &tuple);
 
 	svc = lb4_lookup_service(&key, false);
+
+	if(ip4->protocol == IPPROTO_ICMP && lb4_svc_is_loadbalancer(svc)) {
+		return make_icmp_responce_lb4(ctx, ip4, &l3_off);
+	}
+
 	if (svc) {
 		return nodeport_svc_lb4(ctx, &tuple, svc, &key, ip4, l3_off,
 					has_l4_header, l4_off,
