diff --git a/plugins/cilium-cni/main.go b/plugins/cilium-cni/main.go
index eaa50ed6f27d..340ffc3a1473 100644
--- a/plugins/cilium-cni/main.go
+++ b/plugins/cilium-cni/main.go
@@ -505,7 +505,7 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
 			peer      netlink.Link
 			tmpIfName string
 		)
-		veth, peer, tmpIfName, err = connector.SetupVeth(ep.ContainerID, int(conf.DeviceMTU), int(conf.GROMaxSize), int(conf.GSOMaxSize), ep)
+		veth, peer, tmpIfName, err = connector.SetupVeth(ep.ContainerID, int(conf.RouteMTU), int(conf.GROMaxSize), int(conf.GSOMaxSize), ep)
 		if err != nil {
 			err = fmt.Errorf("unable to set up veth on host side: %s", err)
 			return err
