diff --git a/pkg/ccm/loadbalancer.go b/pkg/ccm/loadbalancer.go
index e57f6c8..6de03db 100644
--- a/pkg/ccm/loadbalancer.go
+++ b/pkg/ccm/loadbalancer.go
@@ -30,6 +30,7 @@ const (
 	sslPortsAnnotation              = `service.beta.kubernetes.io/vcloud-avi-ssl-ports`
 	sslCertAliasAnnotation          = `service.beta.kubernetes.io/vcloud-avi-ssl-cert-alias`
 	skipAviSSLTerminationAnnotation = `service.beta.kubernetes.io/vcloud-avi-ssl-no-termination`
+	loadBalancerIPAnnotation        = `vcd.cpi.flant.com/load-balancer-ip`
 	// TODO: Update controlPlaneLabel to use default K8s constants if available
 	controlPlaneLabel = `node-role.kubernetes.io/control-plane`
 )
@@ -515,6 +516,10 @@ func shouldSkipAviSSLTermination(service *v1.Service) bool {
 
 // getUserSpecifiedLoadBalancerIP returns the specified load balancer IP
 func getUserSpecifiedLoadBalancerIP(service *v1.Service) string {
+	if ip, ok := service.Annotations[loadBalancerIPAnnotation]; ok && ip != "" {
+		return ip
+	}
+
 	return service.Spec.LoadBalancerIP
 }
 
