diff --git a/internal/backend/local/backend_apply.go b/internal/backend/local/backend_apply.go
index cde8fc93f8..bee8557e7f 100644
--- a/internal/backend/local/backend_apply.go
+++ b/internal/backend/local/backend_apply.go
@@ -30,7 +30,9 @@ import (
 var testHookStopPlanApply func()
 
 const (
-	defaultPersistInterval                 = 20 // arbitrary interval that's hopefully a sweet spot
+	// defaultPersistInterval
+	// we redeclare default for force persist
+	defaultPersistInterval                 = 0
 	persistIntervalEnvironmentVariableName = "TF_STATE_PERSIST_INTERVAL"
 )
 
diff --git a/internal/backend/local/hook_state.go b/internal/backend/local/hook_state.go
index 41bebef355..fe73fa858a 100644
--- a/internal/backend/local/hook_state.go
+++ b/internal/backend/local/hook_state.go
@@ -88,7 +88,8 @@ func (h *StateHook) PostStateUpdate(mutate func(*states.SyncState)) (tofu.HookAc
 			return tofu.HookActionHalt, err
 		}
 
-		if mgrPersist, ok := h.StateMgr.(statemgr.Persister); ok && h.PersistInterval != 0 && h.Schemas != nil {
+		// remove h.PersistInterval != 0 check for force persist
+		if mgrPersist, ok := h.StateMgr.(statemgr.Persister); ok && h.Schemas != nil {
 			if h.shouldPersist() {
 				err := mgrPersist.PersistState(context.TODO(), h.Schemas)
 				if err != nil {
