diff --git a/internal/iptables/detect.go b/internal/iptables/detect.go
index a87641b..b1b3715 100644
--- a/internal/iptables/detect.go
+++ b/internal/iptables/detect.go
@@ -22,12 +22,14 @@ import (
 )
 
 // DetectBinaryDir tries to detect the `iptables` location in
-// either /usr/sbin or /sbin. If it's not there, it returns an error.
+// either /usr/sbin, /sbin or /opt/deckhouse/bin. If it's not there, it returns an error.
 func DetectBinaryDir() (string, error) {
 	if files.ExecutableExists("/usr/sbin/iptables") {
 		return "/usr/sbin", nil
 	} else if files.ExecutableExists("/sbin/iptables") {
 		return "/sbin", nil
+	} else if files.ExecutableExists("/opt/deckhouse/bin/xtables-nft-multi") {
+		return "/opt/deckhouse/bin", nil
 	} else {
 		return "", errors.New("iptables is not present in either /usr/sbin or /sbin")
 	}
