Index: internal/k8s/k8s.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/k8s/k8s.go b/internal/k8s/k8s.go
--- a/internal/k8s/k8s.go	(revision 99469a412510da616538825c7a5ecb1ff0dbc59d)
+++ b/internal/k8s/k8s.go	(date 1629756192370)
@@ -27,6 +27,10 @@
 	"k8s.io/client-go/util/workqueue"
 )

+const (
+	LabelNodeExcludeBalancers = "node.kubernetes.io/exclude-from-external-load-balancers"
+)
+
 // Client watches a Kubernetes cluster and translates events into
 // Controller method calls.
 type Client struct {
@@ -570,6 +574,22 @@
 	}
 }

+func (c *Client) HasExcludeLBLabel(nodeName string) (bool, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10 * time.Second)
+	defer cancel()
+
+	node, err := c.client.CoreV1().Nodes().Get(ctx, nodeName, metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	if _, ok := node.Labels[LabelNodeExcludeBalancers]; ok {
+		return true, nil
+	}
+
+	return false, nil
+}
+
 func serviceKeyForSlice(endpointSlice *discovery.EndpointSlice) (svcKey, error) {
 	if endpointSlice == nil {
 		return "", fmt.Errorf("nil EndpointSlice")
Index: speaker/bgp_controller_test.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/speaker/bgp_controller_test.go b/speaker/bgp_controller_test.go
--- a/speaker/bgp_controller_test.go	(revision 99469a412510da616538825c7a5ecb1ff0dbc59d)
+++ b/speaker/bgp_controller_test.go	(date 1629756214110)
@@ -179,6 +179,10 @@
 	panic("never called")
 }

+func (s *testK8S) HasExcludeLBLabel(nodeName string) (bool, error) {
+	return false, nil
+}
+
 func (s *testK8S) Infof(_ *v1.Service, evtType string, msg string, args ...interface{}) {
 	s.t.Logf("k8s Info event %q: %s", evtType, fmt.Sprintf(msg, args...))
 }
Index: speaker/main.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/speaker/main.go b/speaker/main.go
--- a/speaker/main.go	(revision 99469a412510da616538825c7a5ecb1ff0dbc59d)
+++ b/speaker/main.go	(date 1629756192410)
@@ -53,6 +53,7 @@
 	UpdateStatus(svc *v1.Service) error
 	Infof(svc *v1.Service, desc, msg string, args ...interface{})
 	Errorf(svc *v1.Service, desc, msg string, args ...interface{})
+	HasExcludeLBLabel(nodeName string) (bool, error)
 }

 func main() {
@@ -142,6 +143,9 @@
 		os.Exit(1)
 	}
 	ctrl.client = client
+	for _, ctrlImpl := range ctrl.protocols {
+		ctrlImpl.setService(client)
+	}

 	sList.Start(client)
 	defer sList.Stop()
@@ -371,6 +375,7 @@
 	SetBalancer(gokitlog.Logger, string, net.IP, *config.Pool) error
 	DeleteBalancer(gokitlog.Logger, string, string) error
 	SetNode(gokitlog.Logger, *v1.Node) error
+	setService(service)
 }

 // Speakerlist represents a list of healthy speakers.
Index: speaker/bgp_controller.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/speaker/bgp_controller.go b/speaker/bgp_controller.go
--- a/speaker/bgp_controller.go	(revision 99469a412510da616538825c7a5ecb1ff0dbc59d)
+++ b/speaker/bgp_controller.go	(date 1629759141370)
@@ -43,6 +43,11 @@
 	nodeLabels labels.Set
 	peers      []*peer
 	svcAds     map[string][]*bgp.Advertisement
+	service service
+}
+
+func (c *bgpController) setService(svc service) {
+	c.service = svc
 }

 func (c *bgpController) SetConfig(l log.Logger, cfg *config.Config) error {
@@ -168,10 +173,13 @@
 		// First, determine if the peering should be active for this
 		// node.
 		shouldRun := false
-		for _, ns := range p.cfg.NodeSelectors {
-			if ns.Matches(c.nodeLabels) {
-				shouldRun = true
-				break
+
+		if ok, _ := c.service.HasExcludeLBLabel(c.myNode); !ok {
+			for _, ns := range p.cfg.NodeSelectors {
+				if ns.Matches(c.nodeLabels) {
+					shouldRun = true
+					break
+				}
 			}
 		}

Index: speaker/layer2_controller.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/speaker/layer2_controller.go b/speaker/layer2_controller.go
--- a/speaker/layer2_controller.go	(revision 99469a412510da616538825c7a5ecb1ff0dbc59d)
+++ b/speaker/layer2_controller.go	(date 1629756192398)
@@ -31,6 +31,11 @@
 	announcer *layer2.Announce
 	myNode    string
 	sList     SpeakerList
+	service service
+}
+
+func (c *layer2Controller) setService(svc service) {
+	c.service = svc
 }

 func (c *layer2Controller) SetConfig(log.Logger, *config.Config) error {
