Index: pkg/plugins/trivy/plugin.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/plugins/trivy/plugin.go b/pkg/plugins/trivy/plugin.go
--- a/pkg/plugins/trivy/plugin.go	(revision 4b129ca41ec94f5e086e449af77d9bc4e5d3aadb)
+++ b/pkg/plugins/trivy/plugin.go	(date 1677581681674)
@@ -543,6 +543,9 @@
 }

 const (
+	dockerConfigVolumeName      = "docker-config"
+	dockerConfigSecret          = "deckhouse-registry"
+	dockerConfigMountPath       = "/root/.docker"
 	tmpVolumeName               = "tmp"
 	ignoreFileVolumeName        = "ignorefile"
 	ignoreFileName              = ".trivyignore"
@@ -623,6 +626,11 @@
 				MountPath: "/tmp",
 				ReadOnly:  false,
 			},
+			{
+				Name:      dockerConfigVolumeName,
+				MountPath: dockerConfigMountPath,
+				ReadOnly:  true,
+			},
 		},
 	}

@@ -644,6 +652,18 @@
 				},
 			},
 		},
+		{
+			Name: dockerConfigVolumeName,
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: dockerConfigSecret,
+					Items: []corev1.KeyToPath{{
+						Key:  ".dockerconfigjson",
+						Path: "config.json",
+					}},
+				},
+			},
+		},
 	}
 	volumeMounts = append(volumeMounts, getScanResultVolumeMount())
 	volumes = append(volumes, getScanResultVolume())
