diff --git a/cmd/acmesolver/go.mod b/cmd/acmesolver/go.mod
index 4b7d6cb..87ea6d4 100644
--- a/cmd/acmesolver/go.mod
+++ b/cmd/acmesolver/go.mod
@@ -2,6 +2,8 @@ module github.com/cert-manager/cert-manager/acmesolver-binary
 
 go 1.25.0
 
+toolchain go1.25.11
+
 // Do not remove this comment:
 // please place any replace statements here at the top for visibility and add a
 // comment to it as to when it can be removed
diff --git a/cmd/cainjector/go.mod b/cmd/cainjector/go.mod
index 7632608..520ccad 100644
--- a/cmd/cainjector/go.mod
+++ b/cmd/cainjector/go.mod
@@ -2,6 +2,8 @@ module github.com/cert-manager/cert-manager/cainjector-binary
 
 go 1.25.0
 
+toolchain go1.25.11
+
 // Do not remove this comment:
 // please place any replace statements here at the top for visibility and add a
 // comment to it as to when it can be removed
diff --git a/cmd/controller/go.mod b/cmd/controller/go.mod
index 4c80109..dfce00a 100644
--- a/cmd/controller/go.mod
+++ b/cmd/controller/go.mod
@@ -2,6 +2,8 @@ module github.com/cert-manager/cert-manager/controller-binary
 
 go 1.25.0
 
+toolchain go1.25.11
+
 // Do not remove this comment:
 // please place any replace statements here at the top for visibility and add a
 // comment to it as to when it can be removed
diff --git a/cmd/webhook/go.mod b/cmd/webhook/go.mod
index d5e0954..c4bf83d 100644
--- a/cmd/webhook/go.mod
+++ b/cmd/webhook/go.mod
@@ -2,6 +2,8 @@ module github.com/cert-manager/cert-manager/webhook-binary
 
 go 1.25.0
 
+toolchain go1.25.11
+
 // Do not remove this comment:
 // please place any replace statements here at the top for visibility and add a
 // comment to it as to when it can be removed
diff --git a/go.mod b/go.mod
index a4b3e59..b58a1b3 100644
--- a/go.mod
+++ b/go.mod
@@ -2,6 +2,8 @@ module github.com/cert-manager/cert-manager
 
 go 1.25.0
 
+toolchain go1.25.11
+
 // Do not remove this comment:
 // please place any replace statements here at the top for visibility and add a
 // comment to it as to when it can be removed
