diff --git a/cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go b/cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go
index 8fc793cca59..337b0da4886 100644
--- a/cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go
+++ b/cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go
@@ -74,6 +74,7 @@ func NewControlPlaneJoinPhase() workflow.Phase {
 				ArgsValidator:  cobra.NoArgs,
 			},
 			newEtcdLocalSubphase(),
+			NewEtcdJoinPhase(),
 			newMarkControlPlaneSubphase(),
 		},
 	}
@@ -90,7 +91,7 @@ func NewEtcdJoinPhase() workflow.Phase {
 		ArgsValidator: cobra.NoArgs,
 		// TODO: unhide this phase once ControlPlaneKubeletLocalMode goes GA:
 		// https://github.com/kubernetes/enhancements/issues/4471
-		Hidden: true,
+		Hidden: false,
 		// Only run this phase as if `ControlPlaneKubeletLocalMode` is activated.
 		RunIf: func(c workflow.RunData) (bool, error) {
 			return checkFeatureState(c, features.ControlPlaneKubeletLocalMode, true)
