diff --git a/pkg/cloudprovider/vsphere/nodemanager.go b/pkg/cloudprovider/vsphere/nodemanager.go
index be3b693..9e86ebd 100644
--- a/pkg/cloudprovider/vsphere/nodemanager.go
+++ b/pkg/cloudprovider/vsphere/nodemanager.go
@@ -355,6 +355,17 @@ func (nm *NodeManager) DiscoverNode(nodeID string, searchBy cm.FindVM) error {
 		}
 	}
 
+	var internalAddressExists bool
+	for _, addr := range addrs {
+		if addr.Type == v1.NodeInternalIP {
+			internalAddressExists = true
+			break
+		}
+	}
+	if !internalAddressExists {
+		return fmt.Errorf("no Internal IP found in the node %s addresses:\n%+v", nodeID, addrs)
+	}
+
 	klog.V(2).Infof("Found node %s as vm=%+v in vc=%s and datacenter=%s",
 		nodeID, vmDI.VM, vmDI.VcServer, vmDI.DataCenter.Name())
 	klog.V(2).Info("Hostname: ", oVM.Guest.HostName, " UUID: ", vmDI.UUID)
