diff --git a/server/handlers.go b/server/handlers.go
index eb65f490..a939b62e 100644
--- a/server/handlers.go
+++ b/server/handlers.go
@@ -14,6 +14,8 @@ import (
 	"strings"
 	"time"
 
+	"github.com/dexidp/dex/pkg/groups"
+
 	"github.com/coreos/go-oidc/v3/oidc"
 	"github.com/gorilla/mux"
 	jose "gopkg.in/square/go-jose.v2"
@@ -243,6 +245,22 @@ func (s *Server) handleConnectorLogin(w http.ResponseWriter, r *http.Request) {
 	scopes := parseScopes(authReq.Scopes)
 	showBacklink := len(s.connectors) > 1
 
+	client, err := s.storage.GetClient(authReq.ClientID)
+	if err != nil {
+		s.logger.Errorf("Failed to get client %q: %v", authReq.ClientID, err)
+		s.renderError(r, w, http.StatusInternalServerError, "Failed to retrieve client.")
+		return
+	}
+
+	if len(client.AllowedGroups) > 0 {
+		authReq.Claims.Groups = groups.Filter(authReq.Claims.Groups, client.AllowedGroups)
+		if len(authReq.Claims.Groups) == 0 {
+			s.logger.Errorf("user not in allowed groups: %v", client.AllowedGroups)
+			s.renderError(r, w, http.StatusInternalServerError, "User not in allowed groups.")
+			return
+		}
+	}
+
 	switch r.Method {
 	case http.MethodGet:
 		switch conn := conn.Connector.(type) {
diff --git a/storage/kubernetes/types.go b/storage/kubernetes/types.go
index 07e25084..8aefe485 100644
--- a/storage/kubernetes/types.go
+++ b/storage/kubernetes/types.go
@@ -197,6 +197,8 @@ type Client struct {
 
 	Name    string `json:"name,omitempty"`
 	LogoURL string `json:"logoURL,omitempty"`
+
+	AllowedGroups []string `json:"allowedGroups,omitempty"`
 }
 
 // ClientList is a list of Clients.
@@ -216,25 +218,27 @@ func (cli *client) fromStorageClient(c storage.Client) Client {
 			Name:      cli.idToName(c.ID),
 			Namespace: cli.namespace,
 		},
-		ID:           c.ID,
-		Secret:       c.Secret,
-		RedirectURIs: c.RedirectURIs,
-		TrustedPeers: c.TrustedPeers,
-		Public:       c.Public,
-		Name:         c.Name,
-		LogoURL:      c.LogoURL,
+		ID:            c.ID,
+		Secret:        c.Secret,
+		RedirectURIs:  c.RedirectURIs,
+		TrustedPeers:  c.TrustedPeers,
+		Public:        c.Public,
+		Name:          c.Name,
+		LogoURL:       c.LogoURL,
+		AllowedGroups: c.AllowedGroups,
 	}
 }
 
 func toStorageClient(c Client) storage.Client {
 	return storage.Client{
-		ID:           c.ID,
-		Secret:       c.Secret,
-		RedirectURIs: c.RedirectURIs,
-		TrustedPeers: c.TrustedPeers,
-		Public:       c.Public,
-		Name:         c.Name,
-		LogoURL:      c.LogoURL,
+		ID:            c.ID,
+		Secret:        c.Secret,
+		RedirectURIs:  c.RedirectURIs,
+		TrustedPeers:  c.TrustedPeers,
+		Public:        c.Public,
+		Name:          c.Name,
+		LogoURL:       c.LogoURL,
+		AllowedGroups: c.AllowedGroups,
 	}
 }
 
diff --git a/storage/storage.go b/storage/storage.go
index c308ac46..76ff386f 100644
--- a/storage/storage.go
+++ b/storage/storage.go
@@ -164,6 +164,8 @@ type Client struct {
 	// Name and LogoURL used when displaying this client to the end user.
 	Name    string `json:"name" yaml:"name"`
 	LogoURL string `json:"logoURL" yaml:"logoURL"`
+
+	AllowedGroups []string `json:"allowedGroups,omitempty" yaml:"allowedGroups,omitempty"`
 }
 
 // Claims represents the ID Token claims supported by the server.
