diff --git a/server/handlers.go b/server/handlers.go
index d961dbe..4651bf6 100644
--- a/server/handlers.go
+++ b/server/handlers.go
@@ -354,7 +354,7 @@ func (s *Server) handleConnectorLogin(w http.ResponseWriter, r *http.Request) {
 		identity, ok, err := passwordConnector.Login(r.Context(), scopes, username, password)
 		if err != nil {
 			s.logger.Errorf("Failed to login user: %v", err)
-			s.renderError(w, http.StatusInternalServerError, "Login error.")
+			s.renderError(w, http.StatusInternalServerError, fmt.Sprintf("Login error: %s", err))
 			return
 		}
 		if !ok {
