Subject: [PATCH] remove pods in deckhouse namespaces
---
Index: pkg/framework/plugins/defaultevictor/defaultevictor.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/framework/plugins/defaultevictor/defaultevictor.go b/pkg/framework/plugins/defaultevictor/defaultevictor.go
--- a/pkg/framework/plugins/defaultevictor/defaultevictor.go	(revision 35f4ebae24b3a4dbcd15ca64bd208660ebd66661)
+++ b/pkg/framework/plugins/defaultevictor/defaultevictor.go	(date 1724597378535)
@@ -18,6 +18,7 @@
 	"context"
 	"errors"
 	"fmt"
+	"strings"
 	"time"

 	v1 "k8s.io/api/core/v1"
@@ -79,6 +80,14 @@
 		args:   defaultEvictorArgs,
 	}

+	// don't use pods from d8-* and kube-system namespaces
+	ev.constraints = append(ev.constraints, func(pod *v1.Pod) error {
+		if strings.HasPrefix(pod.Namespace, "d8-") || pod.Namespace == "kube-system" {
+			return fmt.Errorf("pod in the deckhouse namespace(%s)", pod.Namespace)
+		}
+		return nil
+	})
+
 	if defaultEvictorArgs.EvictFailedBarePods {
 		klog.V(1).InfoS("Warning: EvictFailedBarePods is set to True. This could cause eviction of pods without ownerReferences.")
 		ev.constraints = append(ev.constraints, func(pod *v1.Pod) error {
