diff --git a/pkg/providers/v1/aws.go b/pkg/providers/v1/aws.go
index e1070ee..feff012 100644
--- a/pkg/providers/v1/aws.go
+++ b/pkg/providers/v1/aws.go
@@ -2733,6 +2733,13 @@ func (c *Cloud) updateInstanceSecurityGroupsForLoadBalancer(ctx context.Context,
 	c.sortELBSecurityGroupList(lbSecurityGroupIDs, annotations, taggedLBSecurityGroups)
 	loadBalancerSecurityGroupID := lbSecurityGroupIDs[0]
 
+	var deleteSecurityGroupID bool
+	// we shouldn't delete Ingress SG rule, if it allows access
+	// from configured "ElbSecurityGroup", so that we won't disrupt access to Nodes from other ELBs
+	if loadBalancerSecurityGroupID != c.cfg.Global.ElbSecurityGroup {
+		deleteSecurityGroupID = true
+	}
+
 	// Get the actual list of groups that allow ingress from the load-balancer
 	actualGroups := make(map[*ec2types.SecurityGroup]bool)
 	{
@@ -2824,7 +2831,7 @@ func (c *Cloud) updateInstanceSecurityGroupsForLoadBalancer(ctx context.Context,
 			if !changed {
 				klog.Warning("Allowing ingress was not needed; concurrent change? groupId=", instanceSecurityGroupID)
 			}
-		} else {
+		} else if deleteSecurityGroupID {
 			changed, err := c.removeSecurityGroupIngress(ctx, instanceSecurityGroupID, permissions)
 			if err != nil {
 				return err
