Subject: [PATCH] Add TemplateOrg field to VCDMachine spec
---
Index: controllers/vcdmachine_controller.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/controllers/vcdmachine_controller.go b/controllers/vcdmachine_controller.go
--- a/controllers/vcdmachine_controller.go	(revision 2fa5bc2f49acddad1ffad59d589afc8b2074304f)
+++ b/controllers/vcdmachine_controller.go	(date 1747229735331)
@@ -552,7 +552,7 @@
 `
 		// vcda-4391 fixed
 		err = vdcManager.AddNewVM(vmName, vcdCluster.Name, 1,
-			vcdMachine.Spec.Catalog, vcdMachine.Spec.Template, vcdMachine.Spec.PlacementPolicy,
+			vcdMachine.Spec.Catalog, vcdMachine.Spec.Template, vcdMachine.Spec.TemplateOrg, vcdMachine.Spec.PlacementPolicy,
 			vcdMachine.Spec.SizingPolicy, vcdMachine.Spec.StorageProfile, guestCustScript, false)
 		if err != nil {
 			err1 := capvcdRdeManager.AddToErrorSet(ctx, capisdk.VCDMachineCreationError, "", machine.Name, fmt.Sprintf("%v", err))
Index: api/v1beta2/vcdmachine_types.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/api/v1beta2/vcdmachine_types.go b/api/v1beta2/vcdmachine_types.go
--- a/api/v1beta2/vcdmachine_types.go	(revision 2fa5bc2f49acddad1ffad59d589afc8b2074304f)
+++ b/api/v1beta2/vcdmachine_types.go	(date 1747229735323)
@@ -46,6 +46,10 @@
 	// +optional
 	Template string `json:"template,omitempty"`
 
+	// TemplateOrg is the organization of the template OVA that is to be used
+	// +optional
+	TemplateOrg string `json:"templateOrg,omitempty"`
+
 	// SizingPolicy is the sizing policy to be used on this machine.
 	// If no sizing policy is specified, default sizing policy will be used to create the nodes
 	// +optional
