diff --git a/tests/e2e/utils/cluster_upgrade_utils.go b/tests/e2e/utils/cluster_upgrade_utils.go
index a1e6fc4..e9a2091 100644
--- a/tests/e2e/utils/cluster_upgrade_utils.go
+++ b/tests/e2e/utils/cluster_upgrade_utils.go
@@ -12,7 +12,7 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/util/wait"
 	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
-	clusterv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterv1beta1 "sigs.k8s.io/cluster-api/api/core/v1beta2"
 	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
 	"strings"
 	"time"
diff --git a/tests/e2e/utils/node_pool_scaling_utils.go b/tests/e2e/utils/node_pool_scaling_utils.go
index 7736b53..3985789 100644
--- a/tests/e2e/utils/node_pool_scaling_utils.go
+++ b/tests/e2e/utils/node_pool_scaling_utils.go
@@ -6,7 +6,7 @@ import (
 	"github.com/vmware/cloud-provider-for-cloud-director/pkg/testingsdk"
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/util/wait"
-	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/core/v1beta2"
 	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"time"
diff --git a/tests/e2e/workload_cluster_resize_test.go b/tests/e2e/workload_cluster_resize_test.go
index c37601d..cd446b6 100644
--- a/tests/e2e/workload_cluster_resize_test.go
+++ b/tests/e2e/workload_cluster_resize_test.go
@@ -14,7 +14,7 @@ import (
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
 	"os"
-	clusterv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterv1beta1 "sigs.k8s.io/cluster-api/api/core/v1beta2"
 	kcfg "sigs.k8s.io/cluster-api/util/kubeconfig"
 	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
diff --git a/tests/e2e/workload_cluster_upgrade_test.go b/tests/e2e/workload_cluster_upgrade_test.go
index 1e170de..1f6e557 100644
--- a/tests/e2e/workload_cluster_upgrade_test.go
+++ b/tests/e2e/workload_cluster_upgrade_test.go
@@ -18,7 +18,7 @@ import (
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
 	"os"
-	clusterv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterv1beta1 "sigs.k8s.io/cluster-api/api/core/v1beta2"
 	kcfg "sigs.k8s.io/cluster-api/util/kubeconfig"
 	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
 	"strings"
