diff --git a/pkg/cloudprovider/yandex/load_balancer.go b/pkg/cloudprovider/yandex/load_balancer.go
index 7b09147..16ccb70 100644
--- a/pkg/cloudprovider/yandex/load_balancer.go
+++ b/pkg/cloudprovider/yandex/load_balancer.go
@@ -19,6 +19,7 @@ const (
 	externalLoadBalancerAnnotation        = "yandex.cpi.flant.com/loadbalancer-external"
 	listenerSubnetIdAnnotation            = "yandex.cpi.flant.com/listener-subnet-id"
 	listenerAddressIPv4                   = "yandex.cpi.flant.com/listener-address-ipv4"
+	loadBalancerInternal                  = "yandex.cpi.flant.com/loadbalancer-internal"
 
 	nodesHealthCheckPath = "/healthz"
 	// NOTE: Please keep the following port in sync with ProxyHealthzPort in pkg/cluster/ports/ports.go
@@ -216,9 +217,16 @@ func (yc *Cloud) getLoadBalancerParameters(svc *v1.Service) (lbParams loadBalanc
 		lbParams.internal = true
 		lbParams.listenerSubnetID = value
 	} else if len(yc.config.lbListenerSubnetID) != 0 {
+		lbParams.internal = true
 		lbParams.listenerSubnetID = yc.config.lbListenerSubnetID
-		_, isExternal := svc.ObjectMeta.Annotations[externalLoadBalancerAnnotation]
-		lbParams.internal = !isExternal
+
+		if _, isExternal := svc.ObjectMeta.Annotations[externalLoadBalancerAnnotation]; isExternal {
+			lbParams.internal = false
+		}
+
+		if _, isInternal := svc.ObjectMeta.Annotations[loadBalancerInternal]; isInternal {
+			lbParams.internal = true
+		}
 	}
 
 	if value, ok := svc.ObjectMeta.Annotations[targetGroupNetworkIdAnnotation]; ok {
