diff --git a/pkg/capisdk/defined_entity.go b/pkg/capisdk/defined_entity.go
index 0a397fed..a25850d0 100644
--- a/pkg/capisdk/defined_entity.go
+++ b/pkg/capisdk/defined_entity.go
@@ -13,7 +13,7 @@ import (
 	"github.com/vmware/cluster-api-provider-cloud-director/pkg/vcdtypes/rde_type_1_1_0"
 	rdeType "github.com/vmware/cluster-api-provider-cloud-director/pkg/vcdtypes/rde_type_1_1_0"
 	"github.com/vmware/cluster-api-provider-cloud-director/release"
-	"k8s.io/klog"
+	"k8s.io/klog/v2"
 	"net/http"
 	"net/url"
 	"reflect"
