--- controllers/capi_objects_utils.go	2026-03-20 13:07:51
+++ b/controllers/capi_objects_utils.go	2026-03-20 13:24:25
@@ -12,7 +12,7 @@
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
 	clusterv1 "sigs.k8s.io/cluster-api/api/core/v1beta2"
-	"sigs.k8s.io/cluster-api/api/bootstrap/kubeadm/v1beta2"
+	kubeadmbootstrapv1 "sigs.k8s.io/cluster-api/api/bootstrap/kubeadm/v1beta2"
 	kcpv1 "sigs.k8s.io/cluster-api/api/controlplane/kubeadm/v1beta2"
 	addonsv1 "sigs.k8s.io/cluster-api/api/addons/v1beta2"
 	"sigs.k8s.io/controller-runtime/pkg/client"
@@ -332,10 +332,10 @@
 }
 
 func getVCDMachineTemplateFromKCP(ctx context.Context, cli client.Client, kcp kcpv1.KubeadmControlPlane) (*infrav1.VCDMachineTemplate, error) {
-	vcdMachineTemplateRef := kcp.Spec.MachineTemplate.InfrastructureRef
+	vcdMachineTemplateRef := kcp.Spec.MachineTemplate.Spec.InfrastructureRef
 	vcdMachineTemplate := &infrav1.VCDMachineTemplate{}
 	vcdMachineTemplateKey := types.NamespacedName{
-		Namespace: vcdMachineTemplateRef.Namespace,
+		Namespace: kcp.Namespace,
 		Name:      vcdMachineTemplateRef.Name,
 	}
 	if err := cli.Get(ctx, vcdMachineTemplateKey, vcdMachineTemplate); err != nil {
@@ -349,7 +349,7 @@
 	vcdMachineTemplateRef := md.Spec.Template.Spec.InfrastructureRef
 	vcdMachineTemplate := &infrav1.VCDMachineTemplate{}
 	vcdMachineTemplateKey := client.ObjectKey{
-		Namespace: vcdMachineTemplateRef.Namespace,
+		Namespace: md.Namespace,
 		Name:      vcdMachineTemplateRef.Name,
 	}
 	if err := cli.Get(ctx, vcdMachineTemplateKey, vcdMachineTemplate); err != nil {
@@ -381,8 +381,8 @@
 	return vcdMachineTemplate, nil
 }
 
-func getKubeadmConfigTemplateByObjRef(ctx context.Context, cli client.Client, objRef v1.ObjectReference) (*v1beta1.KubeadmConfigTemplate, error) {
-	kubeadmConfigTemplate := &v1beta1.KubeadmConfigTemplate{}
+func getKubeadmConfigTemplateByObjRef(ctx context.Context, cli client.Client, objRef v1.ObjectReference) (*kubeadmbootstrapv1.KubeadmConfigTemplate, error) {
+	kubeadmConfigTemplate := &kubeadmbootstrapv1.KubeadmConfigTemplate{}
 	kubeadmConfigTemplateKey := client.ObjectKey{
 		Namespace: objRef.Namespace,
 		Name:      objRef.Name,
@@ -447,6 +447,10 @@
 		if md.Spec.Replicas != nil {
 			desiredReplicasCount = *md.Spec.Replicas
 		}
+		availableReplicas := int32(0)
+		if md.Status.ReadyReplicas != nil {
+			availableReplicas = *md.Status.ReadyReplicas
+		}
 		nodePool := rdeType.NodePool{
 			Name:              md.Name,
 			SizingPolicy:      vcdMachineTemplate.Spec.Template.Spec.SizingPolicy,
@@ -455,7 +459,7 @@
 			StorageProfile:    vcdMachineTemplate.Spec.Template.Spec.StorageProfile,
 			DiskSizeMb:        int32(vcdMachineTemplate.Spec.Template.Spec.DiskSize.Value() / (1024 * 1024)),
 			DesiredReplicas:   desiredReplicasCount,
-			AvailableReplicas: md.Status.ReadyReplicas,
+			AvailableReplicas: availableReplicas,
 			NodeStatus:        nodeStatusMap,
 		}
 		nodePoolList = append(nodePoolList, nodePool)
@@ -484,6 +488,10 @@
 		if kcp.Spec.Replicas != nil {
 			desiredReplicaCount = *kcp.Spec.Replicas
 		}
+		availableReplicas := int32(0)
+		if kcp.Status.ReadyReplicas != nil {
+			availableReplicas = *kcp.Status.ReadyReplicas
+		}
 		nodePool := rdeType.NodePool{
 			Name:              kcp.Name,
 			SizingPolicy:      vcdMachineTemplate.Spec.Template.Spec.SizingPolicy,
@@ -492,7 +500,7 @@
 			StorageProfile:    vcdMachineTemplate.Spec.Template.Spec.StorageProfile,
 			DiskSizeMb:        int32(vcdMachineTemplate.Spec.Template.Spec.DiskSize.Value() / (1024 * 1024)),
 			DesiredReplicas:   desiredReplicaCount,
-			AvailableReplicas: kcp.Status.ReadyReplicas,
+			AvailableReplicas: availableReplicas,
 			NodeStatus:        nodeStatusMap,
 		}
 		nodePoolList = append(nodePoolList, nodePool)
@@ -522,13 +530,25 @@
 
 	vcdMachineTemplateNameToObjRef := make(map[string]v1.ObjectReference)
 	for _, kcp := range kcpList.Items {
-		vcdMachineTemplateNameToObjRef[kcp.Spec.MachineTemplate.InfrastructureRef.Name] = kcp.Spec.MachineTemplate.InfrastructureRef
+		vcdMachineTemplateNameToObjRef[kcp.Spec.MachineTemplate.Spec.InfrastructureRef.Name] = v1.ObjectReference{
+			Kind:      kcp.Spec.MachineTemplate.Spec.InfrastructureRef.Kind,
+			Name:      kcp.Spec.MachineTemplate.Spec.InfrastructureRef.Name,
+			Namespace: kcp.Namespace,
+		}
 	}
 
 	kubeadmConfigTemplateNameToObjRef := make(map[string]*v1.ObjectReference)
 	for _, md := range mdList.Items {
-		vcdMachineTemplateNameToObjRef[md.Spec.Template.Spec.InfrastructureRef.Name] = md.Spec.Template.Spec.InfrastructureRef
-		kubeadmConfigTemplateNameToObjRef[md.Spec.Template.Spec.Bootstrap.ConfigRef.Name] = md.Spec.Template.Spec.Bootstrap.ConfigRef
+		vcdMachineTemplateNameToObjRef[md.Spec.Template.Spec.InfrastructureRef.Name] = v1.ObjectReference{
+			Kind:      md.Spec.Template.Spec.InfrastructureRef.Kind,
+			Name:      md.Spec.Template.Spec.InfrastructureRef.Name,
+			Namespace: md.Namespace,
+		}
+		kubeadmConfigTemplateNameToObjRef[md.Spec.Template.Spec.Bootstrap.ConfigRef.Name] = &v1.ObjectReference{
+			Kind:      md.Spec.Template.Spec.Bootstrap.ConfigRef.Kind,
+			Name:      md.Spec.Template.Spec.Bootstrap.ConfigRef.Name,
+			Namespace: md.Namespace,
+		}
 	}
 
 	vcdMachineTemplates := make([]*infrav1.VCDMachineTemplate, 0)
@@ -540,7 +560,7 @@
 		vcdMachineTemplates = append(vcdMachineTemplates, vcdMachineTemplate)
 	}
 
-	kubeadmConfigTemplates := make([]*v1beta1.KubeadmConfigTemplate, 0)
+	kubeadmConfigTemplates := make([]*kubeadmbootstrapv1.KubeadmConfigTemplate, 0)
 	for _, objRef := range kubeadmConfigTemplateNameToObjRef {
 		kubeadmConifgTemplate, err := getKubeadmConfigTemplateByObjRef(ctx, cli, *objRef)
 		if err != nil {
@@ -641,7 +661,7 @@
 			return false, fmt.Errorf("failed to fetch machines for the kubeadm control plane object [%s] for cluster [%s]: [%v]", kcp.Name, clusterName, err)
 		}
 		for _, machine := range machines {
-			if machine.Spec.Version != nil && *machine.Spec.Version != expectedVersion {
+			if machine.Spec.Version != "" && machine.Spec.Version != expectedVersion {
 				return false, nil
 			}
 		}
@@ -653,7 +673,7 @@
 			return false, fmt.Errorf("failed to fetch machines for the machine deployment [%s] for cluster [%s]: [%v]", md.Name, clusterName, err)
 		}
 		for _, machine := range machineList.Items {
-			if machine.Spec.Version != nil && *machine.Spec.Version != expectedVersion {
+			if machine.Spec.Version != "" && machine.Spec.Version != expectedVersion {
 				return false, nil
 			}
 		}
