diff --git a/business/mesh.go b/business/mesh.go
index ad57f5457..a85fa307c 100644
--- a/business/mesh.go
+++ b/business/mesh.go
@@ -620,25 +620,8 @@ func (in *MeshService) OutboundTrafficPolicy() (*models.OutboundPolicy, error) {
 }

 func (in *MeshService) IstiodResourceThresholds() (*models.IstiodThresholds, error) {
-	conf := config.Get()
-
-	var istioDeployment *v1.Deployment
-	var istioDeploymentConfig = conf.ExternalServices.Istio.IstiodDeploymentName
-	var err error
-
-	if IsNamespaceCached(conf.IstioNamespace) {
-		istioDeployment, err = kialiCache.GetDeployment(conf.IstioNamespace, istioDeploymentConfig)
-	} else {
-		istioDeployment, err = in.k8s.GetDeployment(conf.IstioNamespace, istioDeploymentConfig)
-	}
-	if err != nil && !errors.IsNotFound(err) {
-		return nil, err
-	}
-
-	memoryLimit := istioDeployment.Spec.Template.Spec.Containers[0].Resources.Limits.Memory().AsApproximateFloat64() / 1000000 // in Mb
-	cpuLimit := istioDeployment.Spec.Template.Spec.Containers[0].Resources.Limits.Cpu().AsApproximateFloat64()

-	thresholds := models.IstiodThresholds{Memory: memoryLimit, CPU: cpuLimit}
+	thresholds := models.IstiodThresholds{Memory: 0, CPU: 0}

 	return &thresholds, nil
 }
