diff --git a/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go b/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go
index efd689ba..90795894 100644
--- a/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go
+++ b/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go
@@ -447,6 +447,12 @@ func (r *ReconcileLinstorSatelliteSet) reconcileAllNodesOnController(ctx context
 			continue
 		}
 
+		if pod.Status.Phase != corev1.PodRunning {
+			logger.WithField("pod", pod.Name).Debug("Pod not running.")
+
+			continue
+		}
+
 		// Registration can be done in parallel, so we handle per-node work in a separate go-routine
 		wg.Add(1)
 
@@ -533,6 +539,14 @@ func (r *ReconcileLinstorSatelliteSet) reconcileSingleNodeRegistration(ctx conte
 		return fmt.Errorf("failed to reconcile satellite: %w", err)
 	}
 
+	if mdutil.SliceContains(lNode.Flags, linstor.FlagEvicted) {
+		// The pod exists, so there is no reason not to restore it.
+		err := linstorClient.Nodes.Restore(ctx, lNode.Name, lapi.NodeRestore{})
+		if err != nil {
+			return fmt.Errorf("node '%s' failed to restore: %w", lNode.Name, err)
+		}
+	}
+
 	if lNode.ConnectionStatus != lc.Online {
 		return &reconcileutil.TemporaryError{
 			Source:       fmt.Errorf("node '%s' registered, but not online (%s)", lNode.Name, lNode.ConnectionStatus),
