--- a/public/app/plugins/datasource/prometheus/datasource.ts	2018-12-10 17:08:32.000000000 +0300
+++ b/public/app/plugins/datasource/prometheus/datasource.ts	2018-12-10 19:03:15.791337923 +0300
@@ -235,6 +235,10 @@
         ...this.getRangeScopedVars(),
       });
     }
+    const intervalForRangeVectors = Math.max(kbn.interval_to_seconds(this.interval) * 2, interval);
+    scopedVars = Object.assign({}, scopedVars, {
+    __interval_rv: { text: intervalForRangeVectors + 's', value: intervalForRangeVectors + 's' },
+    });
     query.step = interval;

     let expr = target.expr;
