diff --git a/public/app/plugins/datasource/prometheus/datasource.ts b/public/app/plugins/datasource/prometheus/datasource.ts
index 77c6ab39f8..81919a5b5d 100644
--- a/public/app/plugins/datasource/prometheus/datasource.ts
+++ b/public/app/plugins/datasource/prometheus/datasource.ts
@@ -364,6 +364,24 @@ export class PrometheusDatasource extends DataSourceApi<PromQuery, PromOptions>
         ...this.getRangeScopedVars(options.range),
       });
     }
+    const intervalScrapeX3ForOverTimeVectors = Math.max(kbn.interval_to_seconds(this.interval) * 3, interval);
+    const intervalScrapeX4ForRangeVectors = Math.max(kbn.interval_to_seconds(this.interval) * 4, interval);
+    const intervalForRangeVectors = Math.max(kbn.interval_to_seconds(this.interval) * 2, interval);
+    scopedVars = Object.assign({}, scopedVars, {
+      __interval_rv: {
+        text: intervalForRangeVectors + 's',
+        value: intervalForRangeVectors + 's',
+      },
+      __interval_sx3: {
+        text: intervalScrapeX3ForOverTimeVectors + 's',
+        value: intervalScrapeX3ForOverTimeVectors + 's',
+      },
+      __interval_sx4: {
+        text: intervalScrapeX4ForRangeVectors + 's',
+        value: intervalScrapeX4ForRangeVectors + 's',
+      },
+    });
+
     query.step = interval;

     let expr = target.expr;
diff --git a/public/app/plugins/datasource/prometheus/promql.ts b/public/app/plugins/datasource/prometheus/promql.ts
index d4b68200ba..e82b62614e 100644
--- a/public/app/plugins/datasource/prometheus/promql.ts
+++ b/public/app/plugins/datasource/prometheus/promql.ts
@@ -2,6 +2,8 @@ import { CompletionItem } from '@grafana/ui';

 export const RATE_RANGES: CompletionItem[] = [
   { label: '$__interval', sortText: '$__interval' },
+  { label: '$__interval_sx3', sortText: '$__interval_sx3' },
+  { label: '$__interval_sx4', sortText: '$__interval_sx4' },
   { label: '1m', sortText: '00:01:00' },
   { label: '5m', sortText: '00:05:00' },
   { label: '10m', sortText: '00:10:00' },
