diff --git a/public/app/plugins/datasource/prometheus/datasource.ts b/public/app/plugins/datasource/prometheus/datasource.ts
index f090e947aa..66bdd6503c 100644
--- a/public/app/plugins/datasource/prometheus/datasource.ts
+++ b/public/app/plugins/datasource/prometheus/datasource.ts
@@ -446,6 +446,23 @@ export class PrometheusDatasource extends DataSourceApi<PromQuery, PromOptions>
         ...this.getRangeScopedVars(options.range),
       });
     }
+    const intervalScrapeX3ForOverTimeVectors = Math.max(rangeUtil.intervalToSeconds(this.interval) * 3, interval);
+    const intervalScrapeX4ForRangeVectors = Math.max(rangeUtil.intervalToSeconds(this.interval) * 4, interval);
+    const intervalForRangeVectors = Math.max(rangeUtil.intervalToSeconds(this.interval) * 2, interval);
+    scopedVars = Object.assign({}, scopedVars, {
+      __interval_rv: {
+        text: intervalForRangeVectors + 's',
+        value: intervalForRangeVectors + 's',
+      },
+      __interval_sx3: {
+        text: intervalScrapeX3ForOverTimeVectors + 's',
+        value: intervalScrapeX3ForOverTimeVectors + 's',
+      },
+      __interval_sx4: {
+        text: intervalScrapeX4ForRangeVectors + 's',
+        value: intervalScrapeX4ForRangeVectors + 's',
+      },
+    });
     query.step = interval;

     let expr = target.expr;
diff --git a/public/app/plugins/datasource/prometheus/promql.ts b/public/app/plugins/datasource/prometheus/promql.ts
index 2550c90b99..9925adbd3c 100644
--- a/public/app/plugins/datasource/prometheus/promql.ts
+++ b/public/app/plugins/datasource/prometheus/promql.ts
@@ -5,6 +5,8 @@ import { CompletionItem } from '@grafana/ui';
 // @see public/app/plugins/datasource/loki/language_provider.ts
 export const RATE_RANGES: CompletionItem[] = [
   { label: '$__interval', sortValue: '$__interval' },
+  { label: '$__interval_sx3', sortText: '$__interval_sx3' },
+  { label: '$__interval_sx4', sortText: '$__interval_sx4' },
   { label: '$__rate_interval', sortValue: '$__rate_interval' },
   { label: '$__range', sortValue: '$__range' },
   { label: '1m', sortValue: '00:01:00' },
