diff --git a/pkg/services/provisioning/dashboards/file_reader.go b/pkg/services/provisioning/dashboards/file_reader.go
index 698a3587b4..e9676286ed 100644
--- a/pkg/services/provisioning/dashboards/file_reader.go
+++ b/pkg/services/provisioning/dashboards/file_reader.go
@@ -98,6 +98,16 @@ func (fr *fileReader) startWalkingDisk() error {

 	// save dashboards based on json files
 	for path, fileInfo := range filesFoundOnDisk {
+		_, ok := fr.Cfg.Options["folderFromFileStructure"]
+		if ok {
+			fr.Cfg.Folder = filepath.Base(filepath.Dir(path))
+			fr.log.Debug("folder name", "warning", fr.Cfg.Folder)
+
+			folderId, err = getOrCreateFolderId(fr.Cfg, fr.dashboardProvisioningService)
+			if err != nil && err != ErrFolderNameMissing {
+				return err
+			}
+		}
 		provisioningMetadata, err := fr.saveDashboard(path, folderId, fileInfo, provisionedDashboardRefs)
 		sanityChecker.track(provisioningMetadata)
 		if err != nil {
