diff --git a/pkg/config/constants/constants.go b/pkg/config/constants/constants.go
--- a/pkg/config/constants/constants.go
+++ b/pkg/config/constants/constants.go
@@ -85,7 +85,7 @@ const (
 	IstioIngressLabelValue = "ingressgateway"
 
 	// IstioSystemNamespace is the namespace where Istio's components are deployed
-	IstioSystemNamespace = "istio-system"
+	IstioSystemNamespace = "d8-istio"
 
 	// DefaultAuthenticationPolicyName is the name of the cluster-scoped authentication policy. Only
 	// policy with this name in the cluster-scoped will be considered.
