diff --git a/istioctl/cmd/root.go b/istioctl/cmd/root.go
index 14628cb58e..ee84d9eeef 100644
--- a/istioctl/cmd/root.go
+++ b/istioctl/cmd/root.go
@@ -45,7 +45,6 @@ import (
 	"istio.io/istio/operator/cmd/mesh"
 	"istio.io/istio/pkg/cmd"
 	"istio.io/istio/pkg/collateral"
-	"istio.io/istio/pkg/config/constants"
 	"istio.io/istio/pkg/log"
 	"istio.io/istio/tools/bug-report/pkg/bugreport"
 )
@@ -88,7 +87,7 @@ func ConfigAndEnvProcessing() error {
 }
 
 func init() {
-	viper.SetDefault("istioNamespace", constants.IstioSystemNamespace)
+	viper.SetDefault("istioNamespace", "d8-istio")
 	viper.SetDefault("xds-port", 15012)
 }
 
