Subject: [PATCH] replace-gofsutil
---
Index: go.mod
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/go.mod b/go.mod
--- a/go.mod	(revision 96144dbc4af7e00a336fe5789de2f08d05bff73d)
+++ b/go.mod	(revision e0e6680ff5ed3fbd41e1f602cb5d6d5039d0334e)
@@ -229,4 +229,5 @@
 	k8s.io/sample-cli-plugin => k8s.io/sample-cli-plugin v0.27.10
 	k8s.io/sample-controller => k8s.io/sample-controller v0.27.10
 	sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.15.3
+	github.com/akutz/gofsutil => ../gofsutil
 )
