diff --git a/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go b/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go
index 8fa5257..c430ff8 100644
--- a/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go
+++ b/pkg/controller/linstorsatelliteset/linstorsatelliteset_controller.go
@@ -1261,12 +1261,17 @@ func daemonSetWithDRBDKernelModuleInjection(ds *apps.DaemonSet, satelliteSet *pi
 		MountPath: kubeSpec.ModulesDir,
 	},
 	)
-
+	volumeMounts = append(volumeMounts, corev1.VolumeMount{
+		Name:      kubeSpec.UsrLibDirName,
+		MountPath: kubeSpec.UsrLibDirMountPath,
+	},
+	)
 	ds.Spec.Template.Spec.InitContainers = []corev1.Container{
 		{
-			Name:            "kernel-module-injector",
+			Name:            "wait-drbd-ver-9-2",
 			Image:           satelliteSet.Spec.KernelModuleInjectionImage,
 			ImagePullPolicy: satelliteSet.Spec.ImagePullPolicy,
+			Command:         []string{"/bin/bash", "-c", "exec /entry.sh"},
 			SecurityContext: &corev1.SecurityContext{Privileged: &kubeSpec.Privileged,
 				SELinuxOptions: &corev1.SELinuxOptions{Level: kubeSpec.Level, Type: kubeSpec.Type}},
 			Env:          env,
@@ -1296,6 +1301,15 @@ func daemonSetWithDRBDKernelModuleInjection(ds *apps.DaemonSet, satelliteSet *pi
 				},
 			},
 		})
+		ds.Spec.Template.Spec.Volumes = append(ds.Spec.Template.Spec.Volumes, corev1.Volume{
+			Name: kubeSpec.UsrLibDirName,
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{
+					Path: kubeSpec.UsrLibDir,
+					Type: &kubeSpec.HostPathDirectoryType,
+				},
+			},
+		})
 	}

 	return ds
diff --git a/pkg/k8s/spec/const.go b/pkg/k8s/spec/const.go
index 1fb8051..9ad09fb 100644
--- a/pkg/k8s/spec/const.go
+++ b/pkg/k8s/spec/const.go
@@ -40,6 +40,9 @@ const (
 	LinstorCertDirName          = "linstor-certs"
 	ModulesDir                  = "/lib/modules/" // "/usr/lib/modules/"
 	ModulesDirName              = "modules-dir"
+	UsrLibDir                   = "/usr/lib"
+	UsrLibDirMountPath          = "/host/usr/lib"
+	UsrLibDirName               = "usr-lib-dir"
 	SrcDir                      = "/usr/src"
 	SrcDirName                  = "src-dir"
 	SysDir                      = "/sys/"
