diff --git a/pkg/resourceprovider/provider.go b/pkg/resourceprovider/provider.go
index de59e70..f6b2adf 100644
--- a/pkg/resourceprovider/provider.go
+++ b/pkg/resourceprovider/provider.go
@@ -180,6 +180,7 @@ func (p *resourceProvider) GetPodMetrics(pods ...*metav1.PartialObjectMetadata)
 // also recording the earliest time in resTime.  It will return without operating if
 // any data is missing.
 func (p *resourceProvider) assignForPod(pod *metav1.PartialObjectMetadata, resultsByNs map[string]nsQueryResults) *metrics.PodMetrics {
+	now := time.Now()
 	// check to make sure everything is present
 	nsRes, nsResPresent := resultsByNs[pod.Namespace]
 	if !nsResPresent {
@@ -188,12 +189,23 @@ func (p *resourceProvider) assignForPod(pod *metav1.PartialObjectMetadata, resul
 	}
 	cpuRes, hasResult := nsRes.cpu[pod.Name]
 	if !hasResult {
-		klog.Errorf("unable to fetch CPU metrics for pod %s, skipping", pod.String())
+		if pod.CreationTimestamp.Time.Before(now.Add(-p.window)) {
+			klog.V(4).Infof("Pod %s is too new (age < %v); "+
+				"cpu metrics not yet available, skipping as expected", pod.Name, p.window)
+		} else {
+			klog.Errorf("unable to fetch CPU metrics for pod %s, skipping", pod.String())
+		}
 		return nil
 	}
 	memRes, hasResult := nsRes.mem[pod.Name]
 	if !hasResult {
-		klog.Errorf("unable to fetch memory metrics for pod %s, skipping", pod.String())
+		if pod.CreationTimestamp.Time.Before(now.Add(-p.window)) {
+			klog.V(4).Infof("Pod %s is too new (age < %v); "+
+				"memory metrics not yet available, skipping as expected", pod.Name, p.window)
+		} else {
+			klog.Errorf("unable to fetch memory metrics for pod %s, skipping", pod.String())
+		}
+
 		return nil
 	}

