diff --git a/pkg/resourceprovider/provider.go b/pkg/resourceprovider/provider.go
index de59e70..af756aa 100644
--- a/pkg/resourceprovider/provider.go
+++ b/pkg/resourceprovider/provider.go
@@ -180,6 +180,7 @@ func (p *resourceProvider) GetPodMetrics(pods ...*metav1.PartialObjectMetadata)
 // also recording the earliest time in resTime.  It will return without operating if
 // any data is missing.
 func (p *resourceProvider) assignForPod(pod *metav1.PartialObjectMetadata, resultsByNs map[string]nsQueryResults) *metrics.PodMetrics {
+	now := time.Now()
 	// check to make sure everything is present
 	nsRes, nsResPresent := resultsByNs[pod.Namespace]
 	if !nsResPresent {
@@ -187,13 +188,24 @@ func (p *resourceProvider) assignForPod(pod *metav1.PartialObjectMetadata, resul
 		return nil
 	}
 	cpuRes, hasResult := nsRes.cpu[pod.Name]
-	if !hasResult {
-		klog.Errorf("unable to fetch CPU metrics for pod %s, skipping", pod.String())
+	if !hasResult && pod.CreationTimestamp.Time.After(now.Add(-p.window)) {
+		if pod.CreationTimestamp.Time.After(now.Add(-p.window)) {
+			klog.V(4).Infof("unable to fetch cpu metrics for pod %s (young pod, age < window), skipping",
+				pod.Name)
+		} else {
+			klog.Errorf("unable to fetch CPU metrics for pod %s, skipping", pod.String())
+		}
 		return nil
 	}
 	memRes, hasResult := nsRes.mem[pod.Name]
 	if !hasResult {
-		klog.Errorf("unable to fetch memory metrics for pod %s, skipping", pod.String())
+		if pod.CreationTimestamp.Time.After(now.Add(-p.window)) {
+			klog.V(4).Infof("unable to fetch memory metrics for pod %s (young pod, age < window), skipping",
+				pod.Name)
+		} else {
+			klog.Errorf("unable to fetch memory metrics for pod %s, skipping", pod.String())
+		}
+
 		return nil
 	}
