Index: aws/aws.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/legacy-cloud-providers/aws/aws.go b/legacy-cloud-providers/aws/aws.go
--- a/legacy-cloud-providers/aws/aws.go	(revision fae613ab2e7c402747eff35a5e87250e2c9a494e)
+++ b/legacy-cloud-providers/aws/aws.go	(date 1632466038094)
@@ -4241,6 +4241,13 @@
 	c.sortELBSecurityGroupList(lbSecurityGroupIDs, annotations)
 	loadBalancerSecurityGroupID := lbSecurityGroupIDs[0]

+	var deleteSecurityGroupID bool
+	// we shouldn't delete Ingress SG rule, if it allows access
+	// from configured "ElbSecurityGroup", so that we won't disrupt access to Nodes from other ELBs
+	if loadBalancerSecurityGroupID != c.cfg.Global.ElbSecurityGroup {
+		deleteSecurityGroupID = true
+	}
+
 	// Get the actual list of groups that allow ingress from the load-balancer
 	var actualGroups []*ec2.SecurityGroup
 	{
@@ -4336,7 +4343,7 @@
 			if !changed {
 				klog.Warning("Allowing ingress was not needed; concurrent change? groupId=", instanceSecurityGroupID)
 			}
-		} else {
+		} else if deleteSecurityGroupID {
 			changed, err := c.removeSecurityGroupIngress(instanceSecurityGroupID, permissions)
 			if err != nil {
 				return err
