Subject: [PATCH] 001
---
Index: pkg/cloudprovider/yandex/routes.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/cloudprovider/yandex/routes.go b/pkg/cloudprovider/yandex/routes.go
--- a/pkg/cloudprovider/yandex/routes.go	(revision c35e724eac6114ef570c1c5a2d45d4e8f38703f8)
+++ b/pkg/cloudprovider/yandex/routes.go	(date 1679489832746)
@@ -2,9 +2,9 @@

 import (
 	"context"
-	"fmt"
 	"sync"

+	"github.com/davecgh/go-spew/spew"
 	"github.com/pkg/errors"
 	"github.com/yandex-cloud/go-genproto/yandex/cloud/operation"
 	"github.com/yandex-cloud/go-genproto/yandex/cloud/vpc/v1"
@@ -77,11 +77,22 @@
 	}

 	kubeNodeName := string(route.TargetNode)
-	nextHop, err := yc.getInternalIpByNodeName(kubeNodeName)
+	addresses, err := yc.NodeAddresses(ctx, route.TargetNode)
 	if err != nil {
 		return err
 	}

+	var nextHop string
+	for _, address := range addresses {
+		if address.Type == v1.NodeInternalIP {
+			nextHop = address.Address
+			break
+		}
+	}
+	if len(nextHop) == 0 {
+		return spew.Errorf("could not determine InternalIP from list of IPs for VM %q: %v", route.TargetNode, addresses)
+	}
+
 	newStaticRoutes := filterStaticRoutes(rt.StaticRoutes, routeFilterTerm{
 		termType:        routeFilterAddOrUpdate,
 		nodeName:        kubeNodeName,
@@ -133,25 +144,6 @@
 	return err
 }

-func (yc *Cloud) getInternalIpByNodeName(nodeName string) (string, error) {
-	kubeNode, err := yc.nodeLister.Get(nodeName)
-	if err != nil {
-		return "", err
-	}
-
-	var targetInternalIP string
-	for _, address := range kubeNode.Status.Addresses {
-		if address.Type == v1.NodeInternalIP {
-			targetInternalIP = address.Address
-		}
-	}
-	if len(targetInternalIP) == 0 {
-		return "", fmt.Errorf("no InternalIPs found for Node %q", nodeName)
-	}
-
-	return targetInternalIP, nil
-}
-
 type routeFilterTerm struct {
 	termType        routeFilterTermType
 	nodeName        string
Index: go.mod
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/go.mod b/go.mod
--- a/go.mod	(revision c35e724eac6114ef570c1c5a2d45d4e8f38703f8)
+++ b/go.mod	(date 1679489836846)
@@ -3,6 +3,7 @@
 go 1.18

 require (
+	github.com/davecgh/go-spew v1.1.1
 	github.com/deckarep/golang-set v1.7.1
 	github.com/golang/protobuf v1.5.2
 	github.com/pkg/errors v0.9.1
@@ -29,7 +30,6 @@
 	github.com/cespare/xxhash/v2 v2.1.2 // indirect
 	github.com/coreos/go-semver v0.3.0 // indirect
 	github.com/coreos/go-systemd/v22 v22.3.2 // indirect
-	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/dgrijalva/jwt-go v3.2.0+incompatible // indirect
 	github.com/emicklei/go-restful/v3 v3.8.0 // indirect
 	github.com/evanphx/json-patch v4.12.0+incompatible // indirect
