Subject: [PATCH] add iops calculation
---
Index: pkg/vcdcsiclient/disks.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/vcdcsiclient/disks.go b/pkg/vcdcsiclient/disks.go
--- a/pkg/vcdcsiclient/disks.go	(revision 89ff1aeab1be77cdf6536f1e1002bdc58dde63f2)
+++ b/pkg/vcdcsiclient/disks.go	(date 1701152867311)
@@ -9,17 +9,19 @@
 	"context"
 	"encoding/json"
 	"fmt"
-	"github.com/vmware/cloud-director-named-disk-csi-driver/pkg/util"
-	"github.com/vmware/cloud-director-named-disk-csi-driver/pkg/vcdtypes"
-	"github.com/vmware/cloud-director-named-disk-csi-driver/version"
+	"net/http"
+	"strings"
+	"time"
+
 	"github.com/vmware/cloud-provider-for-cloud-director/pkg/vcdsdk"
 	swaggerClient "github.com/vmware/cloud-provider-for-cloud-director/pkg/vcdswaggerclient_36_0"
 	"github.com/vmware/go-vcloud-director/v2/govcd"
 	"github.com/vmware/go-vcloud-director/v2/types/v56"
 	"k8s.io/klog"
-	"net/http"
-	"strings"
-	"time"
+
+	"github.com/vmware/cloud-director-named-disk-csi-driver/pkg/util"
+	"github.com/vmware/cloud-director-named-disk-csi-driver/pkg/vcdtypes"
+	"github.com/vmware/cloud-director-named-disk-csi-driver/version"
 )

 type DiskManager struct {
@@ -161,6 +163,23 @@
 		diskParams.Disk.StorageProfile = &types.Reference{
 			HREF: storageReference.HREF,
 		}
+
+		storageProfileData, err := diskManager.VCDClient.VCDClient.GetStorageProfileByHref(storageReference.HREF)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get storage profile [%s] for disk [%s] by HREF [%s]",
+				storageProfile, diskName, storageReference.HREF)
+		}
+
+		if storageProfileData.IopsSettings != nil {
+			if storageProfileData.IopsSettings.Enabled {
+				// If storageProfile Max IOPS per Gb is set
+				iops := storageProfileData.IopsSettings.DiskIopsPerGbMax * d.SizeMb / 1024
+				if iops == 0 {
+					iops = storageProfileData.IopsSettings.DiskIopsMax
+				}
+				d.Iops = iops
+			}
+		}
 	}

 	task, err := diskManager.createDisk(diskParams)
Index: go.mod
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/go.mod b/go.mod
--- a/go.mod	(revision 89ff1aeab1be77cdf6536f1e1002bdc58dde63f2)
+++ b/go.mod	(date 1701152867314)
@@ -13,7 +13,7 @@
 	github.com/stretchr/testify v1.7.0
 	github.com/thecodeteam/gofsutil v0.1.2 // indirect
 	github.com/vmware/cloud-provider-for-cloud-director v0.0.0-20230928223648-8134f65b3f61
-	github.com/vmware/go-vcloud-director/v2 v2.14.0-rc.3
+	github.com/vmware/go-vcloud-director/v2 v2.21.0
 	golang.org/x/oauth2 v0.4.0 // indirect
 	golang.org/x/sys v0.5.0
 	google.golang.org/grpc v1.53.0
