Index: pkg/cloudprovider/vsphere/nodemanager.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/cloudprovider/vsphere/nodemanager.go b/pkg/cloudprovider/vsphere/nodemanager.go
--- a/pkg/cloudprovider/vsphere/nodemanager.go	(revision 55b58029493df3ae81d9d085e28dcb7a160ef702)
+++ b/pkg/cloudprovider/vsphere/nodemanager.go	(revision 69cb68d0e885ee9573f68efddaf040caf3534552)
@@ -291,6 +291,17 @@
 		klog.Warningf("Unable to find a suitable IP address. ipFamily: %s", ipFamily)
 	}
 
+	var internalAddressExists bool
+	for _, addr := range addrs {
+		if addr.Type == v1.NodeInternalIP {
+			internalAddressExists = true
+			break
+		}
+	}
+	if !internalAddressExists {
+		return fmt.Errorf("not a single internalAddressFound in Node addresses:\n%+v", addrs)
+	}
+
 	klog.V(2).Infof("Found node %s as vm=%+v in vc=%s and datacenter=%s",
 		nodeID, vmDI.VM, vmDI.VcServer, vmDI.DataCenter.Name())
 	klog.V(2).Info("Hostname: ", oVM.Guest.HostName, " UUID: ", oVM.Summary.Config.Uuid)
