diff --git a/pkg/cloudprovider/vsphere/instances.go b/pkg/cloudprovider/vsphere/instances.go
index ed8160d..577bdf7 100644
--- a/pkg/cloudprovider/vsphere/instances.go
+++ b/pkg/cloudprovider/vsphere/instances.go
@@ -153,6 +153,10 @@ func (i *instances) CurrentNodeName(ctx context.Context, hostname string) (types
 func (i *instances) InstanceExistsByProviderID(ctx context.Context, providerID string) (bool, error) {
 	klog.V(4).Info("instances.InstanceExistsByProviderID() called with ", providerID)
 
+	if providerID == "static://" {
+		return true, nil
+	}
+
 	// Check if node has been discovered already
 	uid := GetUUIDFromProviderID(providerID)
 	err := i.nodeManager.DiscoverNode(uid, cm.FindVMByUUID)
