diff --git a/internal/ingress/controller/nginx.go b/internal/ingress/controller/nginx.go
index 5575009ea..567881cf6 100644
--- a/internal/ingress/controller/nginx.go
+++ b/internal/ingress/controller/nginx.go
@@ -418,10 +418,12 @@ func (n *NGINXController) Stop() error {
 func (n *NGINXController) start(cmd *exec.Cmd) {
 	cmd.Stdout = os.Stdout
 	cmd.Stderr = os.Stderr
-	if err := cmd.Start(); err != nil {
+	if out, err := cmd.CombinedOutput(); err != nil {
 		klog.Fatalf("NGINX error: %v", err)
 		n.ngxErrCh <- err
 		return
+	} else {
+		klog.Infof("%v\n", string(out))
 	}
 
 	go func() {
