diff --git a/internal/ingress/controller/nginx.go b/internal/ingress/controller/nginx.go
index 5575009ea..5ad6ca0e6 100644
--- a/internal/ingress/controller/nginx.go
+++ b/internal/ingress/controller/nginx.go
@@ -416,14 +416,44 @@ func (n *NGINXController) Stop() error {
 }
 
 func (n *NGINXController) start(cmd *exec.Cmd) {
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		klog.Fatalf("NGINX error: %v", err)
+		n.ngxErrCh <- err
+		return
+	}
+	defer stderr.Close()
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		klog.Fatalf("NGINX error: %v", err)
+		n.ngxErrCh <- err
+		return
+	}
+	defer stdout.Close()
+
 	if err := cmd.Start(); err != nil {
 		klog.Fatalf("NGINX error: %v", err)
 		n.ngxErrCh <- err
 		return
 	}
 
+	errors, err := io.ReadAll(stderr)
+	if err != nil {
+		klog.Fatalf("NGINX error: %v", err)
+		n.ngxErrCh <- err
+		return
+	}
+
+	messages, err := io.ReadAll(stdout)
+	if err != nil {
+		klog.Fatalf("NGINX error: %v", err)
+		n.ngxErrCh <- err
+		return
+	}
+
+	klog.Infof("%v\n%v\n", messages, errors)
+
 	go func() {
 		n.ngxErrCh <- cmd.Wait()
 	}()
