diff --git a/internal/ingress/controller/nginx.go b/internal/ingress/controller/nginx.go
index 5575009ea..47a0acfc1 100644
--- a/internal/ingress/controller/nginx.go
+++ b/internal/ingress/controller/nginx.go
@@ -416,12 +416,22 @@ func (n *NGINXController) Stop() error {
 }
 
 func (n *NGINXController) start(cmd *exec.Cmd) {
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
-	if err := cmd.Start(); err != nil {
-		klog.Fatalf("NGINX error: %v", err)
-		n.ngxErrCh <- err
-		return
+	if n.cfg.IsChroot {
+		if out, err := cmd.CombinedOutput(); err != nil {
+			klog.Fatalf("NGINX error: %v", err)
+			n.ngxErrCh <- err
+			return
+		} else {
+			klog.Infof("%v\n", string(out))
+		}
+	} else {
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		if err := cmd.Start(); err != nil {
+			klog.Fatalf("NGINX error: %v", err)
+			n.ngxErrCh <- err
+			return
+		}
 	}
 
 	go func() {
