Subject: [PATCH] Fix getting vcd_catalog_vapp_template data source by org and catalog_id
---
Index: vcd/resource_vcd_catalog_vapp_template.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/vcd/resource_vcd_catalog_vapp_template.go b/vcd/resource_vcd_catalog_vapp_template.go
--- a/vcd/resource_vcd_catalog_vapp_template.go	(revision 2df7e589a92135093f3327e9d8698404bdf65bd0)
+++ b/vcd/resource_vcd_catalog_vapp_template.go	(date 1727329211575)
@@ -382,6 +382,32 @@
 	// No filter: we continue with single item  GET
 
 	if isSearchedByCatalog {
+		orgName, ok := d.GetOk("org")
+		if ok {
+			org, err := vcdClient.GetOrg(orgName.(string))
+			if err != nil {
+				return nil, fmt.Errorf("error retrieving org: %s", err)
+			}
+
+			vAppTemplates, err := catalog.QueryVappTemplateList()
+			if err != nil {
+				return nil, fmt.Errorf("error retrieving vApp Templates from catalog: %s", err)
+			}
+
+			for _, vAppTemplate := range vAppTemplates {
+				if vAppTemplate.Name == identifier && vAppTemplate.Org == org.Org.HREF {
+					vAppTemplateHrefPrefix := vcdClient.Client.VCDHREF
+					vAppTemplateHrefPrefix.Path += "/vAppTemplate/vappTemplate-"
+
+					id, found := strings.CutPrefix(vAppTemplate.HREF, vAppTemplateHrefPrefix.String())
+					if found {
+						identifier = id
+						break
+					}
+				}
+			}
+		}
+
 		// In a resource, this is the only possibility
 		vAppTemplate, err = catalog.GetVAppTemplateByNameOrId(identifier, false)
 	} else {
