Subject: [PATCH] Fix getting vcd_catalog_vapp_template data source by org and catalog_id
---
Index: vcd/resource_vcd_catalog_vapp_template.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/vcd/resource_vcd_catalog_vapp_template.go b/vcd/resource_vcd_catalog_vapp_template.go
--- a/vcd/resource_vcd_catalog_vapp_template.go	(revision 2df7e589a92135093f3327e9d8698404bdf65bd0)
+++ b/vcd/resource_vcd_catalog_vapp_template.go	(date 1730884397001)
@@ -382,6 +382,43 @@
 	// No filter: we continue with single item  GET
 
 	if isSearchedByCatalog {
+		orgName, ok := d.GetOk("org")
+		if ok {
+			var href string
+
+			org, err := vcdClient.GetOrgByName(orgName.(string))
+			if err == nil {
+				href = org.Org.HREF
+			} else {
+				util.Logger.Printf("[TRACE] error retrieving org: %s", err)
+
+				org, err := vcdClient.GetAdminOrgByName(orgName.(string))
+				if err != nil {
+					return nil, fmt.Errorf("error retrieving Admin Org: %s", err)
+				}
+
+				href = org.AdminOrg.HREF
+			}
+
+			vAppTemplates, err := catalog.QueryVappTemplateList()
+			if err != nil {
+				return nil, fmt.Errorf("error retrieving vApp Templates from catalog: %s", err)
+			}
+
+			for _, vAppTemplate := range vAppTemplates {
+				if vAppTemplate.Name == identifier && vAppTemplate.Org == href {
+					vAppTemplateHrefPrefix := vcdClient.Client.VCDHREF
+					vAppTemplateHrefPrefix.Path += "/vAppTemplate/vappTemplate-"
+
+					id, found := strings.CutPrefix(vAppTemplate.HREF, vAppTemplateHrefPrefix.String())
+					if found {
+						identifier = id
+						break
+					}
+				}
+			}
+		}
+
 		// In a resource, this is the only possibility
 		vAppTemplate, err = catalog.GetVAppTemplateByNameOrId(identifier, false)
 	} else {
