diff --git a/pkg/openstack/instances.go b/pkg/openstack/instances.go
index f9b61773..285cad55 100644
--- a/pkg/openstack/instances.go
+++ b/pkg/openstack/instances.go
@@ -19,6 +19,7 @@ package openstack
 import (
 	"context"
 	"fmt"
+	stderrors "errors"
 	sysos "os"
 	"regexp"
 	"strings"
@@ -85,6 +86,9 @@ func (os *OpenStack) InstancesV2() (cloudprovider.InstancesV2, bool) {
 
 // InstanceExists indicates whether a given node exists according to the cloud provider
 func (i *InstancesV2) InstanceExists(ctx context.Context, node *v1.Node) (bool, error) {
+	if strings.HasPrefix(node.Spec.ProviderID, "static://") {
+		return true, nil
+	}
 	_, err := i.getInstance(ctx, node)
 	if err == cloudprovider.InstanceNotFound {
 		klog.V(6).Infof("instance not found for node: %s", node.Name)
@@ -217,11 +221,16 @@ func getServerByName(ctx context.Context, client *gophercloud.ServiceClient, nam
 // If Instances.InstanceID or cloudprovider.GetInstanceProviderID is changed, the regexp should be changed too.
 var providerIDRegexp = regexp.MustCompile(`^` + ProviderName + `://([^/]*)/([^/]+)$`)
 
+var errStaticNode = stderrors.New("static Node detected")
+
 // instanceIDFromProviderID splits a provider's id and return instanceID.
 // A providerID is build out of '${ProviderName}:///${instance-id}' which contains ':///'.
 // or '${ProviderName}://${region}/${instance-id}' which contains '://'.
 // See cloudprovider.GetInstanceProviderID and Instances.InstanceID.
 func instanceIDFromProviderID(providerID string) (instanceID string, region string, err error) {
+	if strings.HasPrefix(providerID, "static://") {
+		return "", "", errStaticNode
+	}
 	// https://github.com/kubernetes/kubernetes/issues/85731
 	if providerID != "" && !strings.Contains(providerID, "://") {
 		providerID = ProviderName + "://" + providerID
