Index: pkg/node/manager/manager.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/node/manager/manager.go b/pkg/node/manager/manager.go
--- a/pkg/node/manager/manager.go	(revision 17b0dbd4a8c4b55ffdc4a65340ad939a2d63e1d3)
+++ b/pkg/node/manager/manager.go	(date 1646911913798)
@@ -412,7 +412,7 @@
 		}

 		if option.Config.IptablesMasqueradingEnabled() &&
-			address.Type != addressing.NodeInternalIP {
+			address.Type == addressing.NodeInternalIP {
 			iptables.AddToNodeIpset(address.IP)
 		}

