diff --git a/pkg/build/cmd.go b/pkg/build/cmd.go
index 5817cee065..a7a19cc30e 100644
--- a/pkg/build/cmd.go
+++ b/pkg/build/cmd.go
@@ -243,7 +243,7 @@ func ldflags(opts BuildOpts) (string, error) {
 	b.WriteString(fmt.Sprintf(" -X main.version=%s", opts.version))
 	b.WriteString(fmt.Sprintf(" -X main.commit=%s", getGitSha()))
 	b.WriteString(fmt.Sprintf(" -X main.buildstamp=%d", buildStamp))
-	b.WriteString(fmt.Sprintf(" -X main.buildBranch=%s", getGitBranch()))
+	b.WriteString(fmt.Sprintf(" -X main.buildBranch=%s", "fix_heatmap,feat_extra_vars"))
 	if v := os.Getenv("LDFLAGS"); v != "" {
 		b.WriteString(fmt.Sprintf(" -extldflags \"%s\"", v))
 	}
