Index: pkg/cloudprovider/providers/openstack/openstack_instances.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/cloudprovider/providers/openstack/openstack_instances.go b/pkg/cloudprovider/providers/openstack/openstack_instances.go
--- a/pkg/cloudprovider/providers/openstack/openstack_instances.go	(revision 445f068d7e01adf6a497a1d98285d1e2fdd8741b)
+++ b/pkg/cloudprovider/providers/openstack/openstack_instances.go	(revision 6b32af75ab05fb90f79241a482d53f0486c32ea0)
@@ -26,6 +26,7 @@
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
 	"k8s.io/klog/v2"
 
+	stderrors "errors"
 	"k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
 	cloudprovider "k8s.io/cloud-provider"
@@ -139,7 +140,9 @@
 // If false is returned with no error, the instance will be immediately deleted by the cloud controller manager.
 func (i *Instances) InstanceExistsByProviderID(ctx context.Context, providerID string) (bool, error) {
 	instanceID, err := instanceIDFromProviderID(providerID)
-	if err != nil {
+	if stderrors.Is(err, errStaticNode) {
+		return true, nil
+	} else if err != nil {
 		return false, err
 	}
 
@@ -286,10 +289,15 @@
 // If Instances.InstanceID or cloudprovider.GetInstanceProviderID is changed, the regexp should be changed too.
 var providerIDRegexp = regexp.MustCompile(`^` + ProviderName + `:///([^/]+)$`)
 
+var errStaticNode = stderrors.New("static Node detected")
+
 // instanceIDFromProviderID splits a provider's id and return instanceID.
 // A providerID is build out of '${ProviderName}:///${instance-id}'which contains ':///'.
 // See cloudprovider.GetInstanceProviderID and Instances.InstanceID.
 func instanceIDFromProviderID(providerID string) (instanceID string, err error) {
+	if providerID == "static://" {
+		return "", errStaticNode
+	}
 
 	// https://github.com/kubernetes/kubernetes/issues/85731
 	if providerID != "" && !strings.Contains(providerID, "://") {
