diff --git a/internal/collector/persistentvolume.go b/internal/collector/persistentvolume.go
index 817631e..dcbd58e 100644
--- a/internal/collector/persistentvolume.go
+++ b/internal/collector/persistentvolume.go
@@ -112,6 +112,22 @@ var (
 				}
 			}),
 		},
+		{
+			Name: "kube_persistentvolume_is_local",
+			Type: metric.Gauge,
+			Help: "Whether persistentvolume is local or not.",
+			GenerateFunc: wrapPersistentVolumeFunc(func(p *v1.PersistentVolume) *metric.Family {
+				return &metric.Family{
+					Metrics: []*metric.Metric{
+						{
+							LabelKeys:   []string{},
+							LabelValues: []string{},
+							Value:       boolFloat64(p.Spec.Local != nil),
+						},
+					},
+				}
+			}),
+		},
 		{
 			Name: "kube_persistentvolume_capacity_bytes",
 			Type: metric.Gauge,
