diff --git a/internal/collector/pod.go b/internal/collector/pod.go
index 777b2f3..7f1b1c8 100644
--- a/internal/collector/pod.go
+++ b/internal/collector/pod.go
@@ -747,6 +747,24 @@ var (
 					}
 				}
 
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+		{
+			Name: "kube_pod_spec_host_network",
+			Type: metric.Gauge,
+			Help: "Describes whether a Pod runs in the hostNetwork mode.",
+			GenerateFunc: wrapPodFunc(func(p *v1.Pod) *metric.Family {
+				ms := []*metric.Metric{}
+
+				ms = append(ms, &metric.Metric{
+					LabelKeys:   []string{},
+					LabelValues: []string{},
+					Value:       boolFloat64(p.Spec.HostNetwork),
+				})
+
 				return &metric.Family{
 					Metrics: ms,
 				}
