diff --git a/pkg/prometheus/promcfg.go b/pkg/prometheus/promcfg.go
index 4c2df3982..9b5693038 100644
--- a/pkg/prometheus/promcfg.go
+++ b/pkg/prometheus/promcfg.go
@@ -1509,6 +1509,10 @@ func (cg *ConfigGenerator) generateServiceMonitorConfig(
 
 	relabelings := initRelabelings()
 
+	if cg.endpointRoleFlavor() == kubernetesSDRoleEndpointSlice {
+		relabelings = generateFallbackEndpointSliceRelabeling(relabelings)
+	}
+
 	// Filter targets by services selected by the monitor.
 
 	// Exact label matches.
@@ -4327,6 +4331,39 @@ func (cg *ConfigGenerator) generateOTLPConfig() (yaml.MapItem, error) {
 	}, nil
 }
 
+func generateFallbackEndpointSliceRelabeling(relabelings []yaml.MapSlice) []yaml.MapSlice {
+	relabelings = append(relabelings, yaml.MapSlice{
+		{Key: "action", Value: "labelmap"},
+		{Key: "regex", Value: "^__meta_kubernetes_endpointslice_(label_.+|labelpresent_.+|annotation_.+|annotationpresent_.+)$"},
+		{Key: "replacement", Value: "__meta_kubernetes_endpoints_${1}"},
+	})
+	relabelings = append(relabelings, yaml.MapSlice{
+		{Key: "action", Value: "labelmap"},
+		{Key: "regex", Value: "^__meta_kubernetes_endpointslice_(port_name|port_protocol|address_target_kind|address_target_name)$"},
+		{Key: "replacement", Value: "__meta_kubernetes_endpoint_${1}"},
+	})
+	relabelings = append(relabelings, yaml.MapSlice{
+		{Key: "source_labels", Value: []string{"__meta_kubernetes_endpointslice_name"}},
+		{Key: "target_label", Value: "__meta_kubernetes_endpoints_name"},
+		{Key: "regex", Value: "(.+)"},
+		{Key: "replacement", Value: "${1}"},
+	})
+	relabelings = append(relabelings, yaml.MapSlice{
+		{Key: "source_labels", Value: []string{"__meta_kubernetes_endpointslice_endpoint_topology_kubernetes_io_hostname"}},
+		{Key: "target_label", Value: "__meta_kubernetes_endpoint_node_name"},
+		{Key: "regex", Value: "(.+)"},
+		{Key: "replacement", Value: "${1}"},
+	})
+	relabelings = append(relabelings, yaml.MapSlice{
+		{Key: "source_labels", Value: []string{"__meta_kubernetes_endpointslice_endpoint_conditions_ready"}},
+		{Key: "target_label", Value: "__meta_kubernetes_endpoint_ready"},
+		{Key: "regex", Value: "(.+)"},
+		{Key: "replacement", Value: "${1}"},
+	})
+
+	return relabelings
+}
+
 func (cg *ConfigGenerator) generateTracingConfig(store *assets.StoreBuilder) (yaml.MapItem, error) {
 	cfg := yaml.MapSlice{}
 	objMeta := cg.prom.GetObjectMeta()
