diff --git a/pkg/alertmanager/amcfg.go b/pkg/alertmanager/amcfg.go
index 6556569..7ada688 100644
--- a/pkg/alertmanager/amcfg.go
+++ b/pkg/alertmanager/amcfg.go
@@ -1077,6 +1077,9 @@ func (cb *configBuilder) convertEmailConfig(ctx context.Context, in monitoringv1
 	}
 
 	if in.TLSConfig != nil {
+		if err := cb.store.AddSafeTLSConfig(ctx, crKey.Namespace, in.TLSConfig); err != nil {
+			return nil, err
+		}
 		out.TLSConfig = cb.convertTLSConfig(in.TLSConfig, crKey)
 	}
 
@@ -1548,5 +1551,8 @@ func (cb *configBuilder) convertHTTPConfig(ctx context.Context, in *monitoringv1
 	}
 
 	if in.TLSConfig != nil {
+		if err := cb.store.AddSafeTLSConfig(ctx, crKey.Namespace, in.TLSConfig); err != nil {
+			return nil, err
+		}
 		out.TLSConfig = cb.convertTLSConfig(in.TLSConfig, crKey)
 	}

