diff --git a/pkg/alertmanager/amcfg.go b/pkg/alertmanager/amcfg.go
index 959bf63..df09fdc 100644
--- a/pkg/alertmanager/amcfg.go
+++ b/pkg/alertmanager/amcfg.go
@@ -1096,6 +1096,9 @@ func (cb *configBuilder) convertEmailConfig(ctx context.Context, in monitoringv1
 	}
 
 	if in.TLSConfig != nil {
+		if err := cb.store.AddSafeTLSConfig(ctx, crKey.Namespace, in.TLSConfig); err != nil {
+			return nil, err
+		}
 		out.TLSConfig = cb.convertTLSConfig(in.TLSConfig, crKey)
 	}
 
@@ -1571,5 +1574,8 @@ func (cb *configBuilder) convertHTTPConfig(ctx context.Context, in *monitoringv1
 	}
 
 	if in.TLSConfig != nil {
+		if err := cb.store.AddSafeTLSConfig(ctx, crKey.Namespace, in.TLSConfig); err != nil {
+			return nil, err
+		}
 		out.TLSConfig = cb.convertTLSConfig(in.TLSConfig, crKey)
 	}
