---
Index: pkg/prometheus/server/operator.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/prometheus/server/operator.go b/pkg/prometheus/server/operator.go
--- a/pkg/prometheus/server/operator.go	(revision 52526e3b0cfb5f3deaa967998ff88a5133f1ec0f)
+++ b/pkg/prometheus/server/operator.go	(date 1695058674835)
@@ -357,7 +357,7 @@
 	// The operator doesn't yet support the endpointslices API.
 	// See https://github.com/prometheus-operator/prometheus-operator/issues/3862
 	// for details.
-	c.endpointSliceSupported = false
+	c.endpointSliceSupported = endpointSliceSupported

 	c.statusReporter = prompkg.StatusReporter{
 		Kclient:         c.kclient,
