diff --git a/pkg/prometheus/statefulset.go b/pkg/prometheus/statefulset.go
index 40701651d..e0edb25ce 100644
--- a/pkg/prometheus/statefulset.go
+++ b/pkg/prometheus/statefulset.go
@@ -390,6 +390,12 @@ func makeStatefulSetSpec(p monitoringv1.Prometheus, c *operator.Config, shard in
 		)
 	}

+	if version.GTE(semver.MustParse("2.30.0")) {
+		// Align scrape timestamps to reduce storage and ram consumption
+		// https://github.com/prometheus/prometheus/pull/9283
+		promArgs = append(promArgs, "-scrape.timestamp-tolerance=10ms")
+	}
+
 	if version.Minor >= 4 {
 		if p.Spec.Rules.Alert.ForOutageTolerance != "" {
 			promArgs = append(promArgs, "-rules.alert.for-outage-tolerance="+p.Spec.Rules.Alert.ForOutageTolerance)
