Index: pkg/providers/v1/aws.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/providers/v1/aws.go b/pkg/providers/v1/aws.go
--- a/pkg/providers/v1/aws.go	(revision d31100461a83378b658b1ab2c689d79987a0d863)
+++ b/pkg/providers/v1/aws.go	(date 1632470363043)
@@ -4451,6 +4451,13 @@
 	c.sortELBSecurityGroupList(lbSecurityGroupIDs, annotations)
 	loadBalancerSecurityGroupID := lbSecurityGroupIDs[0]

+	var deleteSecurityGroupID bool
+	// we shouldn't delete Ingress SG rule, if it allows access
+	// from configured "ElbSecurityGroup", so that we won't disrupt access to Nodes from other ELBs
+	if loadBalancerSecurityGroupID != c.cfg.Global.ElbSecurityGroup {
+		deleteSecurityGroupID = true
+	}
+
 	// Get the actual list of groups that allow ingress from the load-balancer
 	var actualGroups []*ec2.SecurityGroup
 	{
@@ -4546,7 +4553,7 @@
 			if !changed {
 				klog.Warning("Allowing ingress was not needed; concurrent change? groupId=", instanceSecurityGroupID)
 			}
-		} else {
+		} else if deleteSecurityGroupID {
 			changed, err := c.removeSecurityGroupIngress(instanceSecurityGroupID, permissions)
 			if err != nil {
 				return err
