diff --git a/internal/ingress/annotations/class/main.go b/internal/ingress/annotations/class/main.go
index 8c0684af3..36f8baaee 100644
--- a/internal/ingress/annotations/class/main.go
+++ b/internal/ingress/annotations/class/main.go
@@ -62,6 +62,11 @@ func IsValid(ing *networking.Ingress) bool {
 		return IngressClass == DefaultClass
 	}
 
+	// without annotation and IngressClass in spec, fallback to default
+	if ing.Spec.IngressClassName == nil {
+		return IngressClass == DefaultClass
+	}
+
 	// 4. with IngressClass
 	return k8s.IngressClass.Name == *ing.Spec.IngressClassName
 }
