diff --git a/vendor/github.com/containerd/cri/pkg/containerd/opts/spec.go b/vendor/github.com/containerd/cri/pkg/containerd/opts/spec.go
index 1afb4fc1a..7834597ef 100644
--- a/vendor/github.com/containerd/cri/pkg/containerd/opts/spec.go
+++ b/vendor/github.com/containerd/cri/pkg/containerd/opts/spec.go
@@ -64,7 +64,9 @@ func WithProcessArgs(config *runtime.ContainerConfig, image *imagespec.ImageConf
 				args = append([]string{}, image.Cmd...)
 			}
 			if command == nil {
-				command = append([]string{}, image.Entrypoint...)
+				if len(image.Entrypoint) > 0 && image.Entrypoint[0] != "" {
+					command = append([]string{}, image.Entrypoint...)
+				}
 			}
 		}
 		if len(command) == 0 && len(args) == 0 {
