diff --git a/cluster-autoscaler/core/scale_up.go b/cluster-autoscaler/core/scale_up.go
index 5ed3cb3e9..610bd46f7 100644
--- a/cluster-autoscaler/core/scale_up.go
+++ b/cluster-autoscaler/core/scale_up.go
@@ -291,6 +291,17 @@ func computeExpansionOption(context *context.AutoscalingContext, podEquivalenceG
 
 	for _, eg := range podEquivalenceGroups {
 		samplePod := eg.pods[0]
+		samplePod.Annotations["fooooor"] = "baaar"
+		fmt.Println("CHECKING NODE", nodeInfo.Node().Name)
+		fmt.Println("SAMPLE POD", samplePod.Name)
+		fmt.Println("ANNOTS", nodeInfo.Node().Annotations)
+		fmt.Printf("ALLOC: %+v\n", nodeInfo.Allocatable)
+		fmt.Printf("REQ: %+v\n", nodeInfo.Requested)
+		for _, p := range nodeInfo.Pods {
+			fmt.Println("POD", p.Pod.Name)
+			fmt.Println("PCN", p.Pod.Spec.PriorityClassName)
+			fmt.Println("Owner", p.Pod.OwnerReferences)
+		}
 		if err := context.PredicateChecker.CheckPredicates(context.ClusterSnapshot, samplePod, nodeInfo.Node().Name); err == nil {
 			// add pods to option
 			option.Pods = append(option.Pods, eg.pods...)
