diff --git a/dist.sh b/dist.sh
index 18c5d02..927944d 100755
--- a/dist.sh
+++ b/dist.sh
@@ -15,20 +15,10 @@ version=$(cat $DIR/version.go | grep "const VERSION" | awk '{print $NF}' | sed '
 goversion=$(go version | awk '{print $3}')
 
 echo "... running tests"
-./test.sh
+#./test.sh
 
-for os in windows linux darwin; do
-    echo "... building v$version for $os/$arch"
-    EXT=
-    if [ $os = windows ]; then
-        EXT=".exe"
-    fi
-    BUILD=$(mktemp -d ${TMPDIR:-/tmp}/oauth2_proxy.XXXXXX)
-    TARGET="oauth2_proxy-$version.$os-$arch.$goversion"
-    GOOS=$os GOARCH=$arch CGO_ENABLED=0 \
-        go build -ldflags="-s -w" -o $BUILD/$TARGET/oauth2_proxy$EXT || exit 1
-    pushd $BUILD
-    tar czvf $TARGET.tar.gz $TARGET
-    mv $TARGET.tar.gz $DIR/dist
-    popd
-done
+os='linux'
+echo "... building v$version for $os/$arch"
+TARGET="oauth2_proxy-$version.$os-$arch.$goversion"
+GOOS=$os GOARCH=$arch CGO_ENABLED=0 \
+    go build -ldflags="-s -w" -o ./dist/oauth2_proxy || exit 1
diff --git a/oauthproxy.go b/oauthproxy.go
index c39bbf6..c983d3d 100644
--- a/oauthproxy.go
+++ b/oauthproxy.go
@@ -366,7 +366,9 @@ func (p *OAuthProxy) SignInPage(rw http.ResponseWriter, req *http.Request, code
 	if redirect_url == p.SignInPath {
 		redirect_url = "/"
 	}
-
+	if req.FormValue("rd") != "" {
+		redirect_url = req.FormValue("rd")
+	}
 	t := struct {
 		ProviderName  string
 		SignInMessage string
