diff --git a/plugin/cache/cache.go b/plugin/cache/cache.go
index a2ae9a307..98d0041e6 100644
--- a/plugin/cache/cache.go
+++ b/plugin/cache/cache.go
@@ -13,6 +13,9 @@ import (
 	"github.com/coredns/coredns/request"
 
 	"github.com/miekg/dns"
+	"golang.org/x/net/idna"
+	"golang.org/x/net/publicsuffix"
+	"strings"
 )
 
 // Cache is a plugin that looks up responses in a cache and caches replies.
@@ -51,6 +54,9 @@ type Cache struct {
 	// Keep ttl option
 	keepttl bool
 
+	// Metric for
+	enableNXDomainByClientDomain bool
+
 	// Testing.
 	now func() time.Time
 }
@@ -261,6 +267,10 @@ func (w *ResponseWriter) set(m *dns.Msg, key uint64, mt response.Type, duration
 		}
 		if w.ncache.Add(key, i) {
 			evictions.WithLabelValues(w.server, Denial, w.zonesMetricLabel, w.viewMetricLabel).Inc()
+			if w.enableNXDomainByClientDomain {
+				qLabel := RegistrableDomain(m.Question[0].Name)
+				cacheNXDomainByClientDomain.WithLabelValues(w.RemoteAddr().String(), qLabel).Inc()
+			}
 		}
 
 	case response.OtherError:
@@ -320,3 +330,22 @@ const (
 	// Denial is the class defined for negative caching.
 	Denial = "denial"
 )
+
+func normalizeQName(name string) string {
+	n := strings.TrimSuffix(name, ".")
+	n = strings.ToLower(n)
+	ascii, err := idna.Lookup.ToASCII(n)
+	if err != nil || ascii == "" {
+		return n
+	}
+	return ascii
+}
+
+func RegistrableDomain(name string) string {
+	n := normalizeQName(name)
+	etld1, err := publicsuffix.EffectiveTLDPlusOne(n)
+	if err != nil {
+		return n
+	}
+	return etld1
+}
diff --git a/plugin/cache/metrics.go b/plugin/cache/metrics.go
index 77edb0286..d5c0fada1 100644
--- a/plugin/cache/metrics.go
+++ b/plugin/cache/metrics.go
@@ -64,4 +64,13 @@ var (
 		Name:      "evictions_total",
 		Help:      "The count of cache evictions.",
 	}, []string{"server", "type", "zones", "view"})
+	cacheNXDomainByClientDomain = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: plugin.Namespace,
+			Subsystem: "cache",
+			Name:      "nxdomain_by_client_domain_total",
+			Help:      "NXDOMAIN responses served by cache, grouped by client IP and normalized domain.",
+		},
+		[]string{"src_ip", "qname"},
+	)
 )
diff --git a/plugin/cache/setup.go b/plugin/cache/setup.go
index f8278b872..3be8d9e33 100644
--- a/plugin/cache/setup.go
+++ b/plugin/cache/setup.go
@@ -246,6 +246,12 @@ func cacheParse(c *caddy.Controller) (*Cache, error) {
 					return nil, c.ArgErr()
 				}
 				ca.keepttl = true
+			case "nxdomain_verbose":
+				args := c.RemainingArgs()
+				if len(args) != 0 {
+					return nil, c.ArgErr()
+				}
+				ca.enableNXDomainByClientDomain = true
 			default:
 				return nil, c.ArgErr()
 			}
