diff --git a/plugin/cache/cache.go b/plugin/cache/cache.go
index 5c5e3dbfe..f1fc0c4e9 100644
--- a/plugin/cache/cache.go
+++ b/plugin/cache/cache.go
@@ -51,6 +51,9 @@ type Cache struct {
 	// Keep ttl option
 	keepttl bool
 
+	// Metric for
+	enableNXDomainByClient bool
+
 	// Testing.
 	now func() time.Time
 }
@@ -210,6 +213,11 @@ func (w *ResponseWriter) WriteMsg(res *dns.Msg) error {
 			w.set(res, key, mt, duration)
 			cacheSize.WithLabelValues(w.server, Success, w.zonesMetricLabel, w.viewMetricLabel).Set(float64(w.pcache.Len()))
 			cacheSize.WithLabelValues(w.server, Denial, w.zonesMetricLabel, w.viewMetricLabel).Set(float64(w.ncache.Len()))
+			if w.enableNXDomainByClient && mt == response.NameError {
+				qLabel := w.state.Name()
+				ipOnly := w.state.IP()
+				cacheNXDomainByClientDomain.WithLabelValues(ipOnly, qLabel).Inc()
+			}
 		} else {
 			// Don't log it, but increment counter
 			cacheDrops.WithLabelValues(w.server, w.zonesMetricLabel, w.viewMetricLabel).Inc()
diff --git a/plugin/cache/metrics.go b/plugin/cache/metrics.go
index 77edb0286..fb52226cc 100644
--- a/plugin/cache/metrics.go
+++ b/plugin/cache/metrics.go
@@ -64,4 +64,11 @@ var (
 		Name:      "evictions_total",
 		Help:      "The count of cache evictions.",
 	}, []string{"server", "type", "zones", "view"})
+	// cacheNXDomainByClient is counter for top client with NXDOMAIN errors.
+	cacheNXDomainByClientDomain = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: plugin.Namespace,
+		Subsystem: "cache",
+		Name:      "nxdomain_by_client_total",
+		Help:      "NXDOMAIN responses served by cache, grouped by client IP and normalized domain.",
+	}, []string{"src_ip", "qname"})
 )
diff --git a/plugin/cache/setup.go b/plugin/cache/setup.go
index f8278b872..81c9fa709 100644
--- a/plugin/cache/setup.go
+++ b/plugin/cache/setup.go
@@ -246,6 +246,12 @@ func cacheParse(c *caddy.Controller) (*Cache, error) {
 					return nil, c.ArgErr()
 				}
 				ca.keepttl = true
+			case "nxdomain_verbose":
+				args := c.RemainingArgs()
+				if len(args) != 0 {
+					return nil, c.ArgErr()
+				}
+				ca.enableNXDomainByClient = true
 			default:
 				return nil, c.ArgErr()
 			}
