diff --git a/Cargo.lock b/Cargo.lock
index 074592470..54dc92ac3 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -212,6 +212,15 @@ dependencies = [
  "num-traits",
 ]

+[[package]]
+name = "arbitrary"
+version = "1.3.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "e2d098ff73c1ca148721f37baad5ea6a465a13f9573aba8641fbbbae8164a54e"
+dependencies = [
+ "derive_arbitrary",
+]
+
 [[package]]
 name = "arc-swap"
 version = "1.6.0"
@@ -2603,52 +2612,6 @@ version = "0.2.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "8d7439c3735f405729d52c3fbbe4de140eaf938a1fe47d227c27f8254d4302a5"

-[[package]]
-name = "datadog-filter"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
-dependencies = [
- "datadog-search-syntax",
- "dyn-clone",
- "regex",
-]
-
-[[package]]
-name = "datadog-grok"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
-dependencies = [
- "bytes 1.4.0",
- "chrono",
- "chrono-tz",
- "lalrpop",
- "lalrpop-util",
- "nom",
- "once_cell",
- "onig",
- "ordered-float 3.7.0",
- "path",
- "peeking_take_while",
- "regex",
- "serde_json",
- "thiserror",
- "tracing 0.1.37",
- "value",
- "vrl-compiler",
-]
-
-[[package]]
-name = "datadog-search-syntax"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
-dependencies = [
- "itertools",
- "once_cell",
- "pest",
- "pest_derive",
- "regex",
-]
-
 [[package]]
 name = "db-key"
 version = "0.0.5"
@@ -2700,6 +2663,17 @@ dependencies = [
  "syn 1.0.109",
 ]

+[[package]]
+name = "derive_arbitrary"
+version = "1.3.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "f3cdeb9ec472d588e539a818b2dee436825730da08ad0017c4b1a17676bdc8b7"
+dependencies = [
+ "proc-macro2 1.0.57",
+ "quote 1.0.27",
+ "syn 1.0.109",
+]
+
 [[package]]
 name = "derive_more"
 version = "0.99.17"
@@ -5810,17 +5784,6 @@ version = "1.0.12"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "9f746c4065a8fa3fe23974dd82f15431cc8d40779821001404d10d2e79ca7d79"

-[[package]]
-name = "path"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
-dependencies = [
- "once_cell",
- "regex",
- "serde",
- "snafu",
-]
-
 [[package]]
 name = "pbkdf2"
 version = "0.11.0"
@@ -9114,25 +9077,6 @@ version = "0.1.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "830b7e5d4d90034032940e4ace0d9a9a057e7a45cd94e6c007832e39edb82f6d"

-[[package]]
-name = "value"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
-dependencies = [
- "bytes 1.4.0",
- "chrono",
- "mlua",
- "once_cell",
- "ordered-float 3.7.0",
- "path",
- "quickcheck",
- "regex",
- "serde",
- "serde_json",
- "snafu",
- "tracing 0.1.37",
-]
-
 [[package]]
 name = "vcpkg"
 version = "0.2.15"
@@ -9681,120 +9625,14 @@ checksum = "6a02e4885ed3bc0f2de90ea6dd45ebcbb66dacffe03547fadbb0eeae2770887d"

 [[package]]
 name = "vrl"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
-dependencies = [
- "datadog-filter",
- "datadog-grok",
- "datadog-search-syntax",
- "path",
- "value",
- "vrl-cli",
- "vrl-compiler",
- "vrl-core",
- "vrl-diagnostic",
- "vrl-parser",
- "vrl-stdlib",
- "vrl-tests",
-]
-
-[[package]]
-name = "vrl-cli"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
-dependencies = [
- "clap 4.1.14",
- "exitcode",
- "indoc",
- "once_cell",
- "path",
- "prettytable-rs",
- "regex",
- "rustyline",
- "serde_json",
- "thiserror",
- "value",
- "vrl-compiler",
- "vrl-core",
- "vrl-diagnostic",
- "vrl-stdlib",
- "webbrowser",
-]
-
-[[package]]
-name = "vrl-compiler"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
-dependencies = [
- "anymap",
- "bytes 1.4.0",
- "chrono",
- "chrono-tz",
- "dyn-clone",
- "getrandom 0.2.9",
- "indoc",
- "lalrpop-util",
- "ordered-float 3.7.0",
- "paste",
- "path",
- "regex",
- "serde",
- "snafu",
- "thiserror",
- "tracing 0.1.37",
- "value",
- "vrl-diagnostic",
- "vrl-parser",
-]
-
-[[package]]
-name = "vrl-core"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
-dependencies = [
- "bytes 1.4.0",
- "chrono",
- "chrono-tz",
- "derivative",
- "nom",
- "ordered-float 3.7.0",
- "path",
- "serde",
- "serde_json",
- "snafu",
- "value",
- "vrl-diagnostic",
-]
-
-[[package]]
-name = "vrl-diagnostic"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
-dependencies = [
- "codespan-reporting",
- "termcolor",
-]
-
-[[package]]
-name = "vrl-parser"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
-dependencies = [
- "lalrpop",
- "lalrpop-util",
- "ordered-float 3.7.0",
- "paste",
- "path",
- "thiserror",
- "vrl-diagnostic",
-]
-
-[[package]]
-name = "vrl-stdlib"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
+version = "0.4.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "a6236fdfaaa956af732a73630b8a8b43ef75e0a42fed6b94cf3c1c19c99daca5"
 dependencies = [
  "aes",
+ "ansi_term",
+ "anymap",
+ "arbitrary",
  "base16",
  "base64 0.21.0",
  "bytes 1.4.0",
@@ -9802,78 +9640,69 @@ dependencies = [
  "cfb-mode",
  "charset",
  "chrono",
+ "chrono-tz",
  "cidr-utils",
+ "clap 4.1.14",
+ "codespan-reporting",
  "csv",
  "ctr",
  "data-encoding",
- "datadog-filter",
- "datadog-grok",
- "datadog-search-syntax",
  "dns-lookup",
+ "dyn-clone",
+ "exitcode",
  "flate2",
+ "getrandom 0.2.9",
  "grok",
  "hex",
  "hmac",
  "hostname",
  "indexmap",
  "indoc",
+ "itertools",
+ "lalrpop",
+ "lalrpop-util",
  "md-5",
+ "mlua",
  "nom",
  "ofb",
  "once_cell",
+ "onig",
  "ordered-float 3.7.0",
- "path",
+ "paste",
+ "peeking_take_while",
  "percent-encoding",
+ "pest",
+ "pest_derive",
+ "prettydiff",
+ "prettytable-rs",
+ "quickcheck",
  "quoted_printable",
  "rand 0.8.5",
  "regex",
  "roxmltree",
  "rust_decimal",
+ "rustyline",
  "seahash",
  "serde",
  "serde_json",
  "sha-1",
  "sha2 0.10.6",
  "sha3",
+ "snafu",
  "strip-ansi-escapes",
  "syslog_loose",
+ "termcolor",
+ "thiserror",
  "tracing 0.1.37",
  "uaparser",
  "url",
  "utf8-width",
  "uuid",
- "value",
- "vrl-compiler",
- "vrl-core",
- "vrl-diagnostic",
+ "webbrowser",
  "woothee",
  "zstd 0.12.3+zstd.1.5.2",
 ]

-[[package]]
-name = "vrl-tests"
-version = "0.1.0"
-source = "git+https://github.com/vectordotdev/vrl?rev=v0.3.0#113005bcee6cd7b5ea0a53a7db2fc45ba4bc4125"
-dependencies = [
- "ansi_term",
- "chrono",
- "chrono-tz",
- "clap 4.1.14",
- "glob",
- "path",
- "prettydiff",
- "regex",
- "serde",
- "serde_json",
- "tikv-jemallocator",
- "tracing-subscriber",
- "value",
- "vrl-compiler",
- "vrl-core",
- "vrl-diagnostic",
- "vrl-stdlib",
-]
-
 [[package]]
 name = "vsimd"
 version = "0.8.0"
diff --git a/Cargo.toml b/Cargo.toml
index 49fd1c3fc..ce84f09f3 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -226,7 +226,7 @@ hex = { version = "0.4.3", default-features = false, optional = true }
 sha2 = { version = "0.10.6", default-features = false, optional = true }

 # VRL Lang
-vrl = { package = "vrl", git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0", features = ["cli"] }
+vrl = { package = "vrl", git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6", features = ["cli", "test"] }

 # External libs
 arc-swap = { version = "1.6", default-features = false, optional = true }
diff --git a/lib/codecs/Cargo.toml b/lib/codecs/Cargo.toml
index 9fda0b8cd..78f714dc4 100644
--- a/lib/codecs/Cargo.toml
+++ b/lib/codecs/Cargo.toml
@@ -25,7 +25,7 @@ snafu = { version = "0.7.4", default-features = false, features = ["futures"] }
 syslog_loose = { version = "0.18", default-features = false, optional = true }
 tokio-util = { version = "0.7", default-features = false, features = ["codec"] }
 tracing = { version = "0.1", default-features = false }
-vrl = { git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0", default-features = false, features = ["value"] }
+vrl = { git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6", default-features = false, features = ["value"] }
 vector-common = { path = "../vector-common", default-features = false }
 vector-config = { path = "../vector-config", default-features = false }
 vector-config-common = { path = "../vector-config-common", default-features = false }
diff --git a/lib/codecs/src/encoding/format/avro.rs b/lib/codecs/src/encoding/format/avro.rs
index 35bf45890..eb289fd0a 100644
--- a/lib/codecs/src/encoding/format/avro.rs
+++ b/lib/codecs/src/encoding/format/avro.rs
@@ -81,7 +81,7 @@ mod tests {
     use bytes::BytesMut;
     use indoc::indoc;
     use vector_core::event::{LogEvent, Value};
-    use vrl::value::btreemap;
+    use vrl::btreemap;

     use super::*;

diff --git a/lib/codecs/src/encoding/format/gelf.rs b/lib/codecs/src/encoding/format/gelf.rs
index 799fe41ad..3a0c6d046 100644
--- a/lib/codecs/src/encoding/format/gelf.rs
+++ b/lib/codecs/src/encoding/format/gelf.rs
@@ -241,7 +241,7 @@ mod tests {
     use super::*;
     use chrono::{DateTime, NaiveDateTime, Utc};
     use vector_core::event::{Event, EventMetadata};
-    use vrl::value::btreemap;
+    use vrl::btreemap;
     use vrl::value::Value;

     fn do_serialize(
diff --git a/lib/codecs/src/encoding/format/json.rs b/lib/codecs/src/encoding/format/json.rs
index 73ab79c78..943a421c7 100644
--- a/lib/codecs/src/encoding/format/json.rs
+++ b/lib/codecs/src/encoding/format/json.rs
@@ -91,7 +91,7 @@ mod tests {
     use chrono::{TimeZone, Timelike, Utc};
     use vector_core::event::{LogEvent, Metric, MetricKind, MetricValue, StatisticKind, Value};
     use vector_core::metric_tags;
-    use vrl::value::btreemap;
+    use vrl::btreemap;

     use super::*;

diff --git a/lib/codecs/src/encoding/format/logfmt.rs b/lib/codecs/src/encoding/format/logfmt.rs
index 37b0351b0..e656cc2c5 100644
--- a/lib/codecs/src/encoding/format/logfmt.rs
+++ b/lib/codecs/src/encoding/format/logfmt.rs
@@ -60,7 +60,7 @@ mod tests {
     use super::*;
     use bytes::BytesMut;
     use vector_core::event::{LogEvent, Value};
-    use vrl::value::btreemap;
+    use vrl::btreemap;

     #[test]
     fn serialize_logfmt() {
diff --git a/lib/codecs/src/encoding/format/native_json.rs b/lib/codecs/src/encoding/format/native_json.rs
index 6762c9273..854bba9d9 100644
--- a/lib/codecs/src/encoding/format/native_json.rs
+++ b/lib/codecs/src/encoding/format/native_json.rs
@@ -53,7 +53,7 @@ impl Encoder<Event> for NativeJsonSerializer {
 mod tests {
     use bytes::BytesMut;
     use vector_core::event::{LogEvent, Value};
-    use vrl::value::btreemap;
+    use vrl::btreemap;

     use super::*;

diff --git a/lib/enrichment/Cargo.toml b/lib/enrichment/Cargo.toml
index ef118022f..c88d81e2d 100644
--- a/lib/enrichment/Cargo.toml
+++ b/lib/enrichment/Cargo.toml
@@ -10,4 +10,4 @@ arc-swap = { version = "1.6.0", default-features = false }
 dyn-clone = { version = "1.0.11", default-features = false }
 chrono = { version = "0.4.19", default-features = false }
 vector-common = { path = "../vector-common", default-features = false, features = [ "btreemap", "conversion", "serde" ] }
-vrl = { git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0", default-features = false, features = ["diagnostic"] }
+vrl = { git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6", default-features = false, features = ["diagnostic"] }
diff --git a/lib/enrichment/src/find_enrichment_table_records.rs b/lib/enrichment/src/find_enrichment_table_records.rs
index f4fbfd03c..bfe0aa718 100644
--- a/lib/enrichment/src/find_enrichment_table_records.rs
+++ b/lib/enrichment/src/find_enrichment_table_records.rs
@@ -194,6 +194,7 @@ mod tests {
     use vector_common::TimeZone;
     use vrl::compiler::state::RuntimeState;
     use vrl::compiler::TargetValue;
+    use vrl::value;
     use vrl::value::Secrets;

     use super::*;
diff --git a/lib/enrichment/src/get_enrichment_table_record.rs b/lib/enrichment/src/get_enrichment_table_record.rs
index 93c432ccd..edaac8449 100644
--- a/lib/enrichment/src/get_enrichment_table_record.rs
+++ b/lib/enrichment/src/get_enrichment_table_record.rs
@@ -186,6 +186,7 @@ mod tests {
     use vector_common::TimeZone;
     use vrl::compiler::state::RuntimeState;
     use vrl::compiler::TargetValue;
+    use vrl::value;
     use vrl::value::Secrets;

     use super::*;
diff --git a/lib/opentelemetry-proto/Cargo.toml b/lib/opentelemetry-proto/Cargo.toml
index 107fe072f..23c91e4c2 100644
--- a/lib/opentelemetry-proto/Cargo.toml
+++ b/lib/opentelemetry-proto/Cargo.toml
@@ -17,5 +17,5 @@ lookup = { package = "vector-lookup", path = "../vector-lookup", default-feature
 ordered-float = { version = "3.7.0", default-features = false }
 prost = { version = "0.11", default-features = false, features = ["std"] }
 tonic = { version = "0.9", default-features = false, features = ["codegen", "gzip", "prost", "tls", "tls-roots", "transport"] }
-vrl = { git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0", default-features = false, features = ["value"] }
+vrl = { git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6", default-features = false, features = ["value"] }
 vector-core = { path = "../vector-core", default-features = false }
diff --git a/lib/vector-common/Cargo.toml b/lib/vector-common/Cargo.toml
index 4ef8a6628..c2d26eb49 100644
--- a/lib/vector-common/Cargo.toml
+++ b/lib/vector-common/Cargo.toml
@@ -62,7 +62,7 @@ snafu = { version = "0.7", optional = true }
 stream-cancel = { version = "0.8.1", default-features = false }
 tokio = { version = "1.28.1", default-features = false, features = ["macros", "time"] }
 tracing = { version = "0.1.34", default-features = false }
-vrl = { git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0", default-features = false, features = ["value", "core", "compiler"] }
+vrl = { git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6", default-features = false, features = ["value", "core", "compiler"] }
 vector-config = { path = "../vector-config" }
 vector-config-common = { path = "../vector-config-common" }
 vector-config-macros = { path = "../vector-config-macros" }
diff --git a/lib/vector-common/src/lib.rs b/lib/vector-common/src/lib.rs
index bcf58c72d..832eaf0d5 100644
--- a/lib/vector-common/src/lib.rs
+++ b/lib/vector-common/src/lib.rs
@@ -13,7 +13,7 @@
 #![deny(unused_comparisons)]

 #[cfg(feature = "btreemap")]
-pub use vrl::value::btreemap;
+pub use vrl::btreemap;

 #[cfg(feature = "byte_size_of")]
 pub mod byte_size_of;
diff --git a/lib/vector-config/Cargo.toml b/lib/vector-config/Cargo.toml
index 63227b44d..4842b74a9 100644
--- a/lib/vector-config/Cargo.toml
+++ b/lib/vector-config/Cargo.toml
@@ -26,7 +26,7 @@ snafu = { version = "0.7.4", default-features = false }
 toml = { version = "0.7.3", default-features = false }
 tracing = { version = "0.1.34", default-features = false }
 url = { version = "2.3.1", default-features = false, features = ["serde"] }
-vrl = { git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0", default-features = false, features = ["compiler"] }
+vrl = { git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6", default-features = false, features = ["compiler"] }
 vector-config-common = { path = "../vector-config-common" }
 vector-config-macros = { path = "../vector-config-macros" }

diff --git a/lib/vector-core/Cargo.toml b/lib/vector-core/Cargo.toml
index deac7b598..e1e0be5a8 100644
--- a/lib/vector-core/Cargo.toml
+++ b/lib/vector-core/Cargo.toml
@@ -65,7 +65,7 @@ vector-common = { path = "../vector-common" }
 vector-config = { path = "../vector-config" }
 vector-config-common = { path = "../vector-config-common" }
 vector-config-macros = { path = "../vector-config-macros" }
-vrl = { git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0" }
+vrl = { git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6" }

 [target.'cfg(target_os = "macos")'.dependencies]
 security-framework = "2.9.0"
@@ -94,7 +94,7 @@ rand = "0.8.5"
 rand_distr = "0.4.3"
 tracing-subscriber = { version = "0.3.17", default-features = false, features = ["env-filter", "fmt", "ansi", "registry"] }
 vector-common = { path = "../vector-common", default-features = false, features = ["test"] }
-vrl = { git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0", default-features = false, features = ["value", "arbitrary", "lua"] }
+vrl = { git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6", default-features = false, features = ["value", "arbitrary", "lua"] }

 [features]
 api = ["dep:async-graphql"]
diff --git a/lib/vector-core/src/event/vrl_target.rs b/lib/vector-core/src/event/vrl_target.rs
index 861721eb7..f6277cdfe 100644
--- a/lib/vector-core/src/event/vrl_target.rs
+++ b/lib/vector-core/src/event/vrl_target.rs
@@ -588,7 +588,7 @@ mod test {
     use chrono::{offset::TimeZone, Utc};
     use lookup::owned_value_path;
     use similar_asserts::assert_eq;
-    use vrl::value::btreemap;
+    use vrl::btreemap;

     use super::super::MetricValue;
     use super::*;
@@ -858,7 +858,7 @@ mod test {

     #[test]
     fn log_into_events() {
-        use vrl::value::btreemap;
+        use vrl::btreemap;

         let cases = vec![
             (
diff --git a/lib/vector-lookup/Cargo.toml b/lib/vector-lookup/Cargo.toml
index d0255184d..daec79b1e 100644
--- a/lib/vector-lookup/Cargo.toml
+++ b/lib/vector-lookup/Cargo.toml
@@ -10,4 +10,4 @@ license = "MPL-2.0"
 serde = { version = "1.0.163", default-features = false, features = ["derive", "alloc"] }
 vector-config = { path = "../vector-config" }
 vector-config-macros = { path = "../vector-config-macros" }
-vrl = { git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0", default-features = false, features = ["path"] }
+vrl = { git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6", default-features = false, features = ["path"] }
diff --git a/lib/vector-lookup/src/lib.rs b/lib/vector-lookup/src/lib.rs
index 7058cfc6e..e32e73b5f 100644
--- a/lib/vector-lookup/src/lib.rs
+++ b/lib/vector-lookup/src/lib.rs
@@ -1,7 +1,7 @@
 #![deny(warnings)]

-pub use vrl::path::{
-    event_path, metadata_path, owned_value_path, path, OwnedTargetPath, OwnedValuePath, PathPrefix,
-};
+pub use vrl::path::{OwnedTargetPath, OwnedValuePath, PathPrefix};
+
+pub use vrl::{event_path, metadata_path, owned_value_path, path};

 pub mod lookup_v2;
diff --git a/lib/vector-vrl/cli/Cargo.toml b/lib/vector-vrl/cli/Cargo.toml
index 54d027909..dd2c451de 100644
--- a/lib/vector-vrl/cli/Cargo.toml
+++ b/lib/vector-vrl/cli/Cargo.toml
@@ -9,4 +9,4 @@ license = "MPL-2.0"
 [dependencies]
 clap = { version = "4.1.14", features = ["derive"] }
 vector-vrl-functions = { path = "../functions" }
-vrl = { git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0", default-features = false, features = ["stdlib", "cli"] }
+vrl = { git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6", default-features = false, features = ["stdlib", "cli"] }
diff --git a/lib/vector-vrl/functions/Cargo.toml b/lib/vector-vrl/functions/Cargo.toml
index 08fb64069..787637bad 100644
--- a/lib/vector-vrl/functions/Cargo.toml
+++ b/lib/vector-vrl/functions/Cargo.toml
@@ -7,4 +7,4 @@ publish = false
 license = "MPL-2.0"

 [dependencies]
-vrl = { git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0", default-features = false, features = ["compiler", "path", "diagnostic"] }
+vrl = { git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6", default-features = false, features = ["compiler", "path", "diagnostic"] }
diff --git a/lib/vector-vrl/tests/Cargo.toml b/lib/vector-vrl/tests/Cargo.toml
index 12c28a88a..2ff392755 100644
--- a/lib/vector-vrl/tests/Cargo.toml
+++ b/lib/vector-vrl/tests/Cargo.toml
@@ -7,7 +7,7 @@ publish = false

 [dependencies]
 enrichment = { path = "../../enrichment" }
-vrl = { git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0", features = ["test_framework"]}
+vrl = { git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6", features = ["test_framework"]}
 vector-vrl-functions = { path = "../../vector-vrl/functions" }

 ansi_term = "0.12"
diff --git a/lib/vector-vrl/web-playground/Cargo.toml b/lib/vector-vrl/web-playground/Cargo.toml
index 79433c22f..72a911893 100644
--- a/lib/vector-vrl/web-playground/Cargo.toml
+++ b/lib/vector-vrl/web-playground/Cargo.toml
@@ -10,7 +10,7 @@ crate-type = ["cdylib"]

 [dependencies]
 wasm-bindgen = "0.2"
-vrl = { git = "https://github.com/vectordotdev/vrl", rev = "v0.3.0", default-features = false, features = ["value", "stdlib"] }
+vrl = { git = "https://github.com/vectordotdev/vrl", rev="52959f8c24d5524c6e0a4eaef83962049fbfe8f6", default-features = false, features = ["value", "stdlib"] }
 serde = { version = "1.0", features = ["derive"] }
 serde-wasm-bindgen = "0.5"
 gloo-utils = { version = "0.1", features = ["serde"] }
diff --git a/src/config/enterprise.rs b/src/config/enterprise.rs
index 97075716b..bce787aa7 100644
--- a/src/config/enterprise.rs
+++ b/src/config/enterprise.rs
@@ -833,9 +833,9 @@ mod test {
     use indexmap::IndexMap;
     use tokio::time::sleep;
     use vector_core::config::proxy::ProxyConfig;
+    use vrl::btreemap;
     use vrl::compiler::state::ExternalEnv;
     use vrl::compiler::{compile, compile_with_external, CompileConfig};
-    use vrl::value::btreemap;
     use vrl::value::kind::Collection;
     use vrl::value::Kind;
     use wiremock::{matchers, Mock, MockServer, ResponseTemplate};
diff --git a/src/sinks/influxdb/logs.rs b/src/sinks/influxdb/logs.rs
index 80781c34f..0bcd27ef3 100644
--- a/src/sinks/influxdb/logs.rs
+++ b/src/sinks/influxdb/logs.rs
@@ -883,7 +883,7 @@ mod integration_tests {
     use std::sync::Arc;
     use vector_core::config::{LegacyKey, LogNamespace};
     use vector_core::event::{BatchNotifier, BatchStatus, Event, LogEvent};
-    use vrl::value::value;
+    use vrl::value;

     use super::*;
     use crate::{
diff --git a/src/sources/aws_kinesis_firehose/mod.rs b/src/sources/aws_kinesis_firehose/mod.rs
index cefb74bea..d546ce175 100644
--- a/src/sources/aws_kinesis_firehose/mod.rs
+++ b/src/sources/aws_kinesis_firehose/mod.rs
@@ -255,7 +255,7 @@ mod tests {
     use similar_asserts::assert_eq;
     use tokio::time::{sleep, Duration};
     use vector_common::assert_event_data_eq;
-    use vrl::value::value;
+    use vrl::value;

     use super::*;
     use crate::{
diff --git a/src/sources/dnstap/schema.rs b/src/sources/dnstap/schema.rs
index dcf4a6093..90a76dfa3 100644
--- a/src/sources/dnstap/schema.rs
+++ b/src/sources/dnstap/schema.rs
@@ -1,6 +1,6 @@
 use lookup::{owned_value_path, OwnedValuePath};
 use std::collections::BTreeMap;
-use vrl::value::btreemap;
+use vrl::btreemap;
 use vrl::value::{
     kind::{Collection, Field},
     Kind,
diff --git a/src/sources/docker_logs/tests.rs b/src/sources/docker_logs/tests.rs
index e05543127..c95692090 100644
--- a/src/sources/docker_logs/tests.rs
+++ b/src/sources/docker_logs/tests.rs
@@ -45,7 +45,7 @@ mod integration_tests {
     };
     use futures::{stream::TryStreamExt, FutureExt};
     use similar_asserts::assert_eq;
-    use vrl::value::value;
+    use vrl::value;

     /// None if docker is not present on the system
     async fn source_with<'a, L: Into<Option<&'a str>>>(
diff --git a/src/sources/exec/mod.rs b/src/sources/exec/mod.rs
index c540c8e97..1f8d755e7 100644
--- a/src/sources/exec/mod.rs
+++ b/src/sources/exec/mod.rs
@@ -722,7 +722,7 @@ mod tests {
     use bytes::Bytes;
     use std::io::Cursor;
     use vector_core::event::EventMetadata;
-    use vrl::value::value;
+    use vrl::value;

     #[cfg(unix)]
     use futures::task::Poll;
diff --git a/src/sources/file.rs b/src/sources/file.rs
index 8c31338b3..76bec5f10 100644
--- a/src/sources/file.rs
+++ b/src/sources/file.rs
@@ -818,7 +818,7 @@ mod tests {
         sources::file,
         test_util::components::{assert_source_compliance, FILE_SOURCE_TAGS},
     };
-    use vrl::value::value;
+    use vrl::value;

     #[test]
     fn generate_config() {
diff --git a/src/sources/file_descriptors/file_descriptor.rs b/src/sources/file_descriptors/file_descriptor.rs
index 9a97591f0..074067474 100644
--- a/src/sources/file_descriptors/file_descriptor.rs
+++ b/src/sources/file_descriptors/file_descriptor.rs
@@ -112,7 +112,7 @@ mod tests {
         SourceSender,
     };
     use futures::StreamExt;
-    use vrl::value::value;
+    use vrl::value;

     #[test]
     fn generate_config() {
diff --git a/src/sources/file_descriptors/stdin.rs b/src/sources/file_descriptors/stdin.rs
index 3b6d5aece..0f21db79a 100644
--- a/src/sources/file_descriptors/stdin.rs
+++ b/src/sources/file_descriptors/stdin.rs
@@ -117,7 +117,7 @@ mod tests {
     };
     use futures::StreamExt;
     use lookup::path;
-    use vrl::value::value;
+    use vrl::value;

     #[test]
     fn generate_config() {
diff --git a/src/sources/gcp_pubsub.rs b/src/sources/gcp_pubsub.rs
index 10b72a11f..ffef638f2 100644
--- a/src/sources/gcp_pubsub.rs
+++ b/src/sources/gcp_pubsub.rs
@@ -841,7 +841,7 @@ mod integration_tests {
     use once_cell::sync::Lazy;
     use serde_json::{json, Value};
     use tokio::time::{Duration, Instant};
-    use vrl::value::btreemap;
+    use vrl::btreemap;

     use super::*;
     use crate::config::{ComponentKey, ProxyConfig};
diff --git a/src/sources/kafka.rs b/src/sources/kafka.rs
index c9418baab..762c13f9c 100644
--- a/src/sources/kafka.rs
+++ b/src/sources/kafka.rs
@@ -913,7 +913,7 @@ mod integration_test {
     use tokio::time::sleep;
     use vector_buffers::topology::channel::BufferReceiver;
     use vector_core::event::EventStatus;
-    use vrl::value::value;
+    use vrl::value;

     use super::{test::*, *};
     use crate::{
diff --git a/src/sources/kubernetes_logs/parser/cri.rs b/src/sources/kubernetes_logs/parser/cri.rs
index e4d589b8b..ce8f6a903 100644
--- a/src/sources/kubernetes_logs/parser/cri.rs
+++ b/src/sources/kubernetes_logs/parser/cri.rs
@@ -189,7 +189,7 @@ pub mod tests {

     use super::{super::test_util, *};
     use crate::{event::LogEvent, test_util::trace_init};
-    use vrl::value::value;
+    use vrl::value;

     fn make_long_string(base: &str, len: usize) -> String {
         base.chars().cycle().take(len).collect()
diff --git a/src/sources/kubernetes_logs/parser/docker.rs b/src/sources/kubernetes_logs/parser/docker.rs
index db25f52c6..5b3c71201 100644
--- a/src/sources/kubernetes_logs/parser/docker.rs
+++ b/src/sources/kubernetes_logs/parser/docker.rs
@@ -207,7 +207,7 @@ enum NormalizationError {
 pub mod tests {
     use super::{super::test_util, *};
     use crate::test_util::trace_init;
-    use vrl::value::value;
+    use vrl::value;

     fn make_long_string(base: &str, len: usize) -> String {
         base.chars().cycle().take(len).collect()
diff --git a/src/sources/kubernetes_logs/parser/mod.rs b/src/sources/kubernetes_logs/parser/mod.rs
index b8b5b4c6e..fe6f76547 100644
--- a/src/sources/kubernetes_logs/parser/mod.rs
+++ b/src/sources/kubernetes_logs/parser/mod.rs
@@ -84,7 +84,7 @@ impl FunctionTransform for Parser {
 mod tests {
     use bytes::Bytes;
     use lookup::event_path;
-    use vrl::value::value;
+    use vrl::value;

     use super::*;
     use crate::{event::Event, event::LogEvent, test_util::trace_init};
diff --git a/src/sources/kubernetes_logs/parser/test_util.rs b/src/sources/kubernetes_logs/parser/test_util.rs
index 745e849ad..fb89b88d7 100644
--- a/src/sources/kubernetes_logs/parser/test_util.rs
+++ b/src/sources/kubernetes_logs/parser/test_util.rs
@@ -4,7 +4,8 @@ use similar_asserts::assert_eq;
 use chrono::{DateTime, Utc};
 use lookup::{event_path, metadata_path};
 use vector_core::{config::LogNamespace, event};
-use vrl::value::{value, Value};
+use vrl::value;
+use vrl::value::Value;

 use crate::{
     event::{Event, LogEvent},
diff --git a/src/sources/opentelemetry/tests.rs b/src/sources/opentelemetry/tests.rs
index a6085cff2..798759fa1 100644
--- a/src/sources/opentelemetry/tests.rs
+++ b/src/sources/opentelemetry/tests.rs
@@ -12,7 +12,7 @@ use similar_asserts::assert_eq;
 use std::collections::BTreeMap;
 use tonic::Request;
 use vector_core::config::LogNamespace;
-use vrl::value::value;
+use vrl::value;

 use crate::{
     config::{SourceConfig, SourceContext},
diff --git a/src/sources/redis/mod.rs b/src/sources/redis/mod.rs
index 05723b6cc..c8ab0ed6d 100644
--- a/src/sources/redis/mod.rs
+++ b/src/sources/redis/mod.rs
@@ -321,7 +321,7 @@ mod integration_test {
         },
         SourceSender,
     };
-    use vrl::value::value;
+    use vrl::value;

     const REDIS_SERVER: &str = "redis://redis:6379/0";

diff --git a/src/sources/socket/mod.rs b/src/sources/socket/mod.rs
index e0d212a59..93366629f 100644
--- a/src/sources/socket/mod.rs
+++ b/src/sources/socket/mod.rs
@@ -344,8 +344,9 @@ mod test {
         time::{timeout, Duration, Instant},
     };
     use vector_core::event::EventContainer;
-    use vrl::value::value;
-    use vrl::value::{btreemap, Value};
+    use vrl::btreemap;
+    use vrl::value;
+    use vrl::value::Value;

     #[cfg(unix)]
     use {
diff --git a/src/transforms/remap.rs b/src/transforms/remap.rs
index 5107a32f3..877a7dde0 100644
--- a/src/transforms/remap.rs
+++ b/src/transforms/remap.rs
@@ -721,10 +721,8 @@ mod tests {

     use indoc::{formatdoc, indoc};
     use vector_core::{config::GlobalOptions, event::EventMetadata, metric_tags};
-    use vrl::value::{
-        btreemap,
-        kind::{Collection, Index},
-    };
+    use vrl::btreemap;
+    use vrl::value::kind::{Collection, Index};

     use super::*;
     use crate::{
