diff --git a/pkg/kube/multicluster/cluster.go b/pkg/kube/multicluster/cluster.go
index 619a0c0670..ad72089c77 100644
--- a/pkg/kube/multicluster/cluster.go
+++ b/pkg/kube/multicluster/cluster.go
@@ -26,8 +26,10 @@ import (
 	"istio.io/istio/pkg/config/mesh"
 	"istio.io/istio/pkg/kube"
 	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/kube/kubetypes"
 	filter "istio.io/istio/pkg/kube/namespace"
 	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/util/sets"
 )
 
 // Cluster defines cluster struct
@@ -80,8 +82,10 @@ func (c *Cluster) Run(mesh mesh.Watcher, handlers []handler, action ACTION) {
 	// This must be done before we build components, so they can access the filter.
 	namespaces := kclient.New[*corev1.Namespace](c.Client)
 	// This will start a namespace informer and wait for it to be ready. So we must start it in a go routine to avoid blocking.
-	filter := filter.NewDiscoveryNamespacesFilter(namespaces, mesh, c.stop)
-	kube.SetObjectFilter(c.Client, filter)
+	namespaceFilter := filter.NewDiscoveryNamespacesFilter(namespaces, mesh, c.stop)
+	secretFilter := newSecretObjectFilter(features.InformerWatchNamespace)
+	combinedFilter := combineFilters(namespaceFilter, secretFilter)
+	kube.SetObjectFilter(c.Client, combinedFilter)
 
 	syncers := make([]ComponentConstraint, 0, len(handlers))
 	for _, h := range handlers {
@@ -106,6 +110,27 @@ func (c *Cluster) Run(mesh mesh.Watcher, handlers []handler, action ACTION) {
 	c.initialSync.Store(true)
 }
 
+func combineFilters(f1, f2 kubetypes.DynamicObjectFilter) kubetypes.DynamicObjectFilter {
+	return &combinedFilter{
+		f1: f1,
+		f2: f2,
+	}
+}
+
+type combinedFilter struct {
+	f1 kubetypes.DynamicObjectFilter
+	f2 kubetypes.DynamicObjectFilter
+}
+
+func (c *combinedFilter) Filter(obj any) bool {
+	return c.f1.Filter(obj) && c.f2.Filter(obj)
+}
+
+func (c *combinedFilter) AddHandler(handler func(selected, deselected sets.String)) {
+	c.f1.AddHandler(handler)
+	c.f2.AddHandler(handler)
+}
+
 // Stop closes the stop channel, if is safe to be called multi times.
 func (c *Cluster) Stop() {
 	select {
diff --git a/pkg/kube/multicluster/secretcontroller.go b/pkg/kube/multicluster/secretcontroller.go
index 824ac68268..23c02e2657 100644
--- a/pkg/kube/multicluster/secretcontroller.go
+++ b/pkg/kube/multicluster/secretcontroller.go
@@ -33,6 +33,7 @@ import (
 	"istio.io/istio/pkg/kube"
 	"istio.io/istio/pkg/kube/controllers"
 	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/kube/kubetypes"
 	"istio.io/istio/pkg/log"
 	"istio.io/istio/pkg/monitoring"
 	"istio.io/istio/pkg/slices"
@@ -242,6 +243,36 @@ func DefaultBuildClientsFromConfig(kubeConfig []byte, clusterID cluster.ID, conf
 	return clients, nil
 }
 
+func newSecretObjectFilter(watchNamespace string) kubetypes.DynamicObjectFilter {
+	const allowedSecretInAllNamespaces = "d8-istio-sidecar-registry"
+	allowedSecretsInWatchNamespace := map[string]bool{
+		"cacerts":                            true,
+		"d8-remote-clusters-public-metadata": true,
+		"d8-remote-authn-keypair":            true,
+		"d8-istio-sidecar-registry":          true,
+	}
+
+	return kubetypes.NewStaticObjectFilter(func(obj any) bool {
+		secret, ok := obj.(*corev1.Secret)
+		if !ok {
+			return true
+		}
+
+		secretName := secret.GetName()
+		secretNamespace := secret.GetNamespace()
+
+		if secretName == allowedSecretInAllNamespaces {
+			return true
+		}
+
+		if watchNamespace != "" && secretNamespace == watchNamespace {
+			return allowedSecretsInWatchNamespace[secretName]
+		}
+
+		return false
+	})
+}
+
 func (c *Controller) createRemoteCluster(kubeConfig []byte, clusterID string) (*Cluster, error) {
 	clients, err := c.ClientBuilder(kubeConfig, cluster.ID(clusterID), c.configOverrides...)
 	if err != nil {
