diff --git a/cni/pkg/plugin/sidecar_redirect.go b/cni/pkg/plugin/sidecar_redirect.go
index cb9e209726..9e283c6ea2 100644
--- a/cni/pkg/plugin/sidecar_redirect.go
+++ b/cni/pkg/plugin/sidecar_redirect.go
@@ -31,8 +31,8 @@ const (
 	redirectModeTPROXY           = "TPROXY"
 	defaultProxyStatusPort       = "15020"
 	defaultRedirectToPort        = "15001"
-	defaultNoRedirectUID         = "1337"
-	defaultNoRedirectGID         = "1337"
+	defaultNoRedirectUID         = "64535,1337"
+	defaultNoRedirectGID         = "64535,1337"
 	defaultRedirectMode          = redirectModeREDIRECT
 	defaultRedirectIPCidr        = "*"
 	defaultRedirectExcludeIPCidr = ""
@@ -221,13 +221,13 @@ func NewRedirect(pi *PodInfo) (*Redirect, error) {
 	}
 
 	if pi.ProxyUID != nil && *pi.ProxyUID != 0 {
-		redir.noRedirectUID = fmt.Sprintf("%d", *pi.ProxyUID)
+		redir.noRedirectUID = fmt.Sprintf("%d,1337", *pi.ProxyUID)
 	} else {
 		redir.noRedirectUID = defaultNoRedirectUID
 	}
 
 	if pi.ProxyGID != nil && *pi.ProxyGID != 0 {
-		redir.noRedirectGID = fmt.Sprintf("%d", *pi.ProxyGID)
+		redir.noRedirectGID = fmt.Sprintf("%d,1337", *pi.ProxyGID)
 	} else {
 		redir.noRedirectGID = defaultNoRedirectGID
 	}
diff --git a/docker/Dockerfile.base b/docker/Dockerfile.base
index 382eb51855..6b803b27f5 100644
--- a/docker/Dockerfile.base
+++ b/docker/Dockerfile.base
@@ -30,5 +30,5 @@ RUN apt-get update && \
   && update-alternatives --set ip6tables /usr/sbin/ip6tables-legacy
 
 # Sudoers used to allow tcpdump and other debug utilities.
-RUN useradd -m --uid 1337 istio-proxy && \
+RUN useradd -m --uid 64535 istio-proxy && \
   echo "istio-proxy ALL=NOPASSWD: ALL" >> /etc/sudoers
diff --git a/docker/Dockerfile.distroless b/docker/Dockerfile.distroless
index f6d2b0d9c1..38b9f443e7 100644
--- a/docker/Dockerfile.distroless
+++ b/docker/Dockerfile.distroless
@@ -7,8 +7,8 @@ FROM ubuntu:noble AS ubuntu_source
 # Modify contents of container
 COPY --from=distroless_source /etc/ /home/etc
 COPY --from=distroless_source /home/nonroot /home/nonroot
-RUN echo istio-proxy:x:1337: >> /home/etc/group
-RUN echo istio-proxy:x:1337:1337:istio-proxy:/nonexistent:/sbin/nologin >> /home/etc/passwd
+RUN echo istio-proxy:x:64535: >> /home/etc/group
+RUN echo istio-proxy:x:64535:64535:istio-proxy:/nonexistent:/sbin/nologin >> /home/etc/passwd
 
 # Customize distroless with the following:
 # - password file
diff --git a/manifests/charts/istio-control/istio-discovery/files/injection-template.yaml b/manifests/charts/istio-control/istio-discovery/files/injection-template.yaml
index 3b3f69cd9d..b7d84d5121 100644
--- a/manifests/charts/istio-control/istio-discovery/files/injection-template.yaml
+++ b/manifests/charts/istio-control/istio-discovery/files/injection-template.yaml
@@ -88,7 +88,7 @@ spec:
     - "-z"
     - {{ .MeshConfig.ProxyInboundListenPort | default "15006" | quote }}
     - "-u"
-    - {{ .ProxyUID | default "1337" | quote }}
+    - {{ printf "%d,1337" .ProxyUID | quote }}
     - "-m"
     - "{{ annotation .ObjectMeta `sidecar.istio.io/interceptionMode` .ProxyConfig.InterceptionMode }}"
     - "-i"
diff --git a/pkg/config/analysis/analyzers/deployment/pod.go b/pkg/config/analysis/analyzers/deployment/pod.go
index 7d727deb95..cd7560dabe 100644
--- a/pkg/config/analysis/analyzers/deployment/pod.go
+++ b/pkg/config/analysis/analyzers/deployment/pod.go
@@ -31,7 +31,7 @@ type ApplicationUIDAnalyzer struct{}
 var _ analysis.Analyzer = &ApplicationUIDAnalyzer{}
 
 const (
-	UserID = int64(1337)
+	UserID = int64(64535)
 )
 
 func (appUID *ApplicationUIDAnalyzer) Metadata() analysis.Metadata {
diff --git a/pkg/envoy/proxy.go b/pkg/envoy/proxy.go
index 576234fc54..217d8892b5 100644
--- a/pkg/envoy/proxy.go
+++ b/pkg/envoy/proxy.go
@@ -209,8 +209,8 @@ func (e *envoy) Run(abort <-chan error) error {
 	if e.AgentIsRoot {
 		cmd.SysProcAttr = &syscall.SysProcAttr{}
 		cmd.SysProcAttr.Credential = &syscall.Credential{
-			Uid: 1337,
-			Gid: 1337,
+			Uid: 64535,
+			Gid: 64535,
 		}
 	}
 
diff --git a/pkg/kube/inject/inject.go b/pkg/kube/inject/inject.go
index d875f44783..0d6fd41845 100644
--- a/pkg/kube/inject/inject.go
+++ b/pkg/kube/inject/inject.go
@@ -904,7 +904,7 @@ func updateClusterEnvs(container *corev1.Container, newKVs map[string]string) {
 }
 
 // GetProxyIDs returns the UID and GID to be used in the RunAsUser and RunAsGroup fields in the template
-// Inspects the namespace metadata for hints and fallbacks to the usual value of 1337.
+// Inspects the namespace metadata for hints and fallbacks to the usual value of 64535.
 func GetProxyIDs(namespace *corev1.Namespace) (uid int64, gid int64) {
 	uid = constants.DefaultProxyUIDInt
 	gid = constants.DefaultProxyUIDInt
diff --git a/pkg/kube/inject/webhook.go b/pkg/kube/inject/webhook.go
index 0ace3752ca..487fcc7f5c 100644
--- a/pkg/kube/inject/webhook.go
+++ b/pkg/kube/inject/webhook.go
@@ -644,7 +644,7 @@ func adjustInitContainerUser(finalPod *corev1.Pod, originalPod *corev1.Pod, prox
 	}
 	for i := range initContainer.Args {
 		if initContainer.Args[i] == "-u" {
-			initContainer.Args[i+1] = fmt.Sprintf("%d", *userContainer.SecurityContext.RunAsUser)
+			initContainer.Args[i+1] = fmt.Sprintf("%d,1337", *userContainer.SecurityContext.RunAsUser)
 			return
 		}
 	}
diff --git a/tools/istio-iptables/pkg/constants/constants.go b/tools/istio-iptables/pkg/constants/constants.go
index 6ac71c5619..06c8f579f8 100644
--- a/tools/istio-iptables/pkg/constants/constants.go
+++ b/tools/istio-iptables/pkg/constants/constants.go
@@ -116,8 +116,8 @@ Only applies when traffic from all groups (i.e. "*") is being redirected to Envo
 )
 
 const (
-	DefaultProxyUID    = "1337"
-	DefaultProxyUIDInt = int64(1337)
+	DefaultProxyUID    = "64535"
+	DefaultProxyUIDInt = int64(64535)
 )
 
 // Constants used in environment variables
