diff --git a/pkg/collector/collect.go b/pkg/collector/collect.go
index b3b83d9..36ecf10 100644
--- a/pkg/collector/collect.go
+++ b/pkg/collector/collect.go
@@ -52,6 +52,9 @@ func CollectData(cmd *cobra.Command) error {
 	cm := configParams(lp, shellCmd)
 	nodeCommands := cmd.Flag("node-commands").Value.String()
 	commands, err := GetNodesCommands(nodeCommands, cm, nodeType)
+	if err != nil {
+		return err
+	}
 	if len(commands) == 0 {
 		return fmt.Errorf("spec not found")
 	}
@@ -103,6 +106,9 @@ func GetNodesCommands(nodeCommands string, configMap map[string]string, nodeType
 		}
 		err = yaml.Unmarshal([]byte(updatedContent), &specInfo)
 		if err != nil {
+			fmt.Println("COMMANDS: ", nodeCommands)
+			fmt.Println("CM: ", configMap)
+			fmt.Println("NODE TYPE: ", nodeType)
 			return nil, err
 		}
 		for _, c := range specInfo.Commands {
